// google-ads-node/build/lib/services.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-ignore Avoid no module definition defined warnings
const account_budget_proposal_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/account_budget_proposal_service_grpc_pb");
exports.AccountBudgetProposalServiceClient = account_budget_proposal_service_grpc_pb_1.AccountBudgetProposalServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const account_budget_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/account_budget_service_grpc_pb");
exports.AccountBudgetServiceClient = account_budget_service_grpc_pb_1.AccountBudgetServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_ad_asset_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_ad_asset_view_service_grpc_pb");
exports.AdGroupAdAssetViewServiceClient = ad_group_ad_asset_view_service_grpc_pb_1.AdGroupAdAssetViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_ad_label_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_ad_label_service_grpc_pb");
exports.AdGroupAdLabelServiceClient = ad_group_ad_label_service_grpc_pb_1.AdGroupAdLabelServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_ad_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_ad_service_grpc_pb");
exports.AdGroupAdServiceClient = ad_group_ad_service_grpc_pb_1.AdGroupAdServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_audience_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_audience_view_service_grpc_pb");
exports.AdGroupAudienceViewServiceClient = ad_group_audience_view_service_grpc_pb_1.AdGroupAudienceViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_bid_modifier_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_bid_modifier_service_grpc_pb");
exports.AdGroupBidModifierServiceClient = ad_group_bid_modifier_service_grpc_pb_1.AdGroupBidModifierServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_criterion_label_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_criterion_label_service_grpc_pb");
exports.AdGroupCriterionLabelServiceClient = ad_group_criterion_label_service_grpc_pb_1.AdGroupCriterionLabelServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_criterion_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_criterion_service_grpc_pb");
exports.AdGroupCriterionServiceClient = ad_group_criterion_service_grpc_pb_1.AdGroupCriterionServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_criterion_simulation_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_criterion_simulation_service_grpc_pb");
exports.AdGroupCriterionSimulationServiceClient = ad_group_criterion_simulation_service_grpc_pb_1.AdGroupCriterionSimulationServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_extension_setting_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_extension_setting_service_grpc_pb");
exports.AdGroupExtensionSettingServiceClient = ad_group_extension_setting_service_grpc_pb_1.AdGroupExtensionSettingServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_feed_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_feed_service_grpc_pb");
exports.AdGroupFeedServiceClient = ad_group_feed_service_grpc_pb_1.AdGroupFeedServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_label_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_label_service_grpc_pb");
exports.AdGroupLabelServiceClient = ad_group_label_service_grpc_pb_1.AdGroupLabelServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_service_grpc_pb");
exports.AdGroupServiceClient = ad_group_service_grpc_pb_1.AdGroupServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_group_simulation_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_group_simulation_service_grpc_pb");
exports.AdGroupSimulationServiceClient = ad_group_simulation_service_grpc_pb_1.AdGroupSimulationServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_parameter_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_parameter_service_grpc_pb");
exports.AdParameterServiceClient = ad_parameter_service_grpc_pb_1.AdParameterServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_schedule_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_schedule_view_service_grpc_pb");
exports.AdScheduleViewServiceClient = ad_schedule_view_service_grpc_pb_1.AdScheduleViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const ad_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/ad_service_grpc_pb");
exports.AdServiceClient = ad_service_grpc_pb_1.AdServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const age_range_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/age_range_view_service_grpc_pb");
exports.AgeRangeViewServiceClient = age_range_view_service_grpc_pb_1.AgeRangeViewServiceClient;
// @ts-ignore
const asset_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/asset_service_grpc_pb");
exports.AssetServiceClient = asset_service_grpc_pb_1.AssetServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const bidding_strategy_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/bidding_strategy_service_grpc_pb");
exports.BiddingStrategyServiceClient = bidding_strategy_service_grpc_pb_1.BiddingStrategyServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const billing_setup_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/billing_setup_service_grpc_pb");
exports.BillingSetupServiceClient = billing_setup_service_grpc_pb_1.BillingSetupServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const campaign_audience_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/campaign_audience_view_service_grpc_pb");
exports.CampaignAudienceViewServiceClient = campaign_audience_view_service_grpc_pb_1.CampaignAudienceViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const campaign_bid_modifier_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/campaign_bid_modifier_service_grpc_pb");
exports.CampaignBidModifierServiceClient = campaign_bid_modifier_service_grpc_pb_1.CampaignBidModifierServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const campaign_budget_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/campaign_budget_service_grpc_pb");
exports.CampaignBudgetServiceClient = campaign_budget_service_grpc_pb_1.CampaignBudgetServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const campaign_criterion_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/campaign_criterion_service_grpc_pb");
exports.CampaignCriterionServiceClient = campaign_criterion_service_grpc_pb_1.CampaignCriterionServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const campaign_criterion_simulation_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/campaign_criterion_simulation_service_grpc_pb");
exports.CampaignCriterionSimulationServiceClient = campaign_criterion_simulation_service_grpc_pb_1.CampaignCriterionSimulationServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const campaign_extension_setting_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/campaign_extension_setting_service_grpc_pb");
exports.CampaignExtensionSettingServiceClient = campaign_extension_setting_service_grpc_pb_1.CampaignExtensionSettingServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const campaign_feed_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/campaign_feed_service_grpc_pb");
exports.CampaignFeedServiceClient = campaign_feed_service_grpc_pb_1.CampaignFeedServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const campaign_label_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/campaign_label_service_grpc_pb");
exports.CampaignLabelServiceClient = campaign_label_service_grpc_pb_1.CampaignLabelServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const campaign_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/campaign_service_grpc_pb");
exports.CampaignServiceClient = campaign_service_grpc_pb_1.CampaignServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const campaign_shared_set_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/campaign_shared_set_service_grpc_pb");
exports.CampaignSharedSetServiceClient = campaign_shared_set_service_grpc_pb_1.CampaignSharedSetServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const carrier_constant_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/carrier_constant_service_grpc_pb");
exports.CarrierConstantServiceClient = carrier_constant_service_grpc_pb_1.CarrierConstantServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const change_status_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/change_status_service_grpc_pb");
exports.ChangeStatusServiceClient = change_status_service_grpc_pb_1.ChangeStatusServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const click_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/click_view_service_grpc_pb");
exports.ClickViewServiceClient = click_view_service_grpc_pb_1.ClickViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const conversion_action_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/conversion_action_service_grpc_pb");
exports.ConversionActionServiceClient = conversion_action_service_grpc_pb_1.ConversionActionServiceClient;
// @ts-ignore
const conversion_adjustment_upload_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/conversion_adjustment_upload_service_grpc_pb");
exports.ConversionAdjustmentUploadServiceClient = conversion_adjustment_upload_service_grpc_pb_1.ConversionAdjustmentUploadServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const conversion_upload_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/conversion_upload_service_grpc_pb");
exports.ConversionUploadServiceClient = conversion_upload_service_grpc_pb_1.ConversionUploadServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const custom_interest_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/custom_interest_service_grpc_pb");
exports.CustomInterestServiceClient = custom_interest_service_grpc_pb_1.CustomInterestServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const customer_client_link_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/customer_client_link_service_grpc_pb");
exports.CustomerClientLinkServiceClient = customer_client_link_service_grpc_pb_1.CustomerClientLinkServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const customer_client_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/customer_client_service_grpc_pb");
exports.CustomerClientServiceClient = customer_client_service_grpc_pb_1.CustomerClientServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const customer_extension_setting_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/customer_extension_setting_service_grpc_pb");
exports.CustomerExtensionSettingServiceClient = customer_extension_setting_service_grpc_pb_1.CustomerExtensionSettingServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const customer_feed_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/customer_feed_service_grpc_pb");
exports.CustomerFeedServiceClient = customer_feed_service_grpc_pb_1.CustomerFeedServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const customer_label_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/customer_label_service_grpc_pb");
exports.CustomerLabelServiceClient = customer_label_service_grpc_pb_1.CustomerLabelServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const customer_manager_link_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/customer_manager_link_service_grpc_pb");
exports.CustomerManagerLinkServiceClient = customer_manager_link_service_grpc_pb_1.CustomerManagerLinkServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const customer_negative_criterion_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/customer_negative_criterion_service_grpc_pb");
exports.CustomerNegativeCriterionServiceClient = customer_negative_criterion_service_grpc_pb_1.CustomerNegativeCriterionServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const customer_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/customer_service_grpc_pb");
exports.CustomerServiceClient = customer_service_grpc_pb_1.CustomerServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const display_keyword_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/display_keyword_view_service_grpc_pb");
exports.DisplayKeywordViewServiceClient = display_keyword_view_service_grpc_pb_1.DisplayKeywordViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const distance_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/distance_view_service_grpc_pb");
exports.DistanceViewServiceClient = distance_view_service_grpc_pb_1.DistanceViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const domain_category_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/domain_category_service_grpc_pb");
exports.DomainCategoryServiceClient = domain_category_service_grpc_pb_1.DomainCategoryServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const dynamic_search_ads_search_term_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/dynamic_search_ads_search_term_view_service_grpc_pb");
exports.DynamicSearchAdsSearchTermViewServiceClient = dynamic_search_ads_search_term_view_service_grpc_pb_1.DynamicSearchAdsSearchTermViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const extension_feed_item_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/extension_feed_item_service_grpc_pb");
exports.ExtensionFeedItemServiceClient = extension_feed_item_service_grpc_pb_1.ExtensionFeedItemServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const feed_item_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/feed_item_service_grpc_pb");
exports.FeedItemServiceClient = feed_item_service_grpc_pb_1.FeedItemServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const feed_item_target_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/feed_item_target_service_grpc_pb");
exports.FeedItemTargetServiceClient = feed_item_target_service_grpc_pb_1.FeedItemTargetServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const feed_mapping_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/feed_mapping_service_grpc_pb");
exports.FeedMappingServiceClient = feed_mapping_service_grpc_pb_1.FeedMappingServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const feed_placeholder_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/feed_placeholder_view_service_grpc_pb");
exports.FeedPlaceholderViewServiceClient = feed_placeholder_view_service_grpc_pb_1.FeedPlaceholderViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const feed_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/feed_service_grpc_pb");
exports.FeedServiceClient = feed_service_grpc_pb_1.FeedServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const gender_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/gender_view_service_grpc_pb");
exports.GenderViewServiceClient = gender_view_service_grpc_pb_1.GenderViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const geo_target_constant_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/geo_target_constant_service_grpc_pb");
exports.GeoTargetConstantServiceClient = geo_target_constant_service_grpc_pb_1.GeoTargetConstantServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const geographic_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/geographic_view_service_grpc_pb");
exports.GeographicViewServiceClient = geographic_view_service_grpc_pb_1.GeographicViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const google_ads_field_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/google_ads_field_service_grpc_pb");
exports.GoogleAdsFieldServiceClient = google_ads_field_service_grpc_pb_1.GoogleAdsFieldServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const google_ads_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/google_ads_service_grpc_pb");
exports.GoogleAdsServiceClient = google_ads_service_grpc_pb_1.GoogleAdsServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const group_placement_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/group_placement_view_service_grpc_pb");
exports.GroupPlacementViewServiceClient = group_placement_view_service_grpc_pb_1.GroupPlacementViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const hotel_group_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/hotel_group_view_service_grpc_pb");
exports.HotelGroupViewServiceClient = hotel_group_view_service_grpc_pb_1.HotelGroupViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const hotel_performance_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/hotel_performance_view_service_grpc_pb");
exports.HotelPerformanceViewServiceClient = hotel_performance_view_service_grpc_pb_1.HotelPerformanceViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const keyword_plan_ad_group_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/keyword_plan_ad_group_service_grpc_pb");
exports.KeywordPlanAdGroupServiceClient = keyword_plan_ad_group_service_grpc_pb_1.KeywordPlanAdGroupServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const keyword_plan_campaign_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/keyword_plan_campaign_service_grpc_pb");
exports.KeywordPlanCampaignServiceClient = keyword_plan_campaign_service_grpc_pb_1.KeywordPlanCampaignServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const keyword_plan_idea_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/keyword_plan_idea_service_grpc_pb");
exports.KeywordPlanIdeaServiceClient = keyword_plan_idea_service_grpc_pb_1.KeywordPlanIdeaServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const keyword_plan_keyword_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/keyword_plan_keyword_service_grpc_pb");
exports.KeywordPlanKeywordServiceClient = keyword_plan_keyword_service_grpc_pb_1.KeywordPlanKeywordServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const keyword_plan_negative_keyword_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/keyword_plan_negative_keyword_service_grpc_pb");
exports.KeywordPlanNegativeKeywordServiceClient = keyword_plan_negative_keyword_service_grpc_pb_1.KeywordPlanNegativeKeywordServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const keyword_plan_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/keyword_plan_service_grpc_pb");
exports.KeywordPlanServiceClient = keyword_plan_service_grpc_pb_1.KeywordPlanServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const keyword_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/keyword_view_service_grpc_pb");
exports.KeywordViewServiceClient = keyword_view_service_grpc_pb_1.KeywordViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const label_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/label_service_grpc_pb");
exports.LabelServiceClient = label_service_grpc_pb_1.LabelServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const language_constant_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/language_constant_service_grpc_pb");
exports.LanguageConstantServiceClient = language_constant_service_grpc_pb_1.LanguageConstantServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const location_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/location_view_service_grpc_pb");
exports.LocationViewServiceClient = location_view_service_grpc_pb_1.LocationViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const managed_placement_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/managed_placement_view_service_grpc_pb");
exports.ManagedPlacementViewServiceClient = managed_placement_view_service_grpc_pb_1.ManagedPlacementViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const media_file_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/media_file_service_grpc_pb");
exports.MediaFileServiceClient = media_file_service_grpc_pb_1.MediaFileServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const merchant_center_link_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/merchant_center_link_service_grpc_pb");
exports.MerchantCenterLinkServiceClient = merchant_center_link_service_grpc_pb_1.MerchantCenterLinkServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const mobile_app_category_constant_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/mobile_app_category_constant_service_grpc_pb");
exports.MobileAppCategoryConstantServiceClient = mobile_app_category_constant_service_grpc_pb_1.MobileAppCategoryConstantServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const mobile_device_constant_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/mobile_device_constant_service_grpc_pb");
exports.MobileDeviceConstantServiceClient = mobile_device_constant_service_grpc_pb_1.MobileDeviceConstantServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const mutate_job_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/mutate_job_service_grpc_pb");
exports.MutateJobServiceClient = mutate_job_service_grpc_pb_1.MutateJobServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const operating_system_version_constant_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/operating_system_version_constant_service_grpc_pb");
exports.OperatingSystemVersionConstantServiceClient = operating_system_version_constant_service_grpc_pb_1.OperatingSystemVersionConstantServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const parental_status_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/parental_status_view_service_grpc_pb");
exports.ParentalStatusViewServiceClient = parental_status_view_service_grpc_pb_1.ParentalStatusViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const payments_account_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/payments_account_service_grpc_pb");
exports.PaymentsAccountServiceClient = payments_account_service_grpc_pb_1.PaymentsAccountServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const product_bidding_category_constant_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/product_bidding_category_constant_service_grpc_pb");
exports.ProductBiddingCategoryConstantServiceClient = product_bidding_category_constant_service_grpc_pb_1.ProductBiddingCategoryConstantServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const product_group_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/product_group_view_service_grpc_pb");
exports.ProductGroupViewServiceClient = product_group_view_service_grpc_pb_1.ProductGroupViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const recommendation_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/recommendation_service_grpc_pb");
exports.RecommendationServiceClient = recommendation_service_grpc_pb_1.RecommendationServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const remarketing_action_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/remarketing_action_service_grpc_pb");
exports.RemarketingActionServiceClient = remarketing_action_service_grpc_pb_1.RemarketingActionServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const search_term_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/search_term_view_service_grpc_pb");
exports.SearchTermViewServiceClient = search_term_view_service_grpc_pb_1.SearchTermViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const shared_criterion_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/shared_criterion_service_grpc_pb");
exports.SharedCriterionServiceClient = shared_criterion_service_grpc_pb_1.SharedCriterionServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const shared_set_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/shared_set_service_grpc_pb");
exports.SharedSetServiceClient = shared_set_service_grpc_pb_1.SharedSetServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const shopping_performance_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/shopping_performance_view_service_grpc_pb");
exports.ShoppingPerformanceViewServiceClient = shopping_performance_view_service_grpc_pb_1.ShoppingPerformanceViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const topic_constant_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/topic_constant_service_grpc_pb");
exports.TopicConstantServiceClient = topic_constant_service_grpc_pb_1.TopicConstantServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const topic_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/topic_view_service_grpc_pb");
exports.TopicViewServiceClient = topic_view_service_grpc_pb_1.TopicViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const user_interest_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/user_interest_service_grpc_pb");
exports.UserInterestServiceClient = user_interest_service_grpc_pb_1.UserInterestServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const user_list_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/user_list_service_grpc_pb");
exports.UserListServiceClient = user_list_service_grpc_pb_1.UserListServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const user_location_view_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/user_location_view_service_grpc_pb");
exports.UserLocationViewServiceClient = user_location_view_service_grpc_pb_1.UserLocationViewServiceClient;
// @ts-ignore Avoid no module definition defined warnings
const video_service_grpc_pb_1 = require("../protos/google/ads/googleads/v2/services/video_service_grpc_pb");
exports.VideoServiceClient = video_service_grpc_pb_1.VideoServiceClient;
