// google-ads-node/build/lib/utils.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const protobufHelpers = __importStar(require("google-protobuf/google/protobuf/field_mask_pb"));
const lodash_camelcase_1 = __importDefault(require("lodash.camelcase"));
const lodash_get_1 = __importDefault(require("lodash.get"));
const lodash_set_1 = __importDefault(require("lodash.set"));
const lodash_snakecase_1 = __importDefault(require("lodash.snakecase"));
const structs = __importStar(require("./struct"));
const NON_MUTABLE_METHOD_PREFIXES = ["Get", "List", "Generate", "Search"];
// Based on https://github.com/leaves4j/grpc-promisify/blob/master/src/index.js
function promisifyServiceClient(client) {
    Object.keys(Object.getPrototypeOf(client)).forEach((functionName) => {
        if (functionName.charAt(0) === "$" || functionName.charAt(0) === "_") {
            return;
        }
        const originalFunction = client[functionName];
        client[functionName] = (request, callback) => {
            if (callback && typeof callback === "function") {
                return originalFunction.call(client, request, (error, response) => {
                    callback(error, response);
                });
            }
            return new Promise((resolve, reject) => {
                originalFunction.call(client, request, (error, response) => {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(response);
                    }
                });
            });
        };
    });
}
exports.promisifyServiceClient = promisifyServiceClient;
function formatCallResults(resultsList, fieldMask) {
    const parsedResults = [];
    if (fieldMask) {
        const { pathsList } = fieldMask;
        for (const result of resultsList) {
            const parsedRow = parseNestedEntities(result, pathsList);
            parsedResults.push(parsedRow);
        }
    }
    else {
        for (const result of resultsList) {
            const parsedRow = parseNestedEntitiesNoPath(result);
            parsedResults.push(parsedRow);
        }
    }
    return parsedResults;
}
exports.formatCallResults = formatCallResults;
function convertToProtoFormat(data, type, resource_name, nested_path = "") {
    const pb = {};
    const struct = structs[resource_name];
    if (!struct) {
        throw new Error(`Key "${nested_path.replace(".", "")}" not found in resource "${resource_name}"`);
    }
    for (const key of Object.keys(data)) {
        const displayKey = lodash_camelcase_1.default(key);
        const value = data[key];
        /* Resource names are string values, not a protobuf string instance, so just set the string value */
        if (displayKey === "resourceName") {
            pb[displayKey] = value;
            continue;
        }
        /* Build array of proto values */
        if (Array.isArray(value)) {
            pb[displayKey] = value.map((v) => {
                return unroll(v);
            });
            continue;
        }
        pb[displayKey] = unroll(value);
        function unroll(v) {
            return typeof v === "object"
                ? convertToProtoFormat(v, type, resource_name, `${nested_path}.${key}`)
                : toProtoValueFormat(v, struct, `${nested_path}.${key}`.replace(".", ""));
        }
    }
    return pb;
}
exports.convertToProtoFormat = convertToProtoFormat;
function toProtoValueFormat(value, struct, nested_path) {
    const valueType = lodash_get_1.default(struct, nested_path);
    if (!valueType) {
        throw new Error(`Attempted to set value "${value}" on invalid path "${nested_path}" in resource`);
    }
    if (valueType.startsWith("enum_")) {
        return value;
    }
    return {
        value,
    };
}
/* This is different to lodash.camelCase as it leaves any periods (".") */
function convertPathToCamelCase(str) {
    return str.replace(/([-_][a-z])/gi, $1 => {
        return $1
            .toUpperCase()
            .replace("-", "")
            .replace("_", "");
    });
}
function parseNestedEntitiesNoPath(data, _structs = structs) {
    if (typeof data === "string" || typeof data === "number" || typeof data === "boolean") {
        return data;
    }
    const findMatchingStruct = (key) => {
        const capitalcase_key = key.charAt(0).toUpperCase() + key.slice(1);
        const snakecase_key = lodash_snakecase_1.default(key);
        // We need both cases because the structs.ts file exports resources in CapitalCase,
        // but the keys inside each resource are in snake_case. parseNestedEntitiesNoPath() is
        // recursive, so we're not sure which one we'll need.
        // @ts-ignore
        return _structs[capitalcase_key] || _structs[snakecase_key];
    };
    const finalObject = {};
    Object.keys(data).map(key => {
        let displayKey = key;
        /*
          This section regarding "matching structs" is designed to correctly
          trim keys that end in "List". Some of those keys are legit (such as userList)
          while others are bogus (such as pointsList). The bogus ones need trimming so
          that the final result returned to the user matches the format specified in the docs.
    
          This adds other layer of recusive complexity to this function, and I'm open to improvements.
        */
        let matching_struct = findMatchingStruct(displayKey);
        if (!matching_struct && key.endsWith("List")) {
            // Trim "List" only if bogus key
            displayKey = key.split("List")[0];
            matching_struct = findMatchingStruct(displayKey);
        }
        const entity = data[key];
        const isObject = typeof entity === "object";
        const isUndefined = typeof entity === "undefined";
        const isArray = Array.isArray(entity);
        // It needs its "value" unwrapped if it is an object and has "value" as its only key.
        const isValue = isObject
            ? entity.hasOwnProperty("value") && Object.keys(entity).length === 1
            : false;
        if (isUndefined) {
            return;
        }
        if (isArray) {
            finalObject[displayKey] = entity.map((item) => {
                // @ts-ignore
                const parsed = parseNestedEntitiesNoPath({ item }, { item: matching_struct });
                return parsed.item;
            });
        }
        else if (isValue) {
            finalObject[displayKey] = entity.value;
        }
        else if (isObject) {
            finalObject[displayKey] = parseNestedEntitiesNoPath(entity, matching_struct);
        }
        else {
            finalObject[displayKey] = entity;
        }
    });
    return finalObject;
}
// This function first parses the data without regard for the passed props,
// then plucks out the props that it actually cares about.
function parseNestedEntities(data, props) {
    const parsed_data = parseNestedEntitiesNoPath(data);
    const final_object = {};
    for (let path of props) {
        path = convertPathToCamelCase(path);
        // Pluck resource name if available
        const path_elements_trimmed = path.split(".");
        while (path_elements_trimmed.pop()) {
            path_elements_trimmed.push("resourceName");
            const rn_path = path_elements_trimmed.join(".");
            path_elements_trimmed.pop();
            if (lodash_get_1.default(parsed_data, rn_path)) {
                lodash_set_1.default(final_object, rn_path, lodash_get_1.default(parsed_data, rn_path));
            }
        }
        // And pluck the field specified in path
        const plucked_value = lodash_get_1.default(parsed_data, path);
        if (typeof plucked_value !== "undefined") {
            lodash_set_1.default(final_object, path, plucked_value);
        }
    }
    return final_object;
}
function recursiveFieldMaskSearch(data) {
    const paths = [];
    for (const key of Object.keys(data)) {
        if (key === "resource_name") {
            continue;
        }
        const value = data[key];
        if (typeof value === "object" && !Array.isArray(value)) {
            const children = recursiveFieldMaskSearch(value);
            for (const child of children) {
                paths.push(`${key}.${child}`);
            }
            continue;
        }
        paths.push(key);
    }
    return paths;
}
function getFieldMask(data) {
    const fieldMask = new protobufHelpers.FieldMask();
    const paths = recursiveFieldMaskSearch(data);
    fieldMask.setPathsList(paths);
    return fieldMask;
}
exports.getFieldMask = getFieldMask;
function getErrorLocationPath(location) {
    if (!location || !location.hasOwnProperty("fieldPathElementsList")) {
        return "";
    }
    if (!Array.isArray(location.fieldPathElementsList) && location.fieldPathElementsList.length < 1) {
        return "";
    }
    const { fieldPathElementsList } = location;
    const paths = fieldPathElementsList.map((field) => {
        let path = field.fieldName;
        if (field.index && field.index.hasOwnProperty("value")) {
            path += `[${field.index.value}]`;
        }
        return path;
    });
    return paths.join(".");
}
exports.getErrorLocationPath = getErrorLocationPath;
function isMutationRequest(interceptMessage) {
    var _a, _b;
    let isMutation = true;
    if ((_b = (_a = interceptMessage) === null || _a === void 0 ? void 0 : _a.method_definition) === null || _b === void 0 ? void 0 : _b.path) {
        const { path } = interceptMessage.method_definition;
        for (const prefix of NON_MUTABLE_METHOD_PREFIXES) {
            if (path.includes(prefix)) {
                isMutation = false;
                break;
            }
        }
    }
    return isMutation;
}
exports.isMutationRequest = isMutationRequest;
function safeguardMutationProtobufRequest(message, next) {
    var _a, _b, _c;
    // Force validation only for requests if the setValidateOnly method exists
    if ((_a = message) === null || _a === void 0 ? void 0 : _a.setValidateOnly) {
        message.setValidateOnly(true);
        next(message);
        // If the request doesn't support validateOnly, we attempt to clear the operations list
    }
    else if ((_b = message) === null || _b === void 0 ? void 0 : _b.clearOperationsList) {
        message.clearOperationsList();
        next(message);
        // Some request operations are called mutate_operations
    }
    else if ((_c = message) === null || _c === void 0 ? void 0 : _c.clearMutateOperationsList) {
        message.clearMutateOperationsList();
        next(message);
    }
    else {
        // Otherwise, we just give up to prevent any unwanted mutations (this throws an error, shouldn't happen)
        throw new Error(`Prevent mutations mode is enabled, but the request couldn't be safeguarded, giving up.`);
    }
}
exports.safeguardMutationProtobufRequest = safeguardMutationProtobufRequest;
