// @smooth-ui/core-sc/dist/smooth-ui-core-sc.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('prop-types'), require('styled-components'), require('polished'), require('react'), require('react-dom'), require('popper.js')) :
  typeof define === 'function' && define.amd ? define(['exports', 'prop-types', 'styled-components', 'polished', 'react', 'react-dom', 'popper.js'], factory) :
  (global = global || self, factory(global.smoothUI = {}, global.PropTypes, global.styled, global.polished, global.React, global.ReactDom, global.Popper));
}(this, function (exports, PropTypes, styled, polished, React, ReactDOM, PopperJs) { 'use strict';

  var PropTypes__default = 'default' in PropTypes ? PropTypes['default'] : PropTypes;
  var styled__default = 'default' in styled ? styled['default'] : styled;
  var React__default = 'default' in React ? React['default'] : React;
  var ReactDOM__default = 'default' in ReactDOM ? ReactDOM['default'] : ReactDOM;
  PopperJs = PopperJs && PopperJs.hasOwnProperty('default') ? PopperJs['default'] : PopperJs;

  var wrapEvent = function wrapEvent(handler, cb) {
    return function (event) {
      if (handler) {
        handler(event);
      }

      if (!event.defaultPrevented) {
        cb(event);
      }
    };
  };
  var stopPropagation = function stopPropagation(event) {
    return event.stopPropagation();
  };

  var mediaMinWidth = function mediaMinWidth(value) {
    return "@media (min-width: " + value + "px)";
  };
  var mediaMaxWidth = function mediaMaxWidth(value) {
    return "@media (max-width: " + value + "px)";
  };
  var mediaBetweenWidth = function mediaBetweenWidth(min, max) {
    return "@media (min-width: " + min + "px) and (max-width: " + max + "px)";
  };

  var commonjsGlobal = typeof globalThis !== 'undefined' ? globalThis : typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {};

  function createCommonjsModule(fn, module) {
  	return module = { exports: {} }, fn(module, module.exports), module.exports;
  }

  var _extends_1 = createCommonjsModule(function (module) {
  function _extends() {
    module.exports = _extends = Object.assign || function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];

        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }

      return target;
    };

    return _extends.apply(this, arguments);
  }

  module.exports = _extends;
  });

  var is = function is(n) {
    return n !== undefined && n !== null;
  };
  var num = function num(n) {
    return typeof n === 'number' && !Number.isNaN(n);
  };
  var string = function string(n) {
    return typeof n === 'string';
  };
  var func = function func(n) {
    return typeof n === 'function';
  };
  var get = function get(obj) {
    for (var _len = arguments.length, paths = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      paths[_key - 1] = arguments[_key];
    }

    return paths.join('.').split('.').reduce(function (a, b) {
      return a && is(a[b]) ? a[b] : undefined;
    }, obj);
  };
  var cascade = function cascade(fn) {
    if (!func(fn)) return fn;

    for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
      args[_key2 - 1] = arguments[_key2];
    }

    var next = fn.apply(void 0, args);
    return cascade.apply(void 0, [next].concat(args));
  };
  var omit = function omit(obj, fields) {
    var shallowCopy = _extends_1({}, obj);

    for (var i = 0; i < fields.length; i += 1) {
      var key = fields[i];
      delete shallowCopy[key];
    }

    return shallowCopy;
  };

  var getTheme = function getTheme(_ref) {
    var _ref$theme = _ref.theme,
        theme = _ref$theme === void 0 ? {} : _ref$theme;
    return theme;
  };
  var thd = function thd(name, defaultValue, transform) {
    var getFromProps = function getFromProps(p) {
      var value = cascade(function (p) {
        return get(getTheme(p), name);
      }, p);

      if (value === undefined) {
        return cascade(defaultValue, p);
      }

      if (transform) {
        return transform(value);
      }

      return value;
    };

    getFromProps["default"] = defaultValue;
    return getFromProps;
  };
  var th = function th(name, transform) {
    return function (p) {
      var value = cascade(function (p) {
        return get(getTheme(p), name);
      }, p);

      if (!is(value)) {
        throw new Error(name + " not found in theme");
      }

      if (transform) {
        return transform(value);
      }

      return value;
    };
  };
  var mixin = function mixin(name, defaultFn) {
    var mixinFn = function mixinFn() {
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return function (p) {
        var fn = get(getTheme(p), name) || defaultFn;
        return fn.apply(void 0, args)(p);
      };
    };

    mixinFn["default"] = defaultFn;
    return mixinFn;
  };
  function definitionsToTheme(values) {
    var theme = {};
    var themingKeys = Object.keys(values);

    for (var i = 0; i < themingKeys.length; i++) {
      var key = themingKeys[i];
      var value = values[key];
      theme[key] = value && value["default"] !== undefined ? value["default"] : value;
    }

    return theme;
  }

  var unit = function unit(_unit) {
    return function (value) {
      return num(value) ? "" + value + _unit : value;
    };
  };
  var negative = function negative(value) {
    if (!value) return value;
    if (num(value)) return -value;
    if (string(value)) return "-" + value;
    var obj = {};
    var keys = Object.keys(value);

    for (var i = 0; i < keys.length; i++) {
      var key = keys[i];
      obj[key] = negative(value[key]);
    }

    return obj;
  };
  var px =
  /*#__PURE__*/
  unit('px');
  var calc = function calc(value, fn) {
    var _String$match = String(value).match(/([\d.]+)\s*(.*)/),
        num = _String$match[1],
        unit = _String$match[2];

    return "" + fn(Number(num)) + (unit || 'px');
  };

  var black =
  /*#__PURE__*/
  thd('black', '#000');
  var white =
  /*#__PURE__*/
  thd('white', '#fff');
  var gray100 =
  /*#__PURE__*/
  thd('gray100', '#f8f9fa');
  var gray200 =
  /*#__PURE__*/
  thd('gray200', '#e9ecef');
  var gray300 =
  /*#__PURE__*/
  thd('gray300', '#dee2e6');
  var gray400 =
  /*#__PURE__*/
  thd('gray400', '#ced4da');
  var gray500 =
  /*#__PURE__*/
  thd('gray500', '#adb5bd');
  var gray600 =
  /*#__PURE__*/
  thd('gray600', '#6c757d');
  var gray700 =
  /*#__PURE__*/
  thd('gray700', '#495057');
  var gray800 =
  /*#__PURE__*/
  thd('gray800', '#343a40');
  var gray900 =
  /*#__PURE__*/
  thd('gray900', '#212529');
  var blue =
  /*#__PURE__*/
  thd('blue', '#007bff');
  var indigo =
  /*#__PURE__*/
  thd('indigo', '#6610f2');
  var purple =
  /*#__PURE__*/
  thd('purple', '#6f42c1');
  var pink =
  /*#__PURE__*/
  thd('pink', '#e83e8c');
  var red =
  /*#__PURE__*/
  thd('red', '#dc3545');
  var brick =
  /*#__PURE__*/
  thd('brick', '#bd4932');
  var orange =
  /*#__PURE__*/
  thd('orange', '#fd7e14');
  var yellow =
  /*#__PURE__*/
  thd('yellow', '#ffc107');
  var green =
  /*#__PURE__*/
  thd('green', '#28a745');
  var teal =
  /*#__PURE__*/
  thd('teal', '#20c997');
  var cyan =
  /*#__PURE__*/
  thd('cyan', '#17a2b8');
  var primary =
  /*#__PURE__*/
  thd('primary', brick);
  var secondary =
  /*#__PURE__*/
  thd('secondary', gray600);
  var success =
  /*#__PURE__*/
  thd('success', green);
  var info =
  /*#__PURE__*/
  thd('info', cyan);
  var warning =
  /*#__PURE__*/
  thd('warning', yellow);
  var danger =
  /*#__PURE__*/
  thd('danger', red);
  var light =
  /*#__PURE__*/
  thd('light', gray100);
  var dark =
  /*#__PURE__*/
  thd('dark', gray800);
  var primaryLight =
  /*#__PURE__*/
  thd('primaryLight', function (p) {
    return polished.lighten(0.3, primary(p));
  });
  var secondaryLight =
  /*#__PURE__*/
  thd('secondaryLight', function (p) {
    return polished.lighten(0.3, secondary(p));
  });
  var colorVariants =
  /*#__PURE__*/
  thd('colorVariants', ['primary', 'secondary', 'success', 'danger', 'warning', 'info', 'light', 'dark']);
  var colors =
  /*#__PURE__*/
  thd('colors', {
    black: black,
    white: white,
    gray100: gray100,
    gray200: gray200,
    gray300: gray300,
    gray400: gray400,
    gray500: gray500,
    gray600: gray600,
    gray700: gray700,
    gray800: gray800,
    gray900: gray900,
    blue: blue,
    indigo: indigo,
    purple: purple,
    pink: pink,
    red: red,
    brick: brick,
    orange: orange,
    yellow: yellow,
    green: green,
    teal: teal,
    cyan: cyan,
    primary: primary,
    secondary: secondary,
    success: success,
    info: info,
    warning: warning,
    danger: danger,
    light: light,
    dark: dark
  });
  var colorInterval =
  /*#__PURE__*/
  thd('colorInterval', 0.08);
  var colorVariant =
  /*#__PURE__*/
  mixin('colorVariant', function (variant) {
    return function (p) {
      var thValue = thd(variant)(p);
      if (thValue) return thValue;
      var colorValue = colors(p)[variant];
      if (colorValue) return colorValue(p);
      return variant;
    };
  });
  var colorLevel =
  /*#__PURE__*/
  mixin('colorLevel', function (color, level) {
    return function (p) {
      var baseColor = level > 0 ? black(p) : white(p);
      var absLevel = Math.abs(level);
      return polished.mix(absLevel * colorInterval(p), baseColor, color);
    };
  });
  var yiqContrastedThreshold =
  /*#__PURE__*/
  thd('yiqContrastedThreshold', 150);
  var yikTextDark =
  /*#__PURE__*/
  thd('yikTextDark', '#111');
  var yikTextLight =
  /*#__PURE__*/
  thd('yikTextLight', '#fff');
  var colorYik =
  /*#__PURE__*/
  mixin('colorYik', function (color) {
    return function (p) {
      var _parseToRgb = polished.parseToRgb(color),
          r = _parseToRgb.red,
          g = _parseToRgb.green,
          b = _parseToRgb.blue;

      var yik = (r * 299 + g * 587 + b * 114) / 1000;
      return yik >= yiqContrastedThreshold(p) ? yikTextDark(p) : yikTextLight(p);
    };
  });

  var alertPaddingY =
  /*#__PURE__*/
  thd('alertPaddingY', '.75rem');
  var alertPaddingX =
  /*#__PURE__*/
  thd('alertPaddingX', '1.25rem');
  var alertMarginBottom =
  /*#__PURE__*/
  thd('alertMarginBottom', '1rem');
  var alertColorLevel =
  /*#__PURE__*/
  thd('alertColorLevel', 6);
  var alertBgLevel =
  /*#__PURE__*/
  thd('alertBgLevel', -10);
  var alertBorderLevel =
  /*#__PURE__*/
  thd('alertBorderLevel', -9);
  var alertVariant =
  /*#__PURE__*/
  mixin('alertVariant', function (variant) {
    return function (p) {
      var variantColor = colorVariant(variant)(p);
      var color = colorLevel(variantColor, alertColorLevel(p))(p);
      var bgColor = colorLevel(variantColor, alertBgLevel(p))(p);
      var borderColor = colorLevel(variantColor, alertBorderLevel(p))(p);
      var hrColor = polished.darken(0.05, color);
      return styled.css(["color:", ";background-color:", ";border-color:", ";hr{border-top-color:", ";}"], color, bgColor, borderColor, hrColor);
    };
  });

  var borderRadius =
  /*#__PURE__*/
  thd('borderRadius', '.25rem');
  var borderRadiusSm =
  /*#__PURE__*/
  thd('borderRadiusSm', '.2rem');
  var borderRadiusLg =
  /*#__PURE__*/
  thd('borderRadiusLg', '.3rem');
  var borderWidth =
  /*#__PURE__*/
  thd('borderWidth', '1px');

  var baseFocus =
  /*#__PURE__*/
  mixin('baseFocus', function (color) {
    return function () {
      return styled.css(["outline:0;border-color:", ";box-shadow:0 0 2px ", ";"], polished.lighten(0.25, color), polished.transparentize(0.1, color));
    };
  });
  var controlFocus =
  /*#__PURE__*/
  mixin('controlFocus', function (color) {
    return function () {
      return styled.css(["box-shadow:0 0 0 0.2rem ", ";"], polished.transparentize(0.75, color));
    };
  });

  var fontSizes =
  /*#__PURE__*/
  thd('fontSizes', [12, 14, 16, 20, 24, 32, 48, 64, 72]);
  var fontFamily =
  /*#__PURE__*/
  thd('fontFamily', '-apple-system,system-ui,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,sans-serif');
  var fontSizeBase =
  /*#__PURE__*/
  thd('fontSizeBase', '1rem');
  var fontSizeSm =
  /*#__PURE__*/
  thd('fontSizeSm', '1rem', function (s) {
    return polished.modularScale(-1, s);
  });
  var fontSizeLg =
  /*#__PURE__*/
  thd('fontSizeLg', '1rem', function (s) {
    return polished.modularScale(1, s);
  });
  var fontWeightLight =
  /*#__PURE__*/
  thd('fontWeightLight', 300);
  var fontWeightNormal =
  /*#__PURE__*/
  thd('fontWeightNormal', 400);
  var fontWeightBold =
  /*#__PURE__*/
  thd('fontWeightBold', 700);
  var lineHeightBase =
  /*#__PURE__*/
  thd('lineHeightBase', 1.5);
  var lineHeightSm =
  /*#__PURE__*/
  thd('lineHeightSm', 1.5);
  var lineHeightLg =
  /*#__PURE__*/
  thd('lineHeightLg', 1.5);

  var inputBtnPaddingY =
  /*#__PURE__*/
  thd('inputBtnPaddingY', '.375rem');
  var inputBtnPaddingX =
  /*#__PURE__*/
  thd('inputBtnPaddingX', '.75rem');
  var inputBtnLineHeight =
  /*#__PURE__*/
  thd('inputBtnLineHeight', lineHeightBase);
  var inputBtnPaddingYSm =
  /*#__PURE__*/
  thd('inputBtnPaddingYSm', '.25rem');
  var inputBtnPaddingXSm =
  /*#__PURE__*/
  thd('inputBtnPaddingXSm', '.5rem');
  var inputBtnLineHeightSm =
  /*#__PURE__*/
  thd('inputBtnLineHeightSm', lineHeightSm);
  var inputBtnPaddingYLg =
  /*#__PURE__*/
  thd('inputBtnPaddingYLg', '.5rem');
  var inputBtnPaddingXLg =
  /*#__PURE__*/
  thd('inputBtnPaddingXLg', '1rem');
  var inputBtnLineHeightLg =
  /*#__PURE__*/
  thd('inputBtnLineHeightLg', lineHeightLg);
  var inputPaddingY =
  /*#__PURE__*/
  thd('inputPaddingY', inputBtnPaddingY);
  var inputPaddingX =
  /*#__PURE__*/
  thd('inputPaddingX', inputBtnPaddingX);
  var inputLineHeight =
  /*#__PURE__*/
  thd('inputLineHeight', inputBtnLineHeight);
  var inputPaddingYSm =
  /*#__PURE__*/
  thd('inputPaddingYSm', inputBtnPaddingYSm);
  var inputPaddingXSm =
  /*#__PURE__*/
  thd('inputPaddingXSm', inputBtnPaddingXSm);
  var inputLineHeightSm =
  /*#__PURE__*/
  thd('inputLineHeightSm', inputBtnLineHeightSm);
  var inputPaddingYLg =
  /*#__PURE__*/
  thd('inputPaddingYLg', inputBtnPaddingYLg);
  var inputPaddingXLg =
  /*#__PURE__*/
  thd('inputPaddingXLg', inputBtnPaddingXLg);
  var inputLineHeightLg =
  /*#__PURE__*/
  thd('inputLineHeightLg', inputBtnLineHeightLg);
  var inputBorderWidth =
  /*#__PURE__*/
  thd('inputBorderWidth', borderWidth);
  var inputBorderColor =
  /*#__PURE__*/
  thd('inputBorderColor', gray300);
  var inputBgColor =
  /*#__PURE__*/
  thd('inputBgColor', white);
  var inputDisabledBgColor =
  /*#__PURE__*/
  thd('inputDisabledBgColor', gray100);
  var inputDisabledText =
  /*#__PURE__*/
  thd('inputDisabledText', gray600);
  var inputPlaceholderText =
  /*#__PURE__*/
  thd('inputPlaceholderText', gray600);
  var inputTextColor =
  /*#__PURE__*/
  thd('inputTextColor', gray900);

  var btnPaddingY =
  /*#__PURE__*/
  thd('btnPaddingY', inputBtnPaddingY);
  var btnPaddingX =
  /*#__PURE__*/
  thd('btnPaddingX', inputBtnPaddingX);
  var btnLineHeight =
  /*#__PURE__*/
  thd('btnLineHeight', inputBtnLineHeight);
  var btnPaddingYSm =
  /*#__PURE__*/
  thd('btnPaddingYSm', inputBtnPaddingYSm);
  var btnPaddingXSm =
  /*#__PURE__*/
  thd('btnPaddingXSm', inputBtnPaddingXSm);
  var btnLineHeightSm =
  /*#__PURE__*/
  thd('btnLineHeightSm', inputBtnLineHeightSm);
  var btnPaddingYLg =
  /*#__PURE__*/
  thd('btnPaddingYLg', inputBtnPaddingYLg);
  var btnPaddingXLg =
  /*#__PURE__*/
  thd('btnPaddingXLg', inputBtnPaddingXLg);
  var btnLineHeightLg =
  /*#__PURE__*/
  thd('btnLineHeightLg', inputBtnLineHeightLg);
  var btnBorderWidth =
  /*#__PURE__*/
  thd('btnBorderWidth', 0);
  var btnDisabledOpacity =
  /*#__PURE__*/
  thd('btnDisabledOpacity', 0.8);
  var btnVariant =
  /*#__PURE__*/
  mixin('btnVariant', function (variant) {
    return function (p) {
      var color = colorVariant(variant)(p);
      return styled.css(["color:", ";background-color:", ";&:focus{", ";}&:not(:disabled):hover,&:not(:disabled):active{background-color:", ";}"], colorYik(color)(p), color, baseFocus(color)(p), polished.darken(0.05, color));
    };
  });

  var gridColumns =
  /*#__PURE__*/
  thd('gridColumns', 12);
  var gridGutter =
  /*#__PURE__*/
  thd('gridGutter', 1);
  var gridMaxWidths =
  /*#__PURE__*/
  thd('gridMaxWidths', {
    sm: '540px',
    md: '720px',
    lg: '960px',
    xl: '1140px'
  });

  var headingsMarginBottom =
  /*#__PURE__*/
  thd('headingsMarginBottom', '.5rem');
  var headingsFontFamily =
  /*#__PURE__*/
  thd('headingsFontFamily', fontFamily);
  var headingsFontWeight =
  /*#__PURE__*/
  thd('headingsFontWeight', 500);
  var headingsLineHeight =
  /*#__PURE__*/
  thd('headingsLineHeight', 1.2);
  var headingsColor =
  /*#__PURE__*/
  thd('headingsColor', 'inherit');
  var h1FontSize =
  /*#__PURE__*/
  thd('h1FontSize', '2.5rem');
  var h2FontSize =
  /*#__PURE__*/
  thd('h2FontSize', '2rem');
  var h3FontSize =
  /*#__PURE__*/
  thd('h3FontSize', '1.75rem');
  var h4FontSize =
  /*#__PURE__*/
  thd('h4FontSize', '1.5rem');
  var h5FontSize =
  /*#__PURE__*/
  thd('h5FontSize', '1.25rem');
  var h6FontSize =
  /*#__PURE__*/
  thd('h6FontSize', '1rem');
  var display1Size =
  /*#__PURE__*/
  thd('display1Size', '6rem');
  var display2Size =
  /*#__PURE__*/
  thd('display2Size', '5.5rem');
  var display3Size =
  /*#__PURE__*/
  thd('display3Size', '4.4rem');
  var display4Size =
  /*#__PURE__*/
  thd('display4Size', '3.5rem');
  var display1Weight =
  /*#__PURE__*/
  thd('display1Weight', 300);
  var display2Weight =
  /*#__PURE__*/
  thd('display2Weight', 300);
  var display3Weight =
  /*#__PURE__*/
  thd('display3Weight', 300);
  var display4Weight =
  /*#__PURE__*/
  thd('display4Weight', 300);
  var displayLineHeight =
  /*#__PURE__*/
  thd('displayLineHeight', headingsLineHeight);

  var modalBackdropBg =
  /*#__PURE__*/
  thd('modalBackdropBg', 'rgba(0, 0, 0, 0.5)');
  var modalTransitionDuration =
  /*#__PURE__*/
  thd('modalTransitionDuration', 300);
  var modalInnerPadding =
  /*#__PURE__*/
  thd('modalInnerPadding', '1rem');
  var modalDialogMargin =
  /*#__PURE__*/
  thd('modalDialogMargin', '0.5rem');
  var modalDialogMarginYSmUp =
  /*#__PURE__*/
  thd('modalDialogMarginYSmUp', '1.75rem');
  var modalContentBg =
  /*#__PURE__*/
  thd('modalContentBg', white);
  var modalContentBorderWidth =
  /*#__PURE__*/
  thd('modalContentBorderWidth', borderWidth);
  var modalContentBorderColor =
  /*#__PURE__*/
  thd('modalContentBorderColor', black, function (color) {
    return polished.transparentize(0.8, color);
  });
  var modalContentBorderRadius =
  /*#__PURE__*/
  thd('modalContentBorderRadius', borderRadiusLg);
  var modalContentBoxShadowXs =
  /*#__PURE__*/
  thd('modalContentBoxShadowXs', function (p) {
    return styled.css(["box-shadow:0 0.25rem 0.5rem ", ";"], polished.transparentize(0.8, black(p)));
  });
  var modalContentBoxShadowSmUp =
  /*#__PURE__*/
  thd('modalContentBoxShadowSmUp', function (p) {
    return styled.css(["box-shadow:0 0.5rem 1rem ", ";"], polished.transparentize(0.8, black(p)));
  });
  var modalHeaderBorderColor =
  /*#__PURE__*/
  thd('modalHeaderBorderColor', gray200);
  var modalFooterBorderColor =
  /*#__PURE__*/
  thd('modalFooterBorderColor', modalHeaderBorderColor);
  var modalHeaderBorderWidth =
  /*#__PURE__*/
  thd('modalHeaderBorderWidth', modalContentBorderWidth);
  var modalFooterBorderWidth =
  /*#__PURE__*/
  thd('modalFooterBorderWidth', modalHeaderBorderWidth);

  var breakpoints =
  /*#__PURE__*/
  thd('breakpoints', {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200
  });
  /**
   * Minimum breakpoint width.
   * Null for the smallest breakpoint.
   */

  var getBreakpointMin = function getBreakpointMin(name, p) {
    var bks = breakpoints(p);
    var breakPoint = bks[name];
    return breakPoint !== 0 ? breakPoint : null;
  };
  /**
   * Maximum breakpoint width. Null for the largest (last) breakpoint.
   * The maximum value is calculated as the minimum of the next one less 0.02px
   * to work around the limitations of `min-` and `max-` prefixes and viewports with fractional widths.
   * See https://www.w3.org/TR/mediaqueries-4/#mq-min-max
   * Uses 0.02px rather than 0.01px to work around a current rounding bug in Safari.
   * See https://bugs.webkit.org/show_bug.cgi?id=178261
   */


  var getBreakpointMax = function getBreakpointMax(name, p) {
    var bks = breakpoints(p);
    var breakPoint = bks[name];
    return breakPoint !== 0 ? breakPoint - 0.02 : null;
  };

  var up =
  /*#__PURE__*/
  mixin('up', function (name, code) {
    return function (p) {
      var value = getBreakpointMin(name, p);
      if (value === null) return code;
      return styled.css(["", "{", ";}"], mediaMinWidth(value), code);
    };
  });
  var down =
  /*#__PURE__*/
  mixin('down', function (name, code) {
    return function (p) {
      var value = getBreakpointMax(name, p);
      if (value === null) return null;
      return styled.css(["", "{", ";}"], mediaMaxWidth(value), code);
    };
  });
  var between =
  /*#__PURE__*/
  mixin('between', function (lower, upper, code) {
    return function (p) {
      var min = getBreakpointMin(lower, p);
      var max = getBreakpointMax(upper, p);

      if (min !== null && max !== null) {
        return styled.css(["", "{", ";}"], mediaBetweenWidth(min, max), code);
      }

      if (max === null) return up(lower, code)(p);
      if (min === null) return down(upper, code)(p);
      return null;
    };
  });

  var spaces =
  /*#__PURE__*/
  thd('spaces', [0, 8, 16, 24, 32, 40, 48, 56, 64, 72]);

  var safeTransitionAttrs = ['color', 'border-style', 'border-color', 'visibility', 'background', 'background-color', 'text-decoration', 'box-shadow', 'transform', 'opacity'];
  var transitionEnabled =
  /*#__PURE__*/
  thd('transitionEnabled', true);
  var transitionDuration =
  /*#__PURE__*/
  thd('transitionDuration', '.2s');
  var transitionEasingFunc =
  /*#__PURE__*/
  thd('transitionEasingFunc', 'ease-in-out');
  var transition =
  /*#__PURE__*/
  mixin('transition', function (value) {
    return function (p) {
      if (!transitionEnabled(p)) return null;
      return styled.css(["transition:", ";@media screen and (prefers-reduced-motion:reduce){transition:none;}"], value);
    };
  });
  var transitionBase =
  /*#__PURE__*/
  thd('transitionBase', function (p) {
    return transition(safeTransitionAttrs.map(function (attr) {
      return attr + " " + transitionDuration(p) + " " + transitionEasingFunc(p);
    }).join(','))(p);
  });

  var zIndexControl =
  /*#__PURE__*/
  thd('zIndexControl', 1);
  var zIndexInnerSwitch =
  /*#__PURE__*/
  thd('zIndexInnerSwitch', 10);
  var zIndexModal =
  /*#__PURE__*/
  thd('zIndexModal', 1050);
  var zIndexModalBackdrop =
  /*#__PURE__*/
  thd('zIndexModalBackdrop', 1071);



  var theming = /*#__PURE__*/Object.freeze({
    alertPaddingY: alertPaddingY,
    alertPaddingX: alertPaddingX,
    alertMarginBottom: alertMarginBottom,
    alertColorLevel: alertColorLevel,
    alertBgLevel: alertBgLevel,
    alertBorderLevel: alertBorderLevel,
    alertVariant: alertVariant,
    borderRadius: borderRadius,
    borderRadiusSm: borderRadiusSm,
    borderRadiusLg: borderRadiusLg,
    borderWidth: borderWidth,
    btnPaddingY: btnPaddingY,
    btnPaddingX: btnPaddingX,
    btnLineHeight: btnLineHeight,
    btnPaddingYSm: btnPaddingYSm,
    btnPaddingXSm: btnPaddingXSm,
    btnLineHeightSm: btnLineHeightSm,
    btnPaddingYLg: btnPaddingYLg,
    btnPaddingXLg: btnPaddingXLg,
    btnLineHeightLg: btnLineHeightLg,
    btnBorderWidth: btnBorderWidth,
    btnDisabledOpacity: btnDisabledOpacity,
    btnVariant: btnVariant,
    black: black,
    white: white,
    gray100: gray100,
    gray200: gray200,
    gray300: gray300,
    gray400: gray400,
    gray500: gray500,
    gray600: gray600,
    gray700: gray700,
    gray800: gray800,
    gray900: gray900,
    blue: blue,
    indigo: indigo,
    purple: purple,
    pink: pink,
    red: red,
    brick: brick,
    orange: orange,
    yellow: yellow,
    green: green,
    teal: teal,
    cyan: cyan,
    primary: primary,
    secondary: secondary,
    success: success,
    info: info,
    warning: warning,
    danger: danger,
    light: light,
    dark: dark,
    primaryLight: primaryLight,
    secondaryLight: secondaryLight,
    colorVariants: colorVariants,
    colors: colors,
    colorInterval: colorInterval,
    colorVariant: colorVariant,
    colorLevel: colorLevel,
    yiqContrastedThreshold: yiqContrastedThreshold,
    yikTextDark: yikTextDark,
    yikTextLight: yikTextLight,
    colorYik: colorYik,
    baseFocus: baseFocus,
    controlFocus: controlFocus,
    gridColumns: gridColumns,
    gridGutter: gridGutter,
    gridMaxWidths: gridMaxWidths,
    headingsMarginBottom: headingsMarginBottom,
    headingsFontFamily: headingsFontFamily,
    headingsFontWeight: headingsFontWeight,
    headingsLineHeight: headingsLineHeight,
    headingsColor: headingsColor,
    h1FontSize: h1FontSize,
    h2FontSize: h2FontSize,
    h3FontSize: h3FontSize,
    h4FontSize: h4FontSize,
    h5FontSize: h5FontSize,
    h6FontSize: h6FontSize,
    display1Size: display1Size,
    display2Size: display2Size,
    display3Size: display3Size,
    display4Size: display4Size,
    display1Weight: display1Weight,
    display2Weight: display2Weight,
    display3Weight: display3Weight,
    display4Weight: display4Weight,
    displayLineHeight: displayLineHeight,
    inputBtnPaddingY: inputBtnPaddingY,
    inputBtnPaddingX: inputBtnPaddingX,
    inputBtnLineHeight: inputBtnLineHeight,
    inputBtnPaddingYSm: inputBtnPaddingYSm,
    inputBtnPaddingXSm: inputBtnPaddingXSm,
    inputBtnLineHeightSm: inputBtnLineHeightSm,
    inputBtnPaddingYLg: inputBtnPaddingYLg,
    inputBtnPaddingXLg: inputBtnPaddingXLg,
    inputBtnLineHeightLg: inputBtnLineHeightLg,
    inputPaddingY: inputPaddingY,
    inputPaddingX: inputPaddingX,
    inputLineHeight: inputLineHeight,
    inputPaddingYSm: inputPaddingYSm,
    inputPaddingXSm: inputPaddingXSm,
    inputLineHeightSm: inputLineHeightSm,
    inputPaddingYLg: inputPaddingYLg,
    inputPaddingXLg: inputPaddingXLg,
    inputLineHeightLg: inputLineHeightLg,
    inputBorderWidth: inputBorderWidth,
    inputBorderColor: inputBorderColor,
    inputBgColor: inputBgColor,
    inputDisabledBgColor: inputDisabledBgColor,
    inputDisabledText: inputDisabledText,
    inputPlaceholderText: inputPlaceholderText,
    inputTextColor: inputTextColor,
    modalBackdropBg: modalBackdropBg,
    modalTransitionDuration: modalTransitionDuration,
    modalInnerPadding: modalInnerPadding,
    modalDialogMargin: modalDialogMargin,
    modalDialogMarginYSmUp: modalDialogMarginYSmUp,
    modalContentBg: modalContentBg,
    modalContentBorderWidth: modalContentBorderWidth,
    modalContentBorderColor: modalContentBorderColor,
    modalContentBorderRadius: modalContentBorderRadius,
    modalContentBoxShadowXs: modalContentBoxShadowXs,
    modalContentBoxShadowSmUp: modalContentBoxShadowSmUp,
    modalHeaderBorderColor: modalHeaderBorderColor,
    modalFooterBorderColor: modalFooterBorderColor,
    modalHeaderBorderWidth: modalHeaderBorderWidth,
    modalFooterBorderWidth: modalFooterBorderWidth,
    breakpoints: breakpoints,
    up: up,
    down: down,
    between: between,
    spaces: spaces,
    transitionEnabled: transitionEnabled,
    transitionDuration: transitionDuration,
    transitionEasingFunc: transitionEasingFunc,
    transition: transition,
    transitionBase: transitionBase,
    fontSizes: fontSizes,
    fontFamily: fontFamily,
    fontSizeBase: fontSizeBase,
    fontSizeSm: fontSizeSm,
    fontSizeLg: fontSizeLg,
    fontWeightLight: fontWeightLight,
    fontWeightNormal: fontWeightNormal,
    fontWeightBold: fontWeightBold,
    lineHeightBase: lineHeightBase,
    lineHeightSm: lineHeightSm,
    lineHeightLg: lineHeightLg,
    zIndexControl: zIndexControl,
    zIndexInnerSwitch: zIndexInnerSwitch,
    zIndexModal: zIndexModal,
    zIndexModalBackdrop: zIndexModalBackdrop
  });

  function _objectWithoutPropertiesLoose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;

    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }

    return target;
  }

  var objectWithoutPropertiesLoose = _objectWithoutPropertiesLoose;

  var umd = createCommonjsModule(function (module, exports) {
  (function (global, factory) {
    module.exports =
    /*#__PURE__*/
    factory();
  })(commonjsGlobal, function () {

    var isMergeableObject = function isMergeableObject(value) {
      return isNonNullObject(value) && !isSpecial(value);
    };

    function isNonNullObject(value) {
      return !!value && typeof value === 'object';
    }

    function isSpecial(value) {
      var stringValue = Object.prototype.toString.call(value);
      return stringValue === '[object RegExp]' || stringValue === '[object Date]' || isReactElement(value);
    } // see https://github.com/facebook/react/blob/b5ac963fb791d1298e7f396236383bc955f916c1/src/isomorphic/classic/element/ReactElement.js#L21-L25


    var canUseSymbol = typeof Symbol === 'function' && Symbol["for"];
    var REACT_ELEMENT_TYPE = canUseSymbol ? Symbol["for"]('react.element') : 0xeac7;

    function isReactElement(value) {
      return value.$$typeof === REACT_ELEMENT_TYPE;
    }

    function emptyTarget(val) {
      return Array.isArray(val) ? [] : {};
    }

    function cloneUnlessOtherwiseSpecified(value, options) {
      return options.clone !== false && options.isMergeableObject(value) ? deepmerge(emptyTarget(value), value, options) : value;
    }

    function defaultArrayMerge(target, source, options) {
      return target.concat(source).map(function (element) {
        return cloneUnlessOtherwiseSpecified(element, options);
      });
    }

    function getMergeFunction(key, options) {
      if (!options.customMerge) {
        return deepmerge;
      }

      var customMerge = options.customMerge(key);
      return typeof customMerge === 'function' ? customMerge : deepmerge;
    }

    function mergeObject(target, source, options) {
      var destination = {};

      if (options.isMergeableObject(target)) {
        Object.keys(target).forEach(function (key) {
          destination[key] = cloneUnlessOtherwiseSpecified(target[key], options);
        });
      }

      Object.keys(source).forEach(function (key) {
        if (!options.isMergeableObject(source[key]) || !target[key]) {
          destination[key] = cloneUnlessOtherwiseSpecified(source[key], options);
        } else {
          destination[key] = getMergeFunction(key, options)(target[key], source[key], options);
        }
      });
      return destination;
    }

    function deepmerge(target, source, options) {
      options = options || {};
      options.arrayMerge = options.arrayMerge || defaultArrayMerge;
      options.isMergeableObject = options.isMergeableObject || isMergeableObject;
      var sourceIsArray = Array.isArray(source);
      var targetIsArray = Array.isArray(target);
      var sourceAndTargetTypesMatch = sourceIsArray === targetIsArray;

      if (!sourceAndTargetTypesMatch) {
        return cloneUnlessOtherwiseSpecified(source, options);
      } else if (sourceIsArray) {
        return options.arrayMerge(target, source, options);
      } else {
        return mergeObject(target, source, options);
      }
    }

    deepmerge.all = function deepmergeAll(array, options) {
      if (!Array.isArray(array)) {
        throw new Error('first argument should be an array');
      }

      return array.reduce(function (prev, next) {
        return deepmerge(prev, next, options);
      }, {});
    };

    var deepmerge_1 = deepmerge;
    return deepmerge_1;
  });
  });

  var minBreakpoint = function minBreakpoint(breakpoint) {
    return breakpoint !== 0 ? breakpoint : null;
  };

  var minWidth = function minWidth(value) {
    return "@media (min-width: " + value + "px)";
  };

  var DEFAULT_BREAKPOINTS = {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200
  };

  var is$1 = function is(n) {
    return n !== undefined && n !== null;
  };

  var num$1 = function num(n) {
    return typeof n === 'number' && !Number.isNaN(n);
  };

  var string$1 = function string(n) {
    return typeof n === 'string' && n !== '';
  };

  var obj = function obj(n) {
    return typeof n === 'object' && n !== null;
  };

  var func$1 = function func(n) {
    return typeof n === 'function';
  };

  var negative$1 = function negative(n) {
    return n < 0;
  };

  var get$1 = function get(obj, path) {
    return String(path).split('.').reduce(function (a, b) {
      return a && is$1(a[b]) ? a[b] : undefined;
    }, obj);
  };

  function merge(acc, item) {
    if (!item) {
      return acc;
    }

    return umd(acc, item, {
      clone: false // No need to clone deep, it's way faster.

    });
  }

  function callOrReturn(fn, arg) {
    if (!func$1(fn)) return fn;
    var next = fn(arg);
    return callOrReturn(next, arg);
  }

  function getThemeValue(theme, path, initial) {
    if (!theme) return undefined;
    return callOrReturn(get$1(initial || theme, path), {
      theme: theme
    });
  }

  function getValue(value, variants, theme) {
    if (is$1(variants)) {
      var valueFromVariants = getThemeValue(theme, value, variants);

      if (is$1(valueFromVariants)) {
        return valueFromVariants;
      }
    }

    return value;
  }

  function styleFromValue(cssProperties, value, theme, themeKey, transform) {
    var variants = getThemeValue(theme, themeKey);
    var computedValue = getValue(value, variants, theme);

    if (string$1(computedValue) || num$1(computedValue)) {
      var _style = {};

      for (var i = 0; i < cssProperties.length; i++) {
        _style[cssProperties[i]] = transform ? transform(computedValue, {
          rawValue: value,
          variants: variants
        }) : computedValue;
      }

      return _style;
    }

    return null;
  }

  function getThemeFromCssArg(propsOrTheme) {
    if (!propsOrTheme) {
      return null;
    } // Styled Components


    if (propsOrTheme.theme) {
      return propsOrTheme.theme;
    } // Emotion


    return propsOrTheme;
  }

  function getBreakpoints(theme) {
    var themeBreakpoints = getThemeValue(theme, 'breakpoints');

    if (is$1(themeBreakpoints)) {
      return themeBreakpoints;
    }

    return DEFAULT_BREAKPOINTS;
  }

  function createStyleGenerator(getStyle, props, generators) {
    var getStyles = function getStyles(attrs) {
      return function (propsOrTheme) {
        var theme = getThemeFromCssArg(propsOrTheme);
        return getStyle(attrs, theme);
      };
    };

    var getStylesFromProps = function getStylesFromProps(props) {
      var theme = props.theme || null;
      return getStyle(props, theme);
    };

    getStyles.meta = {
      props: props,
      getStyle: getStyle,
      generators: generators
    };
    getStyles.props = getStylesFromProps;
    return getStyles;
  }

  function styleFromBreakPoint(cssProperties, value, theme, themeKey, transform) {
    var breakpoints = getBreakpoints(theme);
    var keys = Object.keys(value);
    var allStyle = {};

    for (var i = 0; i < keys.length; i++) {
      var breakpoint = keys[i];

      var _style2 = styleFromValue(cssProperties, value[breakpoint], theme, themeKey, transform);

      if (_style2 !== null) {
        var breakpointValue = minBreakpoint(breakpoints[breakpoint]);

        if (breakpointValue === null) {
          allStyle = merge(allStyle, _style2);
        } else {
          var _merge;

          allStyle = merge(allStyle, (_merge = {}, _merge[minWidth(breakpointValue)] = _style2, _merge));
        }
      }
    }

    return allStyle;
  }

  function getStyleFactory(prop, cssProperties, themeKey, transform) {
    return function getStyle(attrs, theme) {
      var value = attrs[prop];
      if (!is$1(value)) return null;
      cssProperties = cssProperties || [prop];
      var style = styleFromValue(cssProperties, value, theme, themeKey, transform);

      if (style !== null) {
        return style;
      }

      if (obj(value)) {
        return styleFromBreakPoint(cssProperties, value, theme, themeKey, transform);
      }

      return null;
    };
  }

  function style(_ref) {
    var prop = _ref.prop,
        cssProperties = _ref.cssProperties,
        _ref$themeKey = _ref.themeKey,
        themeKey = _ref$themeKey === void 0 ? null : _ref$themeKey,
        _ref$transform = _ref.transform,
        transform = _ref$transform === void 0 ? null : _ref$transform;
    var getStyle = getStyleFactory(prop, cssProperties, themeKey, transform);
    return createStyleGenerator(getStyle, [prop]);
  }

  function indexGeneratorsByProp(styles) {
    var index = {};

    for (var i = 0; i < styles.length; i++) {
      var _style3 = styles[i];

      if (_style3 && _style3.meta) {
        var propsKeys = Object.keys(_style3.meta.props);

        for (var j = 0; j < propsKeys.length; j++) {
          var prop = _style3.meta.props[propsKeys[j]];
          index[prop] = _style3;
        }
      }
    }

    return index;
  }

  function compose() {
    var flatGenerators = [];

    for (var _len = arguments.length, generators = new Array(_len), _key = 0; _key < _len; _key++) {
      generators[_key] = arguments[_key];
    }

    generators.forEach(function (gen) {
      if (gen.meta.generators) {
        flatGenerators = [].concat(flatGenerators, gen.meta.generators);
      } else {
        flatGenerators.push(gen);
      }
    });
    var generatorsByProp = indexGeneratorsByProp(flatGenerators);

    function getStyle(attrs, theme) {
      var propKeys = Object.keys(attrs);
      var propCount = propKeys.length;
      var allStyle = {};

      for (var i = 0; i < propCount; i++) {
        var propKey = propKeys[i];
        var generator = generatorsByProp[propKey];

        if (generator) {
          allStyle = merge(allStyle, generator.meta.getStyle(attrs, theme));
        }
      }

      return allStyle;
    }

    var props = flatGenerators.reduce(function (keys, generator) {
      return [].concat(keys, generator.meta.props);
    }, []);
    return createStyleGenerator(getStyle, props, generators);
  }

  function universal(generator) {
    var generatorsByProp = indexGeneratorsByProp([generator]);

    function genericGetStyle(opened) {
      return function getStyle(attrs, theme) {
        var propKeys = Object.keys(attrs);
        var propCount = propKeys.length;
        var allStyle = {};

        for (var i = 0; i < propCount; i++) {
          var propKey = propKeys[i];
          var _generator = generatorsByProp[propKey];
          allStyle = merge(allStyle, _generator ? _generator.meta.getStyle(attrs, theme) : opened ? getStyleFactory(propKey)(attrs, theme) : null);
        }

        return allStyle;
      };
    }

    var openedGetStyle = genericGetStyle(true);
    var restrictedGetStyle = genericGetStyle(false);
    var getStyles = createStyleGenerator(openedGetStyle, generator.meta.props, [generator]);

    getStyles.props = function (props) {
      var theme = props.theme || null;
      return restrictedGetStyle(props, theme, props);
    };

    return getStyles;
  }

  var background =
  /*#__PURE__*/
  style({
    prop: 'background'
  });
  var backgroundColor =
  /*#__PURE__*/
  style({
    prop: 'backgroundColor',
    themeKey: 'colors'
  });
  var backgroundImage =
  /*#__PURE__*/
  style({
    prop: 'backgroundImage'
  });
  var backgroundSize =
  /*#__PURE__*/
  style({
    prop: 'backgroundSize'
  });
  var backgroundPosition =
  /*#__PURE__*/
  style({
    prop: 'backgroundPosition'
  });
  var backgroundRepeat =
  /*#__PURE__*/
  style({
    prop: 'backgroundRepeat'
  });
  var backgrounds =
  /*#__PURE__*/
  compose(background, backgroundColor, backgroundImage, backgroundSize, backgroundPosition, backgroundRepeat);
  var opacity =
  /*#__PURE__*/
  style({
    prop: 'opacity'
  });
  var overflow =
  /*#__PURE__*/
  style({
    prop: 'overflow'
  });
  var basics =
  /*#__PURE__*/
  compose(opacity, overflow);

  var unit$1 = function unit(_unit) {
    return function (value) {
      return num$1(value) ? "" + value + _unit : value;
    };
  };

  var px$1 =
  /*#__PURE__*/
  unit$1('px');

  var percent = function percent(n) {
    return !num$1(n) || n > 1 ? px$1(n) : n * 100 + "%";
  };

  var getBorder = function getBorder(n) {
    return num$1(n) && n > 0 ? n + "px solid" : n;
  };

  var border =
  /*#__PURE__*/
  style({
    prop: 'border',
    themeKey: 'borders',
    transform: getBorder
  });
  var borderTop =
  /*#__PURE__*/
  style({
    prop: 'borderTop',
    themeKey: 'borders',
    transform: getBorder
  });
  var borderRight =
  /*#__PURE__*/
  style({
    prop: 'borderRight',
    themeKey: 'borders',
    transform: getBorder
  });
  var borderBottom =
  /*#__PURE__*/
  style({
    prop: 'borderBottom',
    themeKey: 'borders',
    transform: getBorder
  });
  var borderLeft =
  /*#__PURE__*/
  style({
    prop: 'borderLeft',
    themeKey: 'borders',
    transform: getBorder
  });
  var borderColor =
  /*#__PURE__*/
  style({
    prop: 'borderColor',
    themeKey: 'colors'
  });
  var borderRadius$1 =
  /*#__PURE__*/
  style({
    prop: 'borderRadius',
    themeKey: 'radii',
    transform: px$1
  });
  var boxShadow =
  /*#__PURE__*/
  style({
    prop: 'boxShadow',
    themeKey: 'shadows'
  });
  var borders =
  /*#__PURE__*/
  compose(border, borderTop, borderRight, borderBottom, borderLeft, borderColor, borderRadius$1, boxShadow);
  var width =
  /*#__PURE__*/
  style({
    prop: 'width',
    transform: percent,
    themeKey: 'widths'
  });
  var height =
  /*#__PURE__*/
  style({
    prop: 'height',
    transform: percent,
    themeKey: 'heights'
  });
  var maxWidth =
  /*#__PURE__*/
  style({
    prop: 'maxWidth',
    transform: percent,
    themeKey: 'widths'
  });
  var maxHeight =
  /*#__PURE__*/
  style({
    prop: 'maxHeight',
    transform: percent,
    themeKey: 'heights'
  });
  var minWidth$1 =
  /*#__PURE__*/
  style({
    prop: 'minWidth',
    transform: percent,
    themeKey: 'widths'
  });
  var minHeight =
  /*#__PURE__*/
  style({
    prop: 'minHeight',
    transform: percent,
    themeKey: 'heights'
  });
  var dimensions =
  /*#__PURE__*/
  compose(width, height, maxWidth, maxHeight, minWidth$1, minHeight);
  var display =
  /*#__PURE__*/
  style({
    prop: 'display'
  });
  var alignItems =
  /*#__PURE__*/
  style({
    prop: 'alignItems'
  });
  var alignContent =
  /*#__PURE__*/
  style({
    prop: 'alignContent'
  });
  var justifyContent =
  /*#__PURE__*/
  style({
    prop: 'justifyContent'
  });
  var flexWrap =
  /*#__PURE__*/
  style({
    prop: 'flexWrap'
  });
  var flexBasis =
  /*#__PURE__*/
  style({
    prop: 'flexBasis',
    transform: percent
  });
  var flexDirection =
  /*#__PURE__*/
  style({
    prop: 'flexDirection'
  });
  var flex =
  /*#__PURE__*/
  style({
    prop: 'flex'
  });
  var justifySelf =
  /*#__PURE__*/
  style({
    prop: 'justifySelf'
  });
  var alignSelf =
  /*#__PURE__*/
  style({
    prop: 'alignSelf'
  });
  var order =
  /*#__PURE__*/
  style({
    prop: 'order'
  });
  var flexboxes =
  /*#__PURE__*/
  compose(display, alignItems, alignContent, justifyContent, flexWrap, flexBasis, flexDirection, flex, justifySelf, alignSelf, order);
  var position =
  /*#__PURE__*/
  style({
    prop: 'position'
  });
  var zIndex =
  /*#__PURE__*/
  style({
    prop: 'zIndex',
    themeKey: 'zIndexes'
  });
  var top =
  /*#__PURE__*/
  style({
    prop: 'top',
    transform: px$1
  });
  var right =
  /*#__PURE__*/
  style({
    prop: 'right',
    transform: px$1
  });
  var bottom =
  /*#__PURE__*/
  style({
    prop: 'bottom',
    transform: px$1
  });
  var left =
  /*#__PURE__*/
  style({
    prop: 'left',
    transform: px$1
  });
  var positions =
  /*#__PURE__*/
  compose(position, zIndex, top, right, bottom, left);
  var DEFAULT_SPACING = [0, 8, 16, 24, 32, 40, 48, 56, 64, 72];

  function transform(transformedValue, _ref) {
    var rawValue = _ref.rawValue,
        _ref$variants = _ref.variants,
        variants = _ref$variants === void 0 ? DEFAULT_SPACING : _ref$variants;

    if (!num$1(rawValue)) {
      return variants[rawValue] || rawValue;
    }

    var abs = Math.abs(rawValue);
    var neg = negative$1(rawValue);
    var value = variants[abs] || abs;

    if (!num$1(value)) {
      return neg ? "-" + value : value;
    }

    return value * (neg ? -1 : 1);
  }

  var themeKey = 'spaces';
  var m =
  /*#__PURE__*/
  style({
    prop: 'm',
    cssProperties: ['margin'],
    themeKey: themeKey,
    transform: transform
  });
  var mt =
  /*#__PURE__*/
  style({
    prop: 'mt',
    cssProperties: ['marginTop'],
    themeKey: themeKey,
    transform: transform
  });
  var mr =
  /*#__PURE__*/
  style({
    prop: 'mr',
    cssProperties: ['marginRight'],
    themeKey: themeKey,
    transform: transform
  });
  var mb =
  /*#__PURE__*/
  style({
    prop: 'mb',
    cssProperties: ['marginBottom'],
    themeKey: themeKey,
    transform: transform
  });
  var ml =
  /*#__PURE__*/
  style({
    prop: 'ml',
    cssProperties: ['marginLeft'],
    themeKey: themeKey,
    transform: transform
  });
  var mx =
  /*#__PURE__*/
  style({
    prop: 'mx',
    cssProperties: ['marginRight', 'marginLeft'],
    themeKey: themeKey,
    transform: transform
  });
  var my =
  /*#__PURE__*/
  style({
    prop: 'my',
    cssProperties: ['marginTop', 'marginBottom'],
    themeKey: themeKey,
    transform: transform
  });
  var p =
  /*#__PURE__*/
  style({
    prop: 'p',
    cssProperties: ['padding'],
    themeKey: themeKey,
    transform: transform
  });
  var pt =
  /*#__PURE__*/
  style({
    prop: 'pt',
    cssProperties: ['paddingTop'],
    themeKey: themeKey,
    transform: transform
  });
  var pr =
  /*#__PURE__*/
  style({
    prop: 'pr',
    cssProperties: ['paddingRight'],
    themeKey: themeKey,
    transform: transform
  });
  var pb =
  /*#__PURE__*/
  style({
    prop: 'pb',
    cssProperties: ['paddingBottom'],
    themeKey: themeKey,
    transform: transform
  });
  var pl =
  /*#__PURE__*/
  style({
    prop: 'pl',
    cssProperties: ['paddingLeft'],
    themeKey: themeKey,
    transform: transform
  });
  var px$1$1 =
  /*#__PURE__*/
  style({
    prop: 'px',
    cssProperties: ['paddingRight', 'paddingLeft'],
    themeKey: themeKey,
    transform: transform
  });
  var py =
  /*#__PURE__*/
  style({
    prop: 'py',
    cssProperties: ['paddingTop', 'paddingBottom'],
    themeKey: themeKey,
    transform: transform
  });
  var space =
  /*#__PURE__*/
  compose(m, mt, mr, mb, ml, mx, my, p, pt, pr, pb, pl, px$1$1, py);
  var fontFamily$1 =
  /*#__PURE__*/
  style({
    prop: 'fontFamily',
    themeKey: 'fonts'
  });
  var fontSize =
  /*#__PURE__*/
  style({
    prop: 'fontSize',
    themeKey: 'fontSizes'
  });
  var lineHeight =
  /*#__PURE__*/
  style({
    prop: 'lineHeight',
    themeKey: 'lineHeights'
  });
  var fontWeight =
  /*#__PURE__*/
  style({
    prop: 'fontWeight',
    themeKey: 'fontWeights'
  });
  var textAlign =
  /*#__PURE__*/
  style({
    prop: 'textAlign'
  });
  var letterSpacing =
  /*#__PURE__*/
  style({
    prop: 'letterSpacing',
    themeKey: 'letterSpacings',
    transform: px$1
  });
  var color =
  /*#__PURE__*/
  style({
    prop: 'color',
    themeKey: 'colors'
  });
  var textTransform =
  /*#__PURE__*/
  style({
    prop: 'textTransform'
  });
  var typography =
  /*#__PURE__*/
  compose(fontFamily$1, fontSize, lineHeight, fontWeight, textAlign, letterSpacing, color, textTransform);
  var system =
  /*#__PURE__*/
  compose(backgrounds, basics, borders, dimensions, flexboxes, positions, space, typography);

  var universalSystem =
  /*#__PURE__*/
  universal(system);

  function getSystemPropTypes(system) {
    if (!system) return {};
    return system.meta.props.reduce(function (obj, prop) {
      obj[prop] = PropTypes__default.oneOfType([PropTypes__default.number, PropTypes__default.string, PropTypes__default.object]);
      return obj;
    }, {});
  }



  var systemDefs = /*#__PURE__*/Object.freeze({
    colors: colors,
    fontSizes: fontSizes,
    spaces: spaces
  });

  var defaultSystemTheme =
  /*#__PURE__*/
  definitionsToTheme(systemDefs);

  function getProps(props, defaultProps) {
    var theme = _extends_1({}, defaultSystemTheme, getTheme(props));

    return _extends_1({}, defaultProps, props, {
      theme: theme
    });
  }

  function createComponent(getConfig) {
    var _getConfig = getConfig(),
        name = _getConfig.name,
        style = _getConfig.style,
        _getConfig$omitProps = _getConfig.omitProps,
        omitProps = _getConfig$omitProps === void 0 ? [] : _getConfig$omitProps,
        _getConfig$defaultPro = _getConfig.defaultProps,
        defaultProps = _getConfig$defaultPro === void 0 ? {} : _getConfig$defaultPro,
        _getConfig$propTypes = _getConfig.propTypes,
        propTypes = _getConfig$propTypes === void 0 ? {} : _getConfig$propTypes,
        _getConfig$render = _getConfig.render,
        render = _getConfig$render === void 0 ? function (_ref) {
      var Component = _ref.Component,
          props = objectWithoutPropertiesLoose(_ref, ["Component"]);

      return React__default.createElement(Component, props);
    } : _getConfig$render,
        _getConfig$defaultCom = _getConfig.defaultComponent,
        defaultComponent = _getConfig$defaultCom === void 0 ? 'div' : _getConfig$defaultCom,
        _getConfig$system = _getConfig.system,
        system$1 = _getConfig$system === void 0 ? system : _getConfig$system,
        _getConfig$applySyste = _getConfig.applySystem,
        applySystem = _getConfig$applySyste === void 0 ? function (system) {
      return function (props) {
        return {
          '&&': system.props(props)
        };
      };
    } : _getConfig$applySyste,
        InnerComponentFromConfig = _getConfig.InnerComponent;

    var omittedProps = ['theme'].concat(system$1 ? system$1.meta.props : {}, omitProps);
    var baseClassName = "sui-" + name;

    var InnerComponent = InnerComponentFromConfig || function Component(_ref2) {
      var className = _ref2.className,
          as = _ref2.as,
          forwardedRef = _ref2.forwardedRef,
          props = objectWithoutPropertiesLoose(_ref2, ["className", "as", "forwardedRef"]);

      var Component = as || defaultComponent;

      var renderProps = _extends_1({
        ref: forwardedRef,
        Component: Component,
        className: className ? baseClassName + " " + className : baseClassName
      }, omit(props, omittedProps));

      return render(renderProps);
    };

    InnerComponent.displayName = "sui-" + name;

    function forwardRef(props, ref) {
      return React__default.createElement(InnerComponent, _extends_1({}, props, {
        forwardedRef: ref
      }));
    }

    forwardRef.displayName = InnerComponent.displayName;
    var RefComponent = React__default.forwardRef(forwardRef);
    RefComponent.displayName = InnerComponent.displayName;
    var StyledComponent = styled__default(RefComponent)(function (p) {
      var styles = [];
      var props = getProps(p, defaultProps);

      if (func(style)) {
        styles.push(style(props));
      }

      if (func(applySystem)) {
        styles.push(applySystem(system$1)(props));
      }

      return styles;
    });
    StyledComponent.propTypes = _extends_1({
      theme: PropTypes__default.object
    }, getSystemPropTypes(system$1), propTypes);
    StyledComponent.defaultProps = _extends_1({}, defaultProps);
    return StyledComponent;
  }

  var Alert =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'alert',
      omitProps: ['variant'],
      style: function style(p) {
        return styled.css(["position:relative;padding:", " ", ";margin-bottom:", ";border:1px solid transparent;border-radius:", ";", ";"], alertPaddingY(p), alertPaddingX(p), alertMarginBottom(p), borderRadius(p), p.variant && alertVariant(p.variant)(p));
      },
      propTypes: {
        children: PropTypes__default.node,
        variant: PropTypes__default.string
      },
      defaultProps: {
        variant: 'primary'
      }
    };
  });

  var Box =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'box'
    };
  });

  var Breakpoint =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'breakpoint',
      omitProps: ['up', 'down'],
      style: function style(p) {
        return styled.css(["display:none;", ";", ";"], p.up && up(p.up, 'display: block;')(p), p.down && down(p.down, 'display: block;')(p));
      },
      propTypes: {
        children: PropTypes__default.node,
        up: PropTypes__default.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
        down: PropTypes__default.oneOf(['xs', 'sm', 'md', 'lg', 'xl'])
      }
    };
  });

  var btnSize = {
    sm: function sm(p) {
      return styled.css(["padding:", " ", ";font-size:", ";border-radius:", ";line-height:", ";"], btnPaddingYSm(p), btnPaddingXSm(p), fontSizeSm(p), borderRadiusSm(p), btnLineHeightSm(p));
    },
    md: function md(p) {
      return styled.css(["padding:", " ", ";font-size:", ";border-radius:", ";line-height:", ";"], btnPaddingY(p), btnPaddingX(p), fontSizeBase(p), borderRadius(p), btnLineHeight(p));
    },
    lg: function lg(p) {
      return styled.css(["padding:", " ", ";font-size:", ";border-radius:", ";line-height:", ";"], btnPaddingYLg(p), btnPaddingXLg(p), fontSizeLg(p), borderRadiusLg(p), btnLineHeightLg(p));
    }
  };
  var Button =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'button',
      defaultComponent: 'button',
      omitProps: ['size', 'variant'],
      style: function style(p) {
        return styled.css(["display:inline-block;z-index:", ";border-width:", ";cursor:pointer;", ";text-decoration:none;&:disabled{opacity:", ";}", ";", ";"], zIndexControl(p), btnBorderWidth(p), transitionBase(p), btnDisabledOpacity(p), p.size && btnSize[p.size] && btnSize[p.size](p), p.variant && btnVariant(p.variant)(p));
      },
      propTypes: {
        children: PropTypes__default.node,
        disabled: PropTypes__default.bool,
        size: PropTypes__default.oneOf(['sm', 'md', 'lg']),
        variant: PropTypes__default.string
      },
      defaultProps: {
        size: 'md',
        variant: 'primary'
      }
    };
  });

  function _inheritsLoose(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    subClass.__proto__ = superClass;
  }

  var inheritsLoose = _inheritsLoose;

  function create() {
    var SwitchState =
    /*#__PURE__*/
    function (_Component) {
      inheritsLoose(SwitchState, _Component);

      function SwitchState() {
        var _this;

        for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }

        _this = _Component.call.apply(_Component, [this].concat(args)) || this;
        _this.state = {
          focused: false,
          checked: _this.props.checked || _this.props.defaultChecked || false
        };

        _this.handleChange = function (event) {
          if (_this.props.checked === undefined) {
            _this.setState({
              checked: event.currentTarget.checked
            });
          }

          if (_this.props.onChange) {
            _this.props.onChange(event);
          }
        };

        _this.handleFocus = function (event) {
          _this.setState({
            focused: true
          });

          if (_this.props.onFocus) {
            _this.props.onFocus(event);
          }
        };

        _this.handleBlur = function (event) {
          _this.setState({
            focused: false
          });

          if (_this.props.onBlur) {
            _this.props.onBlur(event);
          }
        };

        return _this;
      }

      SwitchState.getDerivedStateFromProps = function getDerivedStateFromProps(props, state) {
        if (props.checked !== undefined && props.checked !== state.checked) {
          return _extends_1({}, state, {
            checked: props.checked
          });
        }

        return state;
      };

      var _proto = SwitchState.prototype;

      _proto.render = function render() {
        var _this$props = this.props,
            checked = _this$props.checked,
            children = _this$props.children,
            defaultChecked = _this$props.defaultChecked,
            onChange = _this$props.onChange,
            onBlur = _this$props.onBlur,
            onFocus = _this$props.onFocus,
            style = _this$props.style,
            props = objectWithoutPropertiesLoose(_this$props, ["checked", "children", "defaultChecked", "onChange", "onBlur", "onFocus", "style"]);

        return this.props.children({
          focused: this.state.focused,
          checked: this.state.checked,
          disabled: props.disabled,
          input: _extends_1({
            checked: this.state.checked,
            onChange: this.handleChange,
            onFocus: this.handleFocus,
            onBlur: this.handleBlur,
            style: _extends_1({
              position: 'absolute',
              margin: 0,
              padding: 0,
              top: 0,
              right: 0,
              left: 0,
              bottom: 0,
              opacity: 0,
              width: '100%',
              height: '100%',
              zIndex: 10
            }, style)
          }, props)
        });
      };

      return SwitchState;
    }(React.Component);

    return SwitchState;
  }

  var SwitchState = /*#__PURE__*/
  create();

  var sizeStyle = {
    sm: function sm(p) {
      return styled.css([".sui-checkbox-content{border-radius:", ";width:0.875rem;height:0.875rem;}.sui-checkbox-check{width:8px;height:8px;}"], borderRadiusSm(p));
    },
    md: function md(p) {
      return styled.css([".sui-checkbox-content{border-radius:", ";width:1rem;height:1rem;}.sui-checkbox-check{width:10px;height:10px;}"], borderRadius(p));
    },
    lg: function lg(p) {
      return styled.css([".sui-checkbox-content{border-radius:", ";width:1.25rem;height:1.25rem;}.sui-checkbox-check{width:12px;height:12px;}"], borderRadiusLg(p));
    }
  };

  var validStyle = function validStyle(p) {
    var valid = p.valid;
    if (valid !== true && valid !== false) return null;
    var color = valid ? success(p) : danger(p);
    return styled.css(["input + .sui-checkbox-content,input:checked + .sui-checkbox-content{border-color:", ";}input:checked + .sui-checkbox-content{background-color:", ";}input:focus + .sui-checkbox-content{border-color:", ";", "}"], color, color, color, controlFocus(color)(p));
  };

  var controlStyle = function controlStyle(p) {
    return styled.css(["input:focus + .sui-checkbox-content{", "}", ";"], controlFocus(primary(p))(p), validStyle(p));
  };

  var containerSystem =
  /*#__PURE__*/
  compose(basics, dimensions, space, flexboxes, positions);
  var contentSystem =
  /*#__PURE__*/
  compose(dimensions, backgrounds, borders);
  var system$1 =
  /*#__PURE__*/
  compose(containerSystem, contentSystem);
  var Checkbox =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'checkbox',
      omitProps: ['control', 'valid'],
      system: system$1,
      applySystem: null,
      render: function render(_ref) {
        var Component = _ref.Component,
            ref = _ref.ref,
            className = _ref.className,
            size = _ref.size,
            props = objectWithoutPropertiesLoose(_ref, ["Component", "ref", "className", "size"]);

        return React__default.createElement(SwitchState, props, function (_ref2) {
          var input = _ref2.input;
          return React__default.createElement(Component, {
            className: className
          }, React__default.createElement("input", _extends_1({
            ref: ref,
            type: "checkbox"
          }, input)), React__default.createElement("div", {
            className: "sui-checkbox-content"
          }, React__default.createElement("svg", {
            className: "sui-checkbox-check",
            viewBox: "0 0 512 512"
          }, React__default.createElement("path", {
            fill: "currentColor",
            d: "M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          }))));
        });
      },
      style: function style(p) {
        return styled.css(["display:inline-flex;align-items:center;justify-content:center;position:relative;width:1.5rem;height:1.5rem;z-index:", ";.sui-checkbox-check{pointer-events:none;transform:scale(0);transform-origin:center;color:", ";", ";}.sui-checkbox-content{display:flex;align-items:center;justify-content:center;width:1rem;height:1rem;background-color:", ";border-radius:", ";border-style:solid;border-width:", ";border-color:", ";", ";}input:checked + .sui-checkbox-content{background-color:", ";border-color:transparent;svg{transform:scale(1);}}input:focus + .sui-checkbox-content{", ";}input:disabled + .sui-checkbox-content{background-color:", ";}", ";", ";", ";.sui-checkbox-content{", ";}"], zIndexControl(p), white(p), transitionBase(p), inputBgColor(p), borderRadius(p), inputBorderWidth(p), inputBorderColor(p), transitionBase(p), primary(p), baseFocus(primary(p))(p), inputDisabledBgColor(p), sizeStyle[p.size] && sizeStyle[p.size] && sizeStyle[p.size](p), p.control && controlStyle(p), containerSystem.props, contentSystem.props);
      },
      propTypes: {
        checked: PropTypes__default.bool,
        control: PropTypes__default.bool,
        disabled: PropTypes__default.bool,
        onChange: PropTypes__default.func,
        size: PropTypes__default.oneOf(['sm', 'md', 'lg']),
        valid: PropTypes__default.bool,
        value: PropTypes__default.string
      },
      defaultProps: {
        size: 'md'
      }
    };
  });

  var common = {
    position: 'relative',
    width: '100%',
    minHeight: '1px'
  };

  function isValidSize(size) {
    return size || size === 0;
  }

  function getSizeWidth(size, nbColumns) {
    return Math.round(size / nbColumns * Math.pow(10, 6)) / Math.pow(10, 4) + "%";
  }

  function getOffsetStyle(offsetSize, nbColumns) {
    if (isValidSize(offsetSize)) {
      var marginLeft = getSizeWidth(offsetSize, nbColumns);
      return {
        marginLeft: marginLeft
      };
    }

    return null;
  }

  function getBreakPointStyle(breakpoint, width, p) {
    var size = p[breakpoint];
    if (!isValidSize(size)) return null;
    var nbColumns = gridColumns(p);
    var mediaQuery = mediaMinWidth(width);

    var media = function media(style) {
      var _ref;

      return width === 0 ? style : (_ref = {}, _ref[mediaQuery] = style, _ref);
    };

    if (size === true) {
      return _extends_1({}, common, media({
        flexBasis: 0,
        flexGrow: 1,
        maxWidth: '100%'
      }));
    }

    if (size === 'auto') {
      return _extends_1({}, common, media({
        flex: "0 0 auto",
        maxWidth: 'none',
        width: 'auto'
      }));
    }

    var sizeWidth = getSizeWidth(size, nbColumns);
    return _extends_1({}, common, media(_extends_1({
      flex: "0 0 " + sizeWidth,
      maxWidth: sizeWidth
    }, getOffsetStyle(p["o" + breakpoint], nbColumns))));
  }

  function getStyleFromProps(p) {
    var _p$gutter = p.gutter,
        gutter = _p$gutter === void 0 ? gridGutter(p) : _p$gutter;
    var bk = breakpoints(p);
    var breakpointsKeys = Object.keys(bk);
    var style = {
      boxSizing: 'border-box',
      flexBasis: 0,
      flexGrow: 1,
      maxWidth: '100%'
    };
    style = merge(style, pl({
      pl: gutter
    })(p));
    style = merge(style, pr({
      pr: gutter
    })(p));
    var index = -1; // eslint-disable-next-line no-plusplus

    while (++index < breakpointsKeys.length) {
      var breakpoint = breakpointsKeys[index];
      style = merge(style, getBreakPointStyle(breakpoint, bk[breakpoint], p));
    }

    return style;
  }

  function create$1() {
    var Col = createComponent(function () {
      return {
        name: 'col',
        omitProps: ['xs', 'sm', 'md', 'lg', 'xl', 'oxs', 'osm', 'omd', 'olg', 'oxl', 'gutter'],
        style: getStyleFromProps,
        propTypes: {
          children: PropTypes__default.node,
          gutter: getSystemPropTypes(pr).pr,
          xs: PropTypes__default.oneOfType([PropTypes__default.bool, PropTypes__default.string, PropTypes__default.number]),
          sm: PropTypes__default.oneOfType([PropTypes__default.bool, PropTypes__default.string, PropTypes__default.number]),
          md: PropTypes__default.oneOfType([PropTypes__default.bool, PropTypes__default.string, PropTypes__default.number]),
          lg: PropTypes__default.oneOfType([PropTypes__default.bool, PropTypes__default.string, PropTypes__default.number]),
          xl: PropTypes__default.oneOfType([PropTypes__default.bool, PropTypes__default.string, PropTypes__default.number]),
          oxs: PropTypes__default.oneOfType([PropTypes__default.string, PropTypes__default.number]),
          osm: PropTypes__default.oneOfType([PropTypes__default.string, PropTypes__default.number]),
          omd: PropTypes__default.oneOfType([PropTypes__default.string, PropTypes__default.number]),
          olg: PropTypes__default.oneOfType([PropTypes__default.string, PropTypes__default.number]),
          oxl: PropTypes__default.oneOfType([PropTypes__default.string, PropTypes__default.number])
        }
      };
    });
    var Split = createComponent(function () {
      return {
        name: 'col-split',
        style: function style() {
          return styled.css(["width:100% !important;"]);
        }
      };
    });
    Col.Split = Split;
    return Col;
  }

  var Col = /*#__PURE__*/
  create$1();

  var ControlFeedback =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'control-feedback',
      omitProps: ['valid'],
      style: function style(p) {
        return styled.css(["width:100%;margin-top:0.25rem;font-size:80%;color:", ";"], p.valid ? success(p) : danger(p));
      },
      propTypes: {
        children: PropTypes__default.node,
        valid: PropTypes__default.bool.isRequired
      }
    };
  });

  var inlineStyle =
  /*#__PURE__*/
  styled.css(["display:inline-flex;margin-right:0.75rem;"]);
  var FormCheck =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'form-check',
      omitProps: ['inline'],
      style: function style(p) {
        return styled.css(["display:flex;align-items:center;", ";"], p.inline && inlineStyle);
      },
      propTypes: {
        children: PropTypes__default.node,
        inline: PropTypes__default.bool
      }
    };
  });

  var FormCheckLabel =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'form-check-label',
      defaultComponent: 'label',
      style: function style(p) {
        return styled.css(["padding-left:0.25rem;[class*='disabled'] ~ &{color:", ";}"], inputDisabledText(p));
      },
      propTypes: {
        children: PropTypes__default.node
      }
    };
  });

  var FormGroup =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'form-group',
      style: function style() {
        return styled.css(["margin-bottom:1rem;"]);
      },
      propTypes: {
        children: PropTypes__default.node
      }
    };
  });

  function _taggedTemplateLiteralLoose(strings, raw) {
    if (!raw) {
      raw = strings.slice(0);
    }

    strings.raw = raw;
    return strings;
  }

  var taggedTemplateLiteralLoose = _taggedTemplateLiteralLoose;

  function _templateObject() {
    var data = taggedTemplateLiteralLoose(["\n  ", "\n  html,\n  body {\n    font-family: ", ";\n    font-size: ", ";\n    line-height: ", ";\n    -webkit-font-smoothing: antialiased;\n  }\n\n  * {\n    box-sizing: border-box;\n  }\n"]);

    _templateObject = function _templateObject() {
      return data;
    };

    return data;
  }
  var Normalize =
  /*#__PURE__*/
  styled.createGlobalStyle(
  /*#__PURE__*/
  _templateObject(),
  /*#__PURE__*/
  polished.normalize(), fontFamily, fontSizeBase, lineHeightBase);

  var styleBreakpoints = function styleBreakpoints(p) {
    var bk = breakpoints(p);
    var maxWidths = gridMaxWidths(p);

    var style = function style(breakpoint) {
      var width = bk[breakpoint];

      var media = function media(s) {
        var _ref;

        return width === 0 ? null : (_ref = {}, _ref[mediaMinWidth(width)] = s, _ref);
      };

      return media({
        maxWidth: maxWidths[breakpoint]
      });
    };

    return Object.keys(bk).reduce(function (obj, breakpoint) {
      return Object.assign(obj, style(breakpoint));
    }, {});
  };

  var Grid =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'grid',
      omitProps: ['gutter', 'fluid'],
      style: function style(p) {
        var _p$gutter = p.gutter,
            gutter = _p$gutter === void 0 ? gridGutter(p) : _p$gutter;
        return styled.css(["width:100%;margin-left:auto;margin-right:auto;", ";", ";", ";"], pr({
          pr: gutter
        })(p), pl({
          pl: gutter
        })(p), p.fluid ? null : styleBreakpoints(p));
      },
      propTypes: {
        children: PropTypes__default.node,
        gutter: getSystemPropTypes(pr).pr,
        fluid: PropTypes__default.bool
      }
    };
  });

  var sizeStyle$1 = {
    sm: function sm(p) {
      return styled.css(["padding:", " ", ";font-size:", ";line-height:", ";border-radius:", ";"], inputPaddingYSm(p), inputPaddingXSm(p), fontSizeSm(p), inputLineHeightSm(p), borderRadiusSm(p));
    },
    md: function md(p) {
      return styled.css(["padding:", " ", ";font-size:", ";line-height:", ";border-radius:", ";"], inputPaddingY(p), inputPaddingX(p), fontSizeBase(p), inputLineHeight(p), borderRadius(p));
    },
    lg: function lg(p) {
      return styled.css(["padding:", " ", ";font-size:", ";line-height:", ";border-radius:", ";"], inputPaddingYLg(p), inputPaddingXLg(p), fontSizeLg(p), inputLineHeightLg(p), borderRadiusLg(p));
    }
  };

  var validStyle$1 = function validStyle(p) {
    var valid = p.valid;
    if (valid !== true && valid !== false) return null;
    var color = valid ? success(p) : danger(p);
    return styled.css(["border-color:", ";&:focus{border-color:", ";", "}"], color, color, controlFocus(color)(p));
  };

  var controlStyle$1 = function controlStyle(p) {
    return styled.css(["display:block;width:100%;&:focus{", "}", ";"], controlFocus(primary(p))(p), validStyle$1(p));
  };

  var Input =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'input',
      defaultComponent: 'input',
      omitProps: ['control', 'size', 'valid'],
      style: function style(p) {
        return styled.css(["display:inline-block;border-width:", ";border-color:", ";border-style:solid;line-height:", ";color:", ";", ";background-color:", ";&[type='number']{padding-right:6px;}&::placeholder{color:", ";}&:focus{", "}&:disabled{background-color:", ";color:", ";}", ";", ";"], inputBorderWidth(p), inputBorderColor(p), inputLineHeight(p), inputTextColor(p), transitionBase(p), inputBgColor(p), inputPlaceholderText(p), baseFocus(primary(p))(p), inputDisabledBgColor(p), inputDisabledText(p), p.size && sizeStyle$1[p.size] && sizeStyle$1[p.size](p), p.control && controlStyle$1(p));
      },
      propTypes: {
        control: PropTypes__default.bool,
        size: PropTypes__default.oneOf(['sm', 'md', 'lg']),
        valid: PropTypes__default.bool
      },
      defaultProps: {
        size: 'md'
      }
    };
  });

  /* eslint-disable jsx-a11y/label-has-for */
  var Label =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'label',
      defaultComponent: 'label',
      style: function style() {
        return styled.css(["display:inline-block;margin-bottom:0.5rem;"]);
      },
      propTypes: {
        children: PropTypes__default.node
      }
    };
  });

  function _assertThisInitialized(self) {
    if (self === void 0) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return self;
  }

  var assertThisInitialized = _assertThisInitialized;

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  var defineProperty = _defineProperty;

  var toArray = function toArray(a) {
    var ret = Array(a.length);

    for (var i = 0; i < a.length; ++i) {
      ret[i] = a[i];
    }

    return ret;
  };
  var arrayFind = function arrayFind(array, search) {
    return array.filter(function (a) {
      return a === search;
    })[0];
  };
  var asArray = function asArray(a) {
    return Array.isArray(a) ? a : [a];
  };

  var tabSort = function tabSort(a, b) {
    var tabDiff = a.tabIndex - b.tabIndex;
    var indexDiff = a.index - b.index;

    if (tabDiff) {
      if (!a.tabIndex) return 1;
      if (!b.tabIndex) return -1;
    }

    return tabDiff || indexDiff;
  };
  var orderByTabIndex = function orderByTabIndex(nodes, filterNegative, keepGuards) {
    return toArray(nodes).map(function (node, index) {
      return {
        node: node,
        index: index,
        tabIndex: keepGuards && node.tabIndex === -1 ? (node.dataset || {}).focusGuard ? 0 : -1 : node.tabIndex
      };
    }).filter(function (data) {
      return !filterNegative || data.tabIndex >= 0;
    }).sort(tabSort);
  };

  var tabbables = ['button:enabled:not([readonly])', 'select:enabled:not([readonly])', 'textarea:enabled:not([readonly])', 'input:enabled:not([readonly])', 'a[href]', 'area[href]', 'iframe', 'object', 'embed', '[tabindex]', '[contenteditable]', '[autofocus]'];

  var FOCUS_GROUP = 'data-focus-lock';
  var FOCUS_DISABLED = 'data-focus-lock-disabled';
  var FOCUS_ALLOW = 'data-no-focus-lock';
  var FOCUS_AUTO = 'data-autofocus-inside';

  var queryTabbables =
  /*#__PURE__*/
  tabbables.join(',');
  var queryGuardTabbables = queryTabbables + ', [data-focus-guard]';
  var getFocusables = function getFocusables(parents, withGuards) {
    return parents.reduce(function (acc, parent) {
      return acc.concat( // add all tabbables inside
      toArray(parent.querySelectorAll(withGuards ? queryGuardTabbables : queryTabbables)), // add if node is tabble itself
      parent.parentNode ? toArray(parent.parentNode.querySelectorAll(tabbables.join(','))).filter(function (node) {
        return node === parent;
      }) : []);
    }, []);
  };
  var getParentAutofocusables = function getParentAutofocusables(parent) {
    var parentFocus = parent.querySelectorAll('[' + FOCUS_AUTO + ']');
    return toArray(parentFocus).map(function (node) {
      return getFocusables([node]);
    }).reduce(function (acc, nodes) {
      return acc.concat(nodes);
    }, []);
  };

  var isElementHidden = function isElementHidden(computedStyle) {
    if (!computedStyle || !computedStyle.getPropertyValue) {
      return false;
    }

    return computedStyle.getPropertyValue('display') === 'none' || computedStyle.getPropertyValue('visibility') === 'hidden';
  };

  var isVisible = function isVisible(node) {
    return !node || node === document || !isElementHidden(window.getComputedStyle(node, null)) && isVisible(node.parentNode);
  };
  var notHiddenInput = function notHiddenInput(node) {
    return !((node.tagName === 'INPUT' || node.tagName === 'BUTTON') && (node.type === 'hidden' || node.disabled));
  };

  var getParents = function getParents(node) {
    var parents = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
    parents.push(node);

    if (node.parentNode) {
      getParents(node.parentNode, parents);
    }

    return parents;
  };

  var getCommonParent = function getCommonParent(nodea, nodeb) {
    var parentsA = getParents(nodea);
    var parentsB = getParents(nodeb);

    for (var i = 0; i < parentsA.length; i += 1) {
      var currentParent = parentsA[i];

      if (parentsB.indexOf(currentParent) >= 0) {
        return currentParent;
      }
    }

    return false;
  };
  var filterFocusable = function filterFocusable(nodes) {
    return toArray(nodes).filter(function (node) {
      return isVisible(node);
    }).filter(function (node) {
      return notHiddenInput(node);
    });
  };
  var getTabbableNodes = function getTabbableNodes(topNodes, withGuards) {
    return orderByTabIndex(filterFocusable(getFocusables(topNodes, withGuards)), true, withGuards);
  };
  var getAllTabbableNodes = function getAllTabbableNodes(topNodes) {
    return orderByTabIndex(filterFocusable(getFocusables(topNodes)), false);
  };
  var parentAutofocusables = function parentAutofocusables(topNode) {
    return filterFocusable(getParentAutofocusables(topNode));
  };

  var isRadio = function isRadio(node) {
    return node.tagName === 'INPUT' && node.type === 'radio';
  };

  var findSelectedRadio = function findSelectedRadio(node, nodes) {
    return nodes.filter(isRadio).filter(function (el) {
      return el.name === node.name;
    }).filter(function (el) {
      return el.checked;
    })[0] || node;
  };

  var pickFirstFocus = function pickFirstFocus(nodes) {
    if (nodes[0] && nodes.length > 1) {
      if (isRadio(nodes[0]) && nodes[0].name) {
        return findSelectedRadio(nodes[0], nodes);
      }
    }

    return nodes[0];
  };

  var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) {
    return typeof obj;
  } : function (obj) {
    return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
  };

  var filterNested = function filterNested(nodes) {
    var l = nodes.length;

    for (var i = 0; i < l; i += 1) {
      var _loop = function _loop(j) {
        if (i !== j) {
          if (nodes[i].contains(nodes[j])) {
            return {
              v: filterNested(nodes.filter(function (x) {
                return x !== nodes[j];
              }))
            };
          }
        }
      };

      for (var j = 0; j < l; j += 1) {
        var _ret = _loop(j);

        if ((typeof _ret === 'undefined' ? 'undefined' : _typeof(_ret)) === "object") return _ret.v;
      }
    }

    return nodes;
  };

  var getTopParent = function getTopParent(node) {
    return node.parentNode ? getTopParent(node.parentNode) : node;
  };

  var getAllAffectedNodes = function getAllAffectedNodes(node) {
    var nodes = asArray(node);
    return nodes.filter(Boolean).reduce(function (acc, currentNode) {
      var group = currentNode.getAttribute(FOCUS_GROUP);
      acc.push.apply(acc, group ? filterNested(toArray(getTopParent(currentNode).querySelectorAll('[' + FOCUS_GROUP + '="' + group + '"]:not([' + FOCUS_DISABLED + '="disabled"])'))) : [currentNode]);
      return acc;
    }, []);
  };

  var findAutoFocused = function findAutoFocused(autoFocusables) {
    return function (node) {
      return !!node.autofocus || node.dataset && !!node.dataset.autofocus || autoFocusables.indexOf(node) >= 0;
    };
  };

  var isGuard = function isGuard(node) {
    return node.dataset && node.dataset.focusGuard;
  };

  var notAGuard = function notAGuard(node) {
    return !isGuard(node);
  };

  var newFocus = function newFocus(innerNodes, outerNodes, activeElement, lastNode, autoFocused) {
    var cnt = innerNodes.length;
    var firstFocus = innerNodes[0];
    var lastFocus = innerNodes[cnt - 1]; // focus is inside

    if (innerNodes.indexOf(activeElement) >= 0) {
      return undefined;
    }

    var activeIndex = outerNodes.indexOf(activeElement);
    var lastIndex = outerNodes.indexOf(lastNode || activeIndex);
    var lastNodeInside = innerNodes.indexOf(lastNode);
    var indexDiff = activeIndex - lastIndex;
    var firstNodeIndex = outerNodes.indexOf(firstFocus);
    var lastNodeIndex = outerNodes.indexOf(lastFocus); // new focus

    if (activeIndex === -1 || lastNodeInside === -1) {
      return innerNodes.indexOf(autoFocused.length ? pickFirstFocus(autoFocused) : pickFirstFocus(innerNodes));
    } // old focus


    if (!indexDiff && lastNodeInside >= 0) {
      return lastNodeInside;
    } // first element


    if (activeIndex <= firstNodeIndex && isGuard(activeElement) && Math.abs(indexDiff) > 1) {
      return 0;
    } // jump out


    if (indexDiff && Math.abs(indexDiff) > 1) {
      return lastNodeInside;
    } // focus above lock


    if (activeIndex <= firstNodeIndex) {
      return cnt - 1;
    } // focus below lock


    if (activeIndex > lastNodeIndex) {
      return 0;
    } // index is inside tab order, but outside Lock


    if (indexDiff) {
      if (Math.abs(indexDiff) > 1) {
        return lastNodeInside;
      }

      return (cnt + lastNodeInside + indexDiff) % cnt;
    } // do nothing


    return undefined;
  };

  var getTopCommonParent = function getTopCommonParent(baseActiveElement, leftEntry, rightEntries) {
    var activeElements = asArray(baseActiveElement);
    var leftEntries = asArray(leftEntry);
    var activeElement = activeElements[0];
    var topCommon = null;
    leftEntries.filter(Boolean).forEach(function (entry) {
      topCommon = getCommonParent(topCommon || entry, entry) || topCommon;
      rightEntries.filter(Boolean).forEach(function (subEntry) {
        var common = getCommonParent(activeElement, subEntry);

        if (common) {
          if (!topCommon || common.contains(topCommon)) {
            topCommon = common;
          } else {
            topCommon = getCommonParent(common, topCommon);
          }
        }
      });
    });
    return topCommon;
  };

  var allParentAutofocusables = function allParentAutofocusables(entries) {
    return entries.reduce(function (acc, node) {
      return acc.concat(parentAutofocusables(node));
    }, []);
  };

  var reorderNodes = function reorderNodes(srcNodes, dstNodes) {
    return srcNodes.map(function (dnode) {
      return dstNodes.find(function (_ref) {
        var node = _ref.node;
        return dnode === node;
      });
    }).filter(Boolean);
  };

  var getFocusabledIn = function getFocusabledIn(topNode) {
    var entries = getAllAffectedNodes(topNode).filter(notAGuard);
    var commonParent = getTopCommonParent(topNode, topNode, entries);
    var outerNodes = getTabbableNodes([commonParent], true);
    var innerElements = getTabbableNodes(entries).filter(function (_ref2) {
      var node = _ref2.node;
      return notAGuard(node);
    }).map(function (_ref3) {
      var node = _ref3.node;
      return node;
    });
    return outerNodes.map(function (_ref4) {
      var node = _ref4.node,
          index = _ref4.index;
      return {
        node: node,
        index: index,
        lockItem: innerElements.indexOf(node) >= 0,
        guard: isGuard(node)
      };
    });
  };

  var getFocusMerge = function getFocusMerge(topNode, lastNode) {
    var activeElement = document && document.activeElement;
    var entries = getAllAffectedNodes(topNode).filter(notAGuard);
    var commonParent = getTopCommonParent(activeElement || topNode, topNode, entries);
    var innerElements = getTabbableNodes(entries).filter(function (_ref5) {
      var node = _ref5.node;
      return notAGuard(node);
    });

    if (!innerElements[0]) {
      innerElements = getAllTabbableNodes(entries).filter(function (_ref6) {
        var node = _ref6.node;
        return notAGuard(node);
      });

      if (!innerElements[0]) {
        return undefined;
      }
    }

    var outerNodes = getTabbableNodes([commonParent]).map(function (_ref7) {
      var node = _ref7.node;
      return node;
    });
    var orderedInnerElements = reorderNodes(outerNodes, innerElements);
    var innerNodes = orderedInnerElements.map(function (_ref8) {
      var node = _ref8.node;
      return node;
    });
    var newId = newFocus(innerNodes, outerNodes, activeElement, lastNode, innerNodes.filter(findAutoFocused(allParentAutofocusables(entries))));

    if (newId === undefined) {
      return newId;
    }

    return orderedInnerElements[newId];
  };

  var focusInFrame = function focusInFrame(frame) {
    return frame === document.activeElement;
  };

  var focusInsideIframe = function focusInsideIframe(topNode) {
    return !!arrayFind(toArray(topNode.querySelectorAll('iframe')), focusInFrame);
  };

  var focusInside = function focusInside(topNode) {
    var activeElement = document && document.activeElement;

    if (!activeElement || activeElement.dataset && activeElement.dataset.focusGuard) {
      return false;
    }

    return getAllAffectedNodes(topNode).reduce(function (result, node) {
      return result || node.contains(activeElement) || focusInsideIframe(node);
    }, false);
  };

  var focusIsHidden = function focusIsHidden() {
    return document && toArray(document.querySelectorAll('[' + FOCUS_ALLOW + ']')).some(function (node) {
      return node.contains(document.activeElement);
    });
  };

  var focusOn = function focusOn(target) {
    target.focus();

    if (target.contentWindow) {
      target.contentWindow.focus();
    }
  };
  var guardCount = 0;
  var lockDisabled = false;
  var setFocus = (function (topNode, lastNode) {
    var focusable = getFocusMerge(topNode, lastNode);

    if (lockDisabled) {
      return;
    }

    if (focusable) {
      if (guardCount > 2) {
        // eslint-disable-next-line no-console
        console.error('FocusLock: focus-fighting detected. Only one focus management system could be active. ' + 'See https://github.com/theKashey/focus-lock/#focus-fighting');
        lockDisabled = true;
        setTimeout(function () {
          lockDisabled = false;
        }, 1);
        return;
      }

      guardCount++;
      focusOn(focusable.node);
      guardCount--;
    }
  });

  //
  var shallowequal = function shallowEqual(objA, objB, compare, compareContext) {
    var ret = compare ? compare.call(compareContext, objA, objB) : void 0;

    if (ret !== void 0) {
      return !!ret;
    }

    if (objA === objB) {
      return true;
    }

    if (typeof objA !== "object" || !objA || typeof objB !== "object" || !objB) {
      return false;
    }

    var keysA = Object.keys(objA);
    var keysB = Object.keys(objB);

    if (keysA.length !== keysB.length) {
      return false;
    }

    var bHasOwnProperty = Object.prototype.hasOwnProperty.bind(objB); // Test for A's keys different from B.

    for (var idx = 0; idx < keysA.length; idx++) {
      var key = keysA[idx];

      if (!bHasOwnProperty(key)) {
        return false;
      }

      var valueA = objA[key];
      var valueB = objB[key];
      ret = compare ? compare.call(compareContext, valueA, valueB, key) : void 0;

      if (ret === false || ret === void 0 && valueA !== valueB) {
        return false;
      }
    }

    return true;
  };

  function withSideEffect(reducePropsToState, handleStateChangeOnClient) {
    {
      if (typeof reducePropsToState !== 'function') {
        throw new Error('Expected reducePropsToState to be a function.');
      }

      if (typeof handleStateChangeOnClient !== 'function') {
        throw new Error('Expected handleStateChangeOnClient to be a function.');
      }
    }

    function getDisplayName(WrappedComponent) {
      return WrappedComponent.displayName || WrappedComponent.name || 'Component';
    }

    return function wrap(WrappedComponent) {
      {
        if (typeof WrappedComponent !== 'function') {
          throw new Error('Expected WrappedComponent to be a React component.');
        }
      }

      var mountedInstances = [];
      var state;

      function emitChange() {
        state = reducePropsToState(mountedInstances.map(function (instance) {
          return instance.props;
        }));
        handleStateChangeOnClient(state);
      }

      var SideEffect =
      /*#__PURE__*/
      function (_Component) {
        inheritsLoose(SideEffect, _Component);

        function SideEffect() {
          return _Component.apply(this, arguments) || this;
        } // Try to use displayName of wrapped component


        SideEffect.peek = function peek() {
          return state;
        };

        var _proto = SideEffect.prototype;

        _proto.shouldComponentUpdate = function shouldComponentUpdate(nextProps) {
          return !shallowequal(nextProps, this.props);
        };

        _proto.componentDidMount = function componentDidMount() {
          mountedInstances.push(this);
          emitChange();
        };

        _proto.componentDidUpdate = function componentDidUpdate() {
          emitChange();
        };

        _proto.componentWillUnmount = function componentWillUnmount() {
          var index = mountedInstances.indexOf(this);
          mountedInstances.splice(index, 1);
          emitChange();
        };

        _proto.render = function render() {
          return React__default.createElement(WrappedComponent, this.props);
        };

        return SideEffect;
      }(React.Component);

      defineProperty(SideEffect, "displayName", "SideEffect(" + getDisplayName(WrappedComponent) + ")");

      return SideEffect;
    };
  }

  function deferAction(action) {
    // Hidding setImmediate from Webpack to avoid inserting polyfill
    var setImmediate = window.setImmediate;

    if (typeof setImmediate !== 'undefined') {
      setImmediate(action);
    } else {
      setTimeout(action, 1);
    }
  }
  var inlineProp = function inlineProp(name, value) {
    var obj = {};
    obj[name] = value;
    return obj;
  };

  var focusOnBody = function focusOnBody() {
    return document && document.activeElement === document.body;
  };

  var isFreeFocus = function isFreeFocus() {
    return focusOnBody() || focusIsHidden();
  };

  var lastActiveTrap = null;
  var lastActiveFocus = null;
  var lastPortaledElement = null;
  var focusWasOutsideWindow = false;

  var defaultWhitelist = function defaultWhitelist() {
    return true;
  };

  var focusWhitelisted = function focusWhitelisted(activeElement) {
    return (lastActiveTrap.whiteList || defaultWhitelist)(activeElement);
  };

  var recordPortal = function recordPortal(observerNode, portaledElement) {
    lastPortaledElement = {
      observerNode: observerNode,
      portaledElement: portaledElement
    };
  };

  var focusIsPortaledPair = function focusIsPortaledPair(element) {
    return lastPortaledElement && lastPortaledElement.portaledElement === element;
  };

  function autoGuard(startIndex, end, step, allNodes) {
    var lastGuard = null;
    var i = startIndex;

    do {
      var item = allNodes[i];

      if (item.guard) {
        if (item.node.dataset.focusAutoGuard) {
          lastGuard = item;
        }
      } else if (item.lockItem) {
        if (i !== startIndex) {
          // we will tab to the next element
          return;
        }

        lastGuard = null;
      } else {
        break;
      }
    } while ((i += step) !== end);

    if (lastGuard) {
      lastGuard.node.tabIndex = 0;
    }
  }

  var extractRef = function extractRef(ref) {
    return ref && 'current' in ref ? ref.current : ref;
  };

  var activateTrap = function activateTrap() {
    var result = false;

    if (lastActiveTrap) {
      var _lastActiveTrap = lastActiveTrap,
          observed = _lastActiveTrap.observed,
          persistentFocus = _lastActiveTrap.persistentFocus,
          autoFocus = _lastActiveTrap.autoFocus,
          shards = _lastActiveTrap.shards;
      var workingNode = observed || lastPortaledElement && lastPortaledElement.portaledElement;
      var activeElement = document && document.activeElement;

      if (workingNode) {
        var workingArea = [workingNode].concat(shards.map(extractRef).filter(Boolean));

        if (!activeElement || focusWhitelisted(activeElement)) {
          if (persistentFocus || focusWasOutsideWindow || !isFreeFocus() || !lastActiveFocus && autoFocus) {
            if (workingNode && !(focusInside(workingArea) || focusIsPortaledPair(activeElement, workingNode))) {
              if (document && !lastActiveFocus && activeElement && !autoFocus) {
                activeElement.blur();
                document.body.focus();
              } else {
                result = setFocus(workingArea, lastActiveFocus);
                lastPortaledElement = {};
              }
            }

            focusWasOutsideWindow = false;
            lastActiveFocus = document && document.activeElement;
          }
        }

        if (document) {
          var newActiveElement = document && document.activeElement;
          var allNodes = getFocusabledIn(workingArea);
          var focusedItem = allNodes.find(function (_ref) {
            var node = _ref.node;
            return node === newActiveElement;
          });

          if (focusedItem) {
            // remove old focus
            allNodes.filter(function (_ref2) {
              var guard = _ref2.guard,
                  node = _ref2.node;
              return guard && node.dataset.focusAutoGuard;
            }).forEach(function (_ref3) {
              var node = _ref3.node;
              return node.removeAttribute('tabIndex');
            });
            var focusedIndex = allNodes.indexOf(focusedItem);
            autoGuard(focusedIndex, allNodes.length, +1, allNodes);
            autoGuard(focusedIndex, -1, -1, allNodes);
          }
        }
      }
    }

    return result;
  };

  var onTrap = function onTrap(event) {
    if (activateTrap() && event) {
      // prevent scroll jump
      event.stopPropagation();
      event.preventDefault();
    }
  };

  var onBlur = function onBlur() {
    return deferAction(activateTrap);
  };
  var onFocus = function onFocus(event) {
    // detect portal
    var source = event.target;
    var currentNode = event.currentTarget;

    if (!currentNode.contains(source)) {
      recordPortal(currentNode, source);
    }
  };

  var FocusWatcher = function FocusWatcher() {
    return null;
  };

  var FocusTrap = function FocusTrap(_ref4) {
    var children = _ref4.children;
    return React__default.createElement("div", {
      onBlur: onBlur,
      onFocus: onFocus
    }, children);
  };

  FocusTrap.propTypes = {
    children: PropTypes__default.node.isRequired
  };

  var onWindowBlur = function onWindowBlur() {
    focusWasOutsideWindow = true;
  };

  var attachHandler = function attachHandler() {
    document.addEventListener('focusin', onTrap, true);
    document.addEventListener('focusout', onBlur);
    window.addEventListener('blur', onWindowBlur);
  };

  var detachHandler = function detachHandler() {
    document.removeEventListener('focusin', onTrap, true);
    document.removeEventListener('focusout', onBlur);
    window.removeEventListener('blur', onWindowBlur);
  };

  function reducePropsToState(propsList) {
    return propsList.filter(function (_ref5) {
      var disabled = _ref5.disabled;
      return !disabled;
    }).slice(-1)[0];
  }

  function handleStateChangeOnClient(trap) {
    if (trap && !lastActiveTrap) {
      attachHandler();
    }

    var lastTrap = lastActiveTrap;
    var sameTrap = lastTrap && trap && trap.onActivation === lastTrap.onActivation;
    lastActiveTrap = trap;

    if (lastTrap && !sameTrap) {
      lastTrap.onDeactivation();
    }

    if (trap) {
      lastActiveFocus = null;

      if (!sameTrap || lastTrap.observed !== trap.observed) {
        trap.onActivation();
      }

      activateTrap(true);
      deferAction(activateTrap);
    } else {
      detachHandler();
      lastActiveFocus = null;
    }
  }

  var FocusTrap$1 = /*#__PURE__*/
  withSideEffect(reducePropsToState, handleStateChangeOnClient)(FocusWatcher);

  var hiddenGuard = {
    width: '1px',
    height: '0px',
    padding: 0,
    overflow: 'hidden',
    position: 'fixed',
    top: '1px',
    left: '1px'
  };

  var InFocusGuard = function InFocusGuard(_ref) {
    var children = _ref.children;
    return React__default.createElement(React__default.Fragment, null, React__default.createElement("div", {
      key: "guard-first",
      "data-focus-guard": true,
      "data-focus-auto-guard": true,
      style: hiddenGuard
    }), children, children && React__default.createElement("div", {
      key: "guard-last",
      "data-focus-guard": true,
      "data-focus-auto-guard": true,
      style: hiddenGuard
    }));
  };

  InFocusGuard.propTypes = {
    children: PropTypes__default.node
  };
  InFocusGuard.defaultProps = {
    children: null
  };

  var RenderChildren = function RenderChildren(_ref) {
    var children = _ref.children;
    return React__default.createElement("div", null, children);
  };

  RenderChildren.propTypes = {
    children: PropTypes__default.node.isRequired
  };
  var Fragment = React__default.Fragment ? React__default.Fragment : RenderChildren;
  var emptyArray = [];

  var FocusLock =
  /*#__PURE__*/
  function (_Component) {
    inheritsLoose(FocusLock, _Component);

    function FocusLock() {
      var _this;

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this = _Component.call.apply(_Component, [this].concat(args)) || this;

      defineProperty(assertThisInitialized(assertThisInitialized(_this)), "state", {
        observed: undefined
      });

      defineProperty(assertThisInitialized(assertThisInitialized(_this)), "onActivation", function () {
        _this.originalFocusedElement = _this.originalFocusedElement || document && document.activeElement;

        if (_this.state.observed && _this.props.onActivation) {
          _this.props.onActivation(_this.state.observed);
        }

        _this.isActive = true;
      });

      defineProperty(assertThisInitialized(assertThisInitialized(_this)), "onDeactivation", function () {
        _this.isActive = false;

        if (_this.props.returnFocus && _this.originalFocusedElement && _this.originalFocusedElement.focus) {
          _this.originalFocusedElement.focus();

          _this.originalFocusedElement = null;
        }

        if (_this.props.onDeactivation) {
          _this.props.onDeactivation(_this.state.observed);
        }
      });

      defineProperty(assertThisInitialized(assertThisInitialized(_this)), "onFocus", function (event) {
        if (_this.isActive) {
          onFocus(event);
        }
      });

      defineProperty(assertThisInitialized(assertThisInitialized(_this)), "onBlur", onBlur);

      defineProperty(assertThisInitialized(assertThisInitialized(_this)), "setObserveNode", function (observed) {
        if (_this.state.observed !== observed) {
          _this.setState({
            observed: observed
          });
        }
      });

      defineProperty(assertThisInitialized(assertThisInitialized(_this)), "isActive", false);

      defineProperty(assertThisInitialized(assertThisInitialized(_this)), "originalFocusedElement", null);

      return _this;
    }

    var _proto = FocusLock.prototype;

    _proto.render = function render() {
      var _extends2;

      var _this$props = this.props,
          children = _this$props.children,
          disabled = _this$props.disabled,
          noFocusGuards = _this$props.noFocusGuards,
          persistentFocus = _this$props.persistentFocus,
          autoFocus = _this$props.autoFocus,
          allowTextSelection = _this$props.allowTextSelection,
          group = _this$props.group,
          className = _this$props.className,
          whiteList = _this$props.whiteList,
          _this$props$shards = _this$props.shards,
          shards = _this$props$shards === void 0 ? emptyArray : _this$props$shards,
          _this$props$as = _this$props.as,
          Container = _this$props$as === void 0 ? 'div' : _this$props$as,
          _this$props$lockProps = _this$props.lockProps,
          containerProps = _this$props$lockProps === void 0 ? {} : _this$props$lockProps;
      var observed = this.state.observed;

      {
        if (typeof allowTextSelection !== 'undefined') {
          // eslint-disable-next-line no-console
          console.warn('React-Focus-Lock: allowTextSelection is deprecated and enabled by default');
        }
      }

      var lockProps = _extends_1((_extends2 = {}, _extends2[FOCUS_DISABLED] = disabled && 'disabled', _extends2[FOCUS_GROUP] = group, _extends2), containerProps);

      var hasLeadingGuards = noFocusGuards !== true;
      var hasTailingGuards = hasLeadingGuards && noFocusGuards !== 'tail';
      return React__default.createElement(Fragment, null, hasLeadingGuards && [React__default.createElement("div", {
        key: "guard-first",
        "data-focus-guard": true,
        tabIndex: disabled ? -1 : 0,
        style: hiddenGuard
      }), // nearest focus guard
      React__default.createElement("div", {
        key: "guard-nearest",
        "data-focus-guard": true,
        tabIndex: disabled ? -1 : 1,
        style: hiddenGuard
      })], React__default.createElement(Container, _extends_1({
        ref: this.setObserveNode
      }, lockProps, {
        className: className,
        onBlur: this.onBlur,
        onFocus: this.onFocus
      }), React__default.createElement(FocusTrap$1, {
        observed: observed,
        disabled: disabled,
        persistentFocus: persistentFocus,
        autoFocus: autoFocus,
        whiteList: whiteList,
        shards: shards,
        onActivation: this.onActivation,
        onDeactivation: this.onDeactivation
      }), children), hasTailingGuards && React__default.createElement("div", {
        "data-focus-guard": true,
        tabIndex: disabled ? -1 : 0,
        style: hiddenGuard
      }));
    };

    return FocusLock;
  }(React.Component);

  FocusLock.propTypes = {
    children: PropTypes__default.node.isRequired,
    disabled: PropTypes__default.bool,
    returnFocus: PropTypes__default.bool,
    noFocusGuards: PropTypes__default.bool,
    allowTextSelection: PropTypes__default.bool,
    autoFocus: PropTypes__default.bool,
    persistentFocus: PropTypes__default.bool,
    group: PropTypes__default.string,
    className: PropTypes__default.string,
    whiteList: PropTypes__default.func,
    shards:
    /*#__PURE__*/
    PropTypes__default.arrayOf(PropTypes__default.any),
    as:
    /*#__PURE__*/
    PropTypes__default.oneOfType([PropTypes__default.string, PropTypes__default.func, PropTypes__default.object]),
    lockProps: PropTypes__default.object,
    onActivation: PropTypes__default.func,
    onDeactivation: PropTypes__default.func
  };
  FocusLock.defaultProps = {
    disabled: false,
    returnFocus: false,
    noFocusGuards: false,
    autoFocus: true,
    persistentFocus: false,
    allowTextSelection: undefined,
    group: undefined,
    className: undefined,
    whiteList: undefined,
    shards: undefined,
    as: 'div',
    lockProps: {},
    onActivation: undefined,
    onDeactivation: undefined
  };

  var AutoFocusInside = function AutoFocusInside(_ref) {
    var disabled = _ref.disabled,
        children = _ref.children,
        className = _ref.className;
    return React__default.createElement("div", _extends_1({}, inlineProp(FOCUS_AUTO, !disabled), {
      className: className
    }), children);
  };

  AutoFocusInside.propTypes = {
    children: PropTypes__default.node.isRequired,
    disabled: PropTypes__default.bool,
    className: PropTypes__default.string
  };
  AutoFocusInside.defaultProps = {
    disabled: false,
    className: undefined
  };

  var MoveFocusInside =
  /*#__PURE__*/
  function (_Component) {
    inheritsLoose(MoveFocusInside, _Component);

    function MoveFocusInside() {
      var _this;

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this = _Component.call.apply(_Component, [this].concat(args)) || this;

      defineProperty(assertThisInitialized(assertThisInitialized(_this)), "setObserveNode", function (ref) {
        _this.observed = ref;

        _this.moveFocus();
      });

      return _this;
    }

    var _proto = MoveFocusInside.prototype;

    _proto.componentDidMount = function componentDidMount() {
      this.moveFocus();
    };

    _proto.componentDidUpdate = function componentDidUpdate(prevProps) {
      if (prevProps.disabled && !this.props.disabled) {
        this.moveFocus();
      }
    };

    _proto.moveFocus = function moveFocus() {
      var observed = this.observed;

      if (!this.props.disabled && observed) {
        if (!focusInside(observed)) {
          setFocus(observed, null);
        }
      }
    };

    _proto.render = function render() {
      var _this$props = this.props,
          children = _this$props.children,
          disabled = _this$props.disabled,
          className = _this$props.className;
      return React__default.createElement("div", _extends_1({}, inlineProp(FOCUS_AUTO, !disabled), {
        ref: this.setObserveNode,
        className: className
      }), children);
    };

    return MoveFocusInside;
  }(React.Component);

  defineProperty(MoveFocusInside, "defaultProps", {
    disabled: false,
    className: undefined
  });
  MoveFocusInside.propTypes = {
    children: PropTypes__default.node.isRequired,
    disabled: PropTypes__default.bool,
    className: PropTypes__default.string
  };

  var FreeFocusInside = function FreeFocusInside(_ref) {
    var children = _ref.children,
        className = _ref.className;
    return React__default.createElement("div", _extends_1({}, inlineProp(FOCUS_ALLOW, true), {
      className: className
    }), children);
  };

  FreeFocusInside.propTypes = {
    children: PropTypes__default.node.isRequired,
    className: PropTypes__default.string
  };
  FreeFocusInside.defaultProps = {
    disabled: false,
    className: undefined
  };

  /*! *****************************************************************************
  Copyright (c) Microsoft Corporation. All rights reserved.
  Licensed under the Apache License, Version 2.0 (the "License"); you may not use
  this file except in compliance with the License. You may obtain a copy of the
  License at http://www.apache.org/licenses/LICENSE-2.0

  THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
  WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
  MERCHANTABLITY OR NON-INFRINGEMENT.

  See the Apache Version 2.0 License for specific language governing permissions
  and limitations under the License.
  ***************************************************************************** */

  /* global Reflect, Promise */
  var _extendStatics = function extendStatics(d, b) {
    _extendStatics = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (d, b) {
      d.__proto__ = b;
    } || function (d, b) {
      for (var p in b) {
        if (b.hasOwnProperty(p)) d[p] = b[p];
      }
    };

    return _extendStatics(d, b);
  };

  function __extends(d, b) {
    _extendStatics(d, b);

    function __() {
      this.constructor = d;
    }

    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
  }

  var _assign = function __assign() {
    _assign = Object.assign || function __assign(t) {
      for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];

        for (var p in s) {
          if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
      }

      return t;
    };

    return _assign.apply(this, arguments);
  };

  function makeStyleTag() {
    if (!document) return null;
    var tag = document.createElement('style');
    tag.type = 'text/css';
    return tag;
  }

  function injectStyles(tag, css) {
    if (tag.styleSheet) {
      tag.styleSheet.cssText = css;
    } else {
      tag.appendChild(document.createTextNode(css));
    }
  }

  function insertStyleTag(tag) {
    var head = document.head || document.getElementsByTagName('head')[0];
    head.appendChild(tag);
  }

  var stylesheetSinglentone = function stylesheetSinglentone() {
    var counter = 0;
    var stylesheet = null;
    return {
      add: function add(style) {
        if (counter == 0) {
          if (stylesheet = makeStyleTag()) {
            injectStyles(stylesheet, style);
            insertStyleTag(stylesheet);
          }
        }

        counter++;
      },
      remove: function remove() {
        counter--;

        if (!counter && stylesheet) {
          stylesheet.parentNode && stylesheet.parentNode.removeChild(stylesheet);
          stylesheet = null;
        }
      }
    };
  };

  var styleSinglentone = function styleSinglentone() {
    var sheet = stylesheetSinglentone();
    return function (_super) {
      __extends(Sheet, _super);

      function Sheet() {
        return _super !== null && _super.apply(this, arguments) || this;
      }

      Sheet.prototype.componentDidMount = function () {
        sheet.add(this.props.styles);
      };

      Sheet.prototype.componentWillUnmount = function () {
        sheet.remove();
      };

      Sheet.prototype.render = function () {
        return null;
      };

      return Sheet;
    }(React.PureComponent);
  };

  var zeroGap = {
    left: 0,
    top: 0,
    right: 0,
    gap: 0
  };

  var getOffset = function getOffset(gapMode) {
    var cs = window.getComputedStyle(document.body);
    var left = cs[gapMode === 'padding' ? 'paddingLeft' : 'marginLeft'];
    var top = cs[gapMode === 'padding' ? 'paddingTop' : 'marginTop'];
    var right = cs[gapMode === 'padding' ? 'paddingRight' : 'marginRight'];
    return [parseInt(left || '', 10) || 0, parseInt(top || '', 10) || 0, parseInt(right || '', 10) || 0];
  };

  var getGapWidth = function getGapWidth(gapMode) {
    if (gapMode === void 0) {
      gapMode = 'margin';
    }

    if (typeof window === 'undefined') {
      return zeroGap;
    }

    var offsets = getOffset(gapMode);
    var documentWidth = document.documentElement.clientWidth;
    var windowWidth = window.innerWidth;
    return {
      left: offsets[0],
      top: offsets[1],
      right: offsets[2],
      gap: Math.max(0, windowWidth - documentWidth + offsets[2] - offsets[0])
    };
  };

  var Style =
  /*#__PURE__*/
  styleSinglentone();

  var getStyles = function getStyles(_a, allowRelative, gapMode, important) {
    var left = _a.left,
        top = _a.top,
        right = _a.right,
        gap = _a.gap;

    if (gapMode === void 0) {
      gapMode = 'margin';
    }

    return "\n  body {\n    overflow: hidden " + important + ";\n    " + [allowRelative && "position: relative " + important + ";", gapMode == 'margin' && "\n    padding-left: " + left + "px;\n    padding-top: " + top + "px;\n    padding-right: " + right + "px;\n    margin-left:0;\n    margin-top:0;\n    margin-right: " + gap + "px " + important + ";\n    ", gapMode == 'padding' && "padding-right: " + gap + "px " + important + ";"].filter(Boolean).join('') + "\n  }\n  \n  .right-scroll-bar-position {\n    right: " + gap + "px " + important + ";\n  }\n  \n  .width-before-scroll-bar {\n    margin-right: " + gap + "px " + important + ";\n  }\n  \n  .right-scroll-bar-position .right-scroll-bar-position {\n    right: 0 " + important + ";\n  }\n  \n  .width-before-scroll-bar .width-before-scroll-bar {\n    margin-right: 0 " + important + ";\n  }\n";
  };

  var zeroRightClassName = 'right-scroll-bar-position';
  var fullWidthClassName = 'width-before-scroll-bar';

  var RemoveScrollBar =
  /*#__PURE__*/
  function (_super) {
    __extends(RemoveScrollBar, _super);

    function RemoveScrollBar() {
      var _this = _super !== null && _super.apply(this, arguments) || this;

      _this.state = {
        gap: getGapWidth(_this.props.gapMode)
      };

      _this.onResize = function () {
        _this.forceUpdate();

        if (_this.state.gap && _this.props.dynamic) {
          if (window.innerHeight > document.body.offsetHeight) {
            _this.setState({
              gap: zeroGap
            });
          }
        }
      };

      return _this;
    }

    RemoveScrollBar.prototype.componentDidMount = function () {
      var gap = getGapWidth(this.props.gapMode);

      if (gap !== this.state.gap) {
        this.setState({
          gap: gap
        });
      }

      if (typeof window !== 'undefined') {
        window.addEventListener('resize', this.onResize);
      }
    };

    RemoveScrollBar.prototype.componentWillUnmount = function () {
      if (typeof window !== 'undefined') {
        window.removeEventListener('resize', this.onResize);
      }
    };

    RemoveScrollBar.prototype.componentDidUpdate = function () {
      if (!this.state.gap) {
        var gap = getGapWidth(this.props.gapMode);

        if (gap !== this.state.gap) {
          this.setState({
            gap: gap
          });
        }
      }
    };

    RemoveScrollBar.prototype.render = function () {
      var _a = this.props,
          noRelative = _a.noRelative,
          noImportant = _a.noImportant,
          _b = _a.gapMode,
          gapMode = _b === void 0 ? 'margin' : _b;
      var gap = this.state.gap;
      return React.createElement(Style, {
        styles: getStyles(gap, !noRelative, gapMode, !noImportant ? "!important" : '')
      });
    };

    return RemoveScrollBar;
  }(React.Component);

  var elementCouldBeScrolled = function elementCouldBeScrolled(node) {
    return window.getComputedStyle(node).overflowY !== 'hidden';
  };

  var handleScroll = function handleScroll(endTarget, event, sourceDelta) {
    var delta = sourceDelta;
    var target = event.target;
    var targetInLock = endTarget.contains(target);
    var shouldCancelScroll = false;
    var isDeltaPositive = delta > 0;
    var availableScroll = 0;
    var availableScrollTop = 0;

    do {
      var scrollTop = target.scrollTop,
          scrollHeight = target.scrollHeight,
          clientHeight = target.clientHeight;
      var elementScroll = scrollHeight - clientHeight - scrollTop;

      if (scrollTop || elementScroll) {
        if (elementCouldBeScrolled(target)) {
          availableScroll += elementScroll;
          availableScrollTop += scrollTop;
        }
      }

      target = target.parentNode;
    } while (!targetInLock && target !== document.body || targetInLock && (endTarget.contains(target) || endTarget === target));

    if (isDeltaPositive && delta > availableScroll) {
      shouldCancelScroll = true;
    } else if (!isDeltaPositive && -delta > availableScrollTop) {
      shouldCancelScroll = true;
    }

    return shouldCancelScroll;
  };

  var passiveSupported = false;

  if (typeof window !== 'undefined') {
    try {
      var options =
      /*#__PURE__*/
      Object.defineProperty({}, 'passive', {
        get: function get() {
          passiveSupported = true;
          return true;
        }
      });
      window.addEventListener('test', options, options);
      window.removeEventListener('test', options, options);
    } catch (err) {
      passiveSupported = false;
    }
  }

  var aggressive = passiveSupported ? {
    passive: false
  } : false;

  var getTouchY = function getTouchY(event) {
    return event.changedTouches ? event.changedTouches[0].clientY : 0;
  };
  var classNames = {
    fullWidth: fullWidthClassName,
    zeroRight: zeroRightClassName
  };

  var extractRef$1 = function extractRef(ref) {
    return ref && 'current' in ref ? ref.current : ref;
  };

  var RemoveScroll =
  /*#__PURE__*/
  function (_super) {
    __extends(RemoveScroll, _super);

    function RemoveScroll() {
      var _this = _super !== null && _super.apply(this, arguments) || this;

      _this.shouldPreventQueue = [];
      _this.touchStart = 0;
      _this.ref = React.createRef();

      _this.shouldPrevent = function (event) {
        var stack = RemoveScroll.stack.filter(function (el) {
          return el.props.enabled;
        });

        if (!stack.length || stack[stack.length - 1] !== _this) {
          return;
        }

        var delta = event.deltaY || getTouchY(event);

        var sourceEvent = _this.shouldPreventQueue.filter(function (e) {
          return e.name === event.type && e.delta === delta && e.target === event.target;
        })[0];

        if (sourceEvent && sourceEvent.should) {
          event.preventDefault();
          return;
        }

        if (!sourceEvent) {
          var shardNodes = (_this.props.shards || []).map(extractRef$1).filter(Boolean).filter(function (node) {
            return node.contains(event.target);
          });
          var shouldStop = shardNodes.length > 0 ? _this.shouldCancelEvent(event, shardNodes[0]) : !_this.props.noIsolation;

          if (shouldStop) {
            event.preventDefault();
          }
        }
      };

      _this.shouldCancel = function (name, delta, target, should) {
        var event = {
          name: name,
          delta: delta,
          target: target,
          should: should
        };

        _this.shouldPreventQueue.push(event);

        setTimeout(function () {
          _this.shouldPreventQueue = _this.shouldPreventQueue.filter(function (e) {
            return e !== event;
          });
        }, 1);
      };

      _this.scrollTouchStart = function (event) {
        _this.touchStart = getTouchY(event);
      };

      _this.scrollWheel = function (event) {
        _this.shouldCancel(event.type, event.deltaY, event.target, _this.shouldCancelEvent(event, _this.ref.current));
      };

      _this.scrollTouchMove = function (event) {
        _this.shouldCancel(event.type, getTouchY(event), event.target, _this.shouldCancelEvent(event, _this.ref.current));
      };

      return _this;
    }

    RemoveScroll.prototype.componentDidMount = function () {
      RemoveScroll.stack.push(this);
      this.componentDidUpdate({
        enabled: false
      });
    };

    RemoveScroll.prototype.componentWillUnmount = function () {
      var _this = this;

      RemoveScroll.stack = RemoveScroll.stack.filter(function (inst) {
        return inst !== _this;
      });
      this.disable();
    };

    RemoveScroll.prototype.componentDidUpdate = function (oldProps) {
      if (oldProps.enabled !== this.props.enabled) {
        if (this.props.enabled) {
          this.enable();
        } else {
          this.disable();
        }
      }
    };

    RemoveScroll.prototype.enable = function () {
      if (typeof document !== 'undefined') {
        document.addEventListener('wheel', this.shouldPrevent, aggressive);
        document.addEventListener('touchmove', this.shouldPrevent, aggressive);
        document.addEventListener('touchstart', this.scrollTouchStart, aggressive);
      }
    };

    RemoveScroll.prototype.disable = function () {
      if (typeof window !== 'undefined') {
        document.removeEventListener('wheel', this.shouldPrevent, aggressive);
        document.removeEventListener('touchmove', this.shouldPrevent, aggressive);
        document.removeEventListener('touchstart', this.scrollTouchStart, aggressive);
      }
    };

    RemoveScroll.prototype.shouldCancelEvent = function (event, parent) {
      switch (event.type) {
        case 'wheel':
        case 'scroll':
          return handleScroll(parent, event, event.deltaY);

        case 'touchmove':
          return handleScroll(parent, event, this.touchStart - getTouchY(event));
      }

      return false;
    };

    RemoveScroll.prototype.render = function () {
      var _a = this.props,
          forwardProps = _a.forwardProps,
          children = _a.children,
          className = _a.className,
          removeScrollBar = _a.removeScrollBar,
          enabled = _a.enabled;
      var props = {
        ref: this.ref,
        onScrollCapture: this.scrollWheel,
        onWheelCapture: this.scrollWheel,
        onTouchMoveCapture: this.scrollTouchMove
      };
      return React.createElement(React.Fragment, null, removeScrollBar && enabled && React.createElement(RemoveScrollBar, {
        gapMode: "margin"
      }), forwardProps ? React.cloneElement(React.Children.only(children), props) : React.createElement("div", _assign({}, props, {
        className: className
      }), children));
    };

    RemoveScroll.classNames = classNames;
    RemoveScroll.defaultProps = {
      enabled: true,
      removeScrollBar: true
    };
    RemoveScroll.stack = [];
    return RemoveScroll;
  }(React.Component);

  function _objectWithoutPropertiesLoose$1(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;

    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }

    return target;
  }

  function _inheritsLoose$1(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    subClass.__proto__ = superClass;
  }

  var timeoutsShape = PropTypes__default.oneOfType([PropTypes__default.number,
  /*#__PURE__*/
  PropTypes__default.shape({
    enter: PropTypes__default.number,
    exit: PropTypes__default.number,
    appear: PropTypes__default.number
  }).isRequired]);
  var classNamesShape = PropTypes__default.oneOfType([PropTypes__default.string,
  /*#__PURE__*/
  PropTypes__default.shape({
    enter: PropTypes__default.string,
    exit: PropTypes__default.string,
    active: PropTypes__default.string
  }),
  /*#__PURE__*/
  PropTypes__default.shape({
    enter: PropTypes__default.string,
    enterDone: PropTypes__default.string,
    enterActive: PropTypes__default.string,
    exit: PropTypes__default.string,
    exitDone: PropTypes__default.string,
    exitActive: PropTypes__default.string
  })]);

  var TransitionGroupContext = /*#__PURE__*/
  React__default.createContext(null);

  var UNMOUNTED = 'unmounted';
  var EXITED = 'exited';
  var ENTERING = 'entering';
  var ENTERED = 'entered';
  var EXITING = 'exiting';
  /**
   * The Transition component lets you describe a transition from one component
   * state to another _over time_ with a simple declarative API. Most commonly
   * it's used to animate the mounting and unmounting of a component, but can also
   * be used to describe in-place transition states as well.
   *
   * ---
   *
   * **Note**: `Transition` is a platform-agnostic base component. If you're using
   * transitions in CSS, you'll probably want to use
   * [`CSSTransition`](https://reactcommunity.org/react-transition-group/css-transition)
   * instead. It inherits all the features of `Transition`, but contains
   * additional features necessary to play nice with CSS transitions (hence the
   * name of the component).
   *
   * ---
   *
   * By default the `Transition` component does not alter the behavior of the
   * component it renders, it only tracks "enter" and "exit" states for the
   * components. It's up to you to give meaning and effect to those states. For
   * example we can add styles to a component when it enters or exits:
   *
   * ```jsx
   * import { Transition } from 'react-transition-group';
   *
   * const duration = 300;
   *
   * const defaultStyle = {
   *   transition: `opacity ${duration}ms ease-in-out`,
   *   opacity: 0,
   * }
   *
   * const transitionStyles = {
   *   entering: { opacity: 1 },
   *   entered:  { opacity: 1 },
   *   exiting:  { opacity: 0 },
   *   exited:  { opacity: 0 },
   * };
   *
   * const Fade = ({ in: inProp }) => (
   *   <Transition in={inProp} timeout={duration}>
   *     {state => (
   *       <div style={{
   *         ...defaultStyle,
   *         ...transitionStyles[state]
   *       }}>
   *         I'm a fade Transition!
   *       </div>
   *     )}
   *   </Transition>
   * );
   * ```
   *
   * There are 4 main states a Transition can be in:
   *  - `'entering'`
   *  - `'entered'`
   *  - `'exiting'`
   *  - `'exited'`
   *
   * Transition state is toggled via the `in` prop. When `true` the component
   * begins the "Enter" stage. During this stage, the component will shift from
   * its current transition state, to `'entering'` for the duration of the
   * transition and then to the `'entered'` stage once it's complete. Let's take
   * the following example (we'll use the
   * [useState](https://reactjs.org/docs/hooks-reference.html#usestate) hook):
   *
   * ```jsx
   * function App() {
   *   const [inProp, setInProp] = useState(false);
   *   return (
   *     <div>
   *       <Transition in={inProp} timeout={500}>
   *         {state => (
   *           // ...
   *         )}
   *       </Transition>
   *       <button onClick={() => setInProp(true)}>
   *         Click to Enter
   *       </button>
   *     </div>
   *   );
   * }
   * ```
   *
   * When the button is clicked the component will shift to the `'entering'` state
   * and stay there for 500ms (the value of `timeout`) before it finally switches
   * to `'entered'`.
   *
   * When `in` is `false` the same thing happens except the state moves from
   * `'exiting'` to `'exited'`.
   */

  var Transition =
  /*#__PURE__*/
  function (_React$Component) {
    _inheritsLoose$1(Transition, _React$Component);

    function Transition(props, context) {
      var _this;

      _this = _React$Component.call(this, props, context) || this;
      var parentGroup = context; // In the context of a TransitionGroup all enters are really appears

      var appear = parentGroup && !parentGroup.isMounting ? props.enter : props.appear;
      var initialStatus;
      _this.appearStatus = null;

      if (props["in"]) {
        if (appear) {
          initialStatus = EXITED;
          _this.appearStatus = ENTERING;
        } else {
          initialStatus = ENTERED;
        }
      } else {
        if (props.unmountOnExit || props.mountOnEnter) {
          initialStatus = UNMOUNTED;
        } else {
          initialStatus = EXITED;
        }
      }

      _this.state = {
        status: initialStatus
      };
      _this.nextCallback = null;
      return _this;
    }

    Transition.getDerivedStateFromProps = function getDerivedStateFromProps(_ref, prevState) {
      var nextIn = _ref["in"];

      if (nextIn && prevState.status === UNMOUNTED) {
        return {
          status: EXITED
        };
      }

      return null;
    }; // getSnapshotBeforeUpdate(prevProps) {
    //   let nextStatus = null
    //   if (prevProps !== this.props) {
    //     const { status } = this.state
    //     if (this.props.in) {
    //       if (status !== ENTERING && status !== ENTERED) {
    //         nextStatus = ENTERING
    //       }
    //     } else {
    //       if (status === ENTERING || status === ENTERED) {
    //         nextStatus = EXITING
    //       }
    //     }
    //   }
    //   return { nextStatus }
    // }


    var _proto = Transition.prototype;

    _proto.componentDidMount = function componentDidMount() {
      this.updateStatus(true, this.appearStatus);
    };

    _proto.componentDidUpdate = function componentDidUpdate(prevProps) {
      var nextStatus = null;

      if (prevProps !== this.props) {
        var status = this.state.status;

        if (this.props["in"]) {
          if (status !== ENTERING && status !== ENTERED) {
            nextStatus = ENTERING;
          }
        } else {
          if (status === ENTERING || status === ENTERED) {
            nextStatus = EXITING;
          }
        }
      }

      this.updateStatus(false, nextStatus);
    };

    _proto.componentWillUnmount = function componentWillUnmount() {
      this.cancelNextCallback();
    };

    _proto.getTimeouts = function getTimeouts() {
      var timeout = this.props.timeout;
      var exit, enter, appear;
      exit = enter = appear = timeout;

      if (timeout != null && typeof timeout !== 'number') {
        exit = timeout.exit;
        enter = timeout.enter; // TODO: remove fallback for next major

        appear = timeout.appear !== undefined ? timeout.appear : enter;
      }

      return {
        exit: exit,
        enter: enter,
        appear: appear
      };
    };

    _proto.updateStatus = function updateStatus(mounting, nextStatus) {
      if (mounting === void 0) {
        mounting = false;
      }

      if (nextStatus !== null) {
        // nextStatus will always be ENTERING or EXITING.
        this.cancelNextCallback();
        var node = ReactDOM__default.findDOMNode(this);

        if (nextStatus === ENTERING) {
          this.performEnter(node, mounting);
        } else {
          this.performExit(node);
        }
      } else if (this.props.unmountOnExit && this.state.status === EXITED) {
        this.setState({
          status: UNMOUNTED
        });
      }
    };

    _proto.performEnter = function performEnter(node, mounting) {
      var _this2 = this;

      var enter = this.props.enter;
      var appearing = this.context ? this.context.isMounting : mounting;
      var timeouts = this.getTimeouts();
      var enterTimeout = appearing ? timeouts.appear : timeouts.enter; // no enter animation skip right to ENTERED
      // if we are mounting and running this it means appear _must_ be set

      if (!mounting && !enter) {
        this.safeSetState({
          status: ENTERED
        }, function () {
          _this2.props.onEntered(node);
        });
        return;
      }

      this.props.onEnter(node, appearing);
      this.safeSetState({
        status: ENTERING
      }, function () {
        _this2.props.onEntering(node, appearing);

        _this2.onTransitionEnd(node, enterTimeout, function () {
          _this2.safeSetState({
            status: ENTERED
          }, function () {
            _this2.props.onEntered(node, appearing);
          });
        });
      });
    };

    _proto.performExit = function performExit(node) {
      var _this3 = this;

      var exit = this.props.exit;
      var timeouts = this.getTimeouts(); // no exit animation skip right to EXITED

      if (!exit) {
        this.safeSetState({
          status: EXITED
        }, function () {
          _this3.props.onExited(node);
        });
        return;
      }

      this.props.onExit(node);
      this.safeSetState({
        status: EXITING
      }, function () {
        _this3.props.onExiting(node);

        _this3.onTransitionEnd(node, timeouts.exit, function () {
          _this3.safeSetState({
            status: EXITED
          }, function () {
            _this3.props.onExited(node);
          });
        });
      });
    };

    _proto.cancelNextCallback = function cancelNextCallback() {
      if (this.nextCallback !== null) {
        this.nextCallback.cancel();
        this.nextCallback = null;
      }
    };

    _proto.safeSetState = function safeSetState(nextState, callback) {
      // This shouldn't be necessary, but there are weird race conditions with
      // setState callbacks and unmounting in testing, so always make sure that
      // we can cancel any pending setState callbacks after we unmount.
      callback = this.setNextCallback(callback);
      this.setState(nextState, callback);
    };

    _proto.setNextCallback = function setNextCallback(callback) {
      var _this4 = this;

      var active = true;

      this.nextCallback = function (event) {
        if (active) {
          active = false;
          _this4.nextCallback = null;
          callback(event);
        }
      };

      this.nextCallback.cancel = function () {
        active = false;
      };

      return this.nextCallback;
    };

    _proto.onTransitionEnd = function onTransitionEnd(node, timeout, handler) {
      this.setNextCallback(handler);
      var doesNotHaveTimeoutOrListener = timeout == null && !this.props.addEndListener;

      if (!node || doesNotHaveTimeoutOrListener) {
        setTimeout(this.nextCallback, 0);
        return;
      }

      if (this.props.addEndListener) {
        this.props.addEndListener(node, this.nextCallback);
      }

      if (timeout != null) {
        setTimeout(this.nextCallback, timeout);
      }
    };

    _proto.render = function render() {
      var status = this.state.status;

      if (status === UNMOUNTED) {
        return null;
      }

      var _this$props = this.props,
          children = _this$props.children,
          childProps = _objectWithoutPropertiesLoose$1(_this$props, ["children"]); // filter props for Transtition


      delete childProps["in"];
      delete childProps.mountOnEnter;
      delete childProps.unmountOnExit;
      delete childProps.appear;
      delete childProps.enter;
      delete childProps.exit;
      delete childProps.timeout;
      delete childProps.addEndListener;
      delete childProps.onEnter;
      delete childProps.onEntering;
      delete childProps.onEntered;
      delete childProps.onExit;
      delete childProps.onExiting;
      delete childProps.onExited;

      if (typeof children === 'function') {
        // allows for nested Transitions
        return React__default.createElement(TransitionGroupContext.Provider, {
          value: null
        }, children(status, childProps));
      }

      var child = React__default.Children.only(children);
      return (// allows for nested Transitions
        React__default.createElement(TransitionGroupContext.Provider, {
          value: null
        }, React__default.cloneElement(child, childProps))
      );
    };

    return Transition;
  }(React__default.Component);

  Transition.contextType = TransitionGroupContext;
  Transition.propTypes = {
    /**
     * A `function` child can be used instead of a React element. This function is
     * called with the current transition status (`'entering'`, `'entered'`,
     * `'exiting'`, `'exited'`, `'unmounted'`), which can be used to apply context
     * specific props to a component.
     *
     * ```jsx
     * <Transition in={this.state.in} timeout={150}>
     *   {state => (
     *     <MyComponent className={`fade fade-${state}`} />
     *   )}
     * </Transition>
     * ```
     */
    children:
    /*#__PURE__*/
    PropTypes.oneOfType([PropTypes.func.isRequired, PropTypes.element.isRequired]).isRequired,

    /**
     * Show the component; triggers the enter or exit states
     */
    "in": PropTypes.bool,

    /**
     * By default the child component is mounted immediately along with
     * the parent `Transition` component. If you want to "lazy mount" the component on the
     * first `in={true}` you can set `mountOnEnter`. After the first enter transition the component will stay
     * mounted, even on "exited", unless you also specify `unmountOnExit`.
     */
    mountOnEnter: PropTypes.bool,

    /**
     * By default the child component stays mounted after it reaches the `'exited'` state.
     * Set `unmountOnExit` if you'd prefer to unmount the component after it finishes exiting.
     */
    unmountOnExit: PropTypes.bool,

    /**
     * Normally a component is not transitioned if it is shown when the `<Transition>` component mounts.
     * If you want to transition on the first mount set `appear` to `true`, and the
     * component will transition in as soon as the `<Transition>` mounts.
     *
     * > Note: there are no specific "appear" states. `appear` only adds an additional `enter` transition.
     */
    appear: PropTypes.bool,

    /**
     * Enable or disable enter transitions.
     */
    enter: PropTypes.bool,

    /**
     * Enable or disable exit transitions.
     */
    exit: PropTypes.bool,

    /**
     * The duration of the transition, in milliseconds.
     * Required unless `addEndListener` is provided.
     *
     * You may specify a single timeout for all transitions:
     *
     * ```jsx
     * timeout={500}
     * ```
     *
     * or individually:
     *
     * ```jsx
     * timeout={{
     *  appear: 500,
     *  enter: 300,
     *  exit: 500,
     * }}
     * ```
     *
     * - `appear` defaults to the value of `enter`
     * - `enter` defaults to `0`
     * - `exit` defaults to `0`
     *
     * @type {number | { enter?: number, exit?: number, appear?: number }}
     */
    timeout: function timeout(props) {
      var pt = timeoutsShape;
      if (!props.addEndListener) pt = pt.isRequired;

      for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }

      return pt.apply(void 0, [props].concat(args));
    },

    /**
     * Add a custom transition end trigger. Called with the transitioning
     * DOM node and a `done` callback. Allows for more fine grained transition end
     * logic. **Note:** Timeouts are still used as a fallback if provided.
     *
     * ```jsx
     * addEndListener={(node, done) => {
     *   // use the css transitionend event to mark the finish of a transition
     *   node.addEventListener('transitionend', done, false);
     * }}
     * ```
     */
    addEndListener: PropTypes.func,

    /**
     * Callback fired before the "entering" status is applied. An extra parameter
     * `isAppearing` is supplied to indicate if the enter stage is occurring on the initial mount
     *
     * @type Function(node: HtmlElement, isAppearing: bool) -> void
     */
    onEnter: PropTypes.func,

    /**
     * Callback fired after the "entering" status is applied. An extra parameter
     * `isAppearing` is supplied to indicate if the enter stage is occurring on the initial mount
     *
     * @type Function(node: HtmlElement, isAppearing: bool)
     */
    onEntering: PropTypes.func,

    /**
     * Callback fired after the "entered" status is applied. An extra parameter
     * `isAppearing` is supplied to indicate if the enter stage is occurring on the initial mount
     *
     * @type Function(node: HtmlElement, isAppearing: bool) -> void
     */
    onEntered: PropTypes.func,

    /**
     * Callback fired before the "exiting" status is applied.
     *
     * @type Function(node: HtmlElement) -> void
     */
    onExit: PropTypes.func,

    /**
     * Callback fired after the "exiting" status is applied.
     *
     * @type Function(node: HtmlElement) -> void
     */
    onExiting: PropTypes.func,

    /**
     * Callback fired after the "exited" status is applied.
     *
     * @type Function(node: HtmlElement) -> void
     */
    onExited: PropTypes.func // Name the function so it is clearer in the documentation

  };

  function noop() {}

  Transition.defaultProps = {
    "in": false,
    mountOnEnter: false,
    unmountOnExit: false,
    appear: false,
    enter: true,
    exit: true,
    onEnter: noop,
    onEntering: noop,
    onEntered: noop,
    onExit: noop,
    onExiting: noop,
    onExited: noop
  };
  Transition.UNMOUNTED = 0;
  Transition.EXITED = 1;
  Transition.ENTERING = 2;
  Transition.ENTERED = 3;
  Transition.EXITING = 4;

  function useNode(type) {
    var _useState = React.useState(null),
        node = _useState[0],
        setNode = _useState[1];

    React.useEffect(function () {
      var node = document.createElement(type);
      document.body.appendChild(node);
      setNode(node);
      return function () {
        return document.body.removeChild(node);
      };
    }, [type]);
    return node;
  }

  function create$2() {
    function Portal(_ref) {
      var children = _ref.children,
          _ref$type = _ref.type,
          type = _ref$type === void 0 ? 'sui-portal' : _ref$type;
      var node = useNode(type);
      if (!node) return null;
      return ReactDOM.createPortal(children, node);
    }

    Portal.propTypes = {
      type: PropTypes__default.string,
      children: PropTypes__default.node
    };
    return Portal;
  }

  var Portal = /*#__PURE__*/
  create$2();

  var ModalContext = /*#__PURE__*/
  React.createContext();

  function createModalComponent() {
    var createAriaHider = function createAriaHider(dialogNode) {
      var originalValues = [];
      var rootNodes = [];
      Array.prototype.forEach.call(document.querySelectorAll('body > *'), function (node) {
        if (node === dialogNode.parentNode) {
          return;
        }

        var attr = node.getAttribute('aria-hidden');
        var alreadyHidden = attr !== null && attr !== 'false';

        if (alreadyHidden) {
          return;
        }

        originalValues.push(attr);
        rootNodes.push(node);
        node.setAttribute('aria-hidden', 'true');
      });
      return function () {
        rootNodes.forEach(function (node, index) {
          var originalValue = originalValues[index];

          if (originalValue === null) {
            node.removeAttribute('aria-hidden');
          } else {
            node.setAttribute('aria-hidden', originalValue);
          }
        });
      };
    };

    var cx = function cx() {
      for (var _len = arguments.length, classNames = new Array(_len), _key = 0; _key < _len; _key++) {
        classNames[_key] = arguments[_key];
      }

      return classNames.filter(Boolean).join(' ');
    };

    var ModalComponent =
    /*#__PURE__*/
    function (_Component) {
      inheritsLoose(ModalComponent, _Component);

      function ModalComponent() {
        var _this;

        for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
          args[_key2] = arguments[_key2];
        }

        _this = _Component.call.apply(_Component, [this].concat(args)) || this;
        _this.contentRef = React__default.createRef();
        _this.trap = null;
        _this.disposeAriaHider = null;

        _this.handleDialogRef = function (dialogNode) {
          if (dialogNode) {
            _this.disposeAriaHider = createAriaHider(dialogNode);
          } else {
            _this.disposeAriaHider();
          }
        };

        _this.onFocusLockActivation = function () {
          var initialFocusRef = _this.props.initialFocusRef;

          if (initialFocusRef && initialFocusRef.current) {
            initialFocusRef.current.focus();
          }
        };

        return _this;
      }

      var _proto = ModalComponent.prototype;

      _proto.render = function render() {
        var _this2 = this;

        var _this$props = this.props,
            className = _this$props.className,
            forwardedRef = _this$props.forwardedRef,
            as = _this$props.as,
            theme = _this$props.theme,
            opened = _this$props.opened,
            onClose = _this$props.onClose,
            children = _this$props.children,
            persistent = _this$props.persistent,
            onClick = _this$props.onClick,
            onKeyDown = _this$props.onKeyDown,
            initialFocusRef = _this$props.initialFocusRef,
            props = objectWithoutPropertiesLoose(_this$props, ["className", "forwardedRef", "as", "theme", "opened", "onClose", "children", "persistent", "onClick", "onKeyDown", "initialFocusRef"]);

        var timeout = modalTransitionDuration({
          theme: theme || {}
        });
        return React__default.createElement(Transition, {
          timeout: timeout,
          "in": opened
        }, function (status) {
          if (status === 'exited') return null;
          return React__default.createElement(Portal, null, React__default.createElement(FocusLock, {
            returnFocus: true,
            onActivation: _this2.onFocusLockActivation
          }, React__default.createElement(RemoveScroll, {
            forwardProps: true
          }, React__default.createElement("div", _extends_1({
            className: cx('sui-modal', "sui-modal-transition-" + status, className),
            ref: _this2.handleDialogRef,
            onClick: wrapEvent(onClick, function (event) {
              event.stopPropagation();
              onClose();
            }),
            onKeyDown: wrapEvent(onKeyDown, function (event) {
              if (event.key === 'Escape') {
                event.stopPropagation();
                onClose();
              }
            })
          }, props), React__default.createElement(ModalContext.Provider, {
            value: {
              contentRef: _this2.contentRef,
              onClose: onClose
            }
          }, children)))));
        });
      };

      return ModalComponent;
    }(React.Component);

    var ModalComponentWithTheme = styled.withTheme(ModalComponent);
    return ModalComponentWithTheme;
  }

  var Modal =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'modal',
      InnerComponent: createModalComponent(),
      style: function style(p) {
        return styled.css(["position:fixed;top:0;right:0;bottom:0;left:0;z-index:", ";visibility:visible;overflow-x:hidden;overflow-y:auto;opacity:0;outline:0;background-color:", ";", ";&.sui-modal-transition-entering{opacity:0;}&.sui-modal-transition-entered{opacity:1;}&.sui-modal-transition-exiting{opacity:0;}"], zIndexModal(p), modalBackdropBg(p), transition("opacity " + modalTransitionDuration(p) + "ms ease-in-out")(p));
      },
      propTypes: {
        children: PropTypes__default.node,
        opened: PropTypes__default.bool,
        onClose: PropTypes__default.func,
        initialFocusRef: PropTypes__default.object
      },
      defaultProps: {
        persistent: true
      }
    };
  });

  var ModalBody =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'modal-body',
      style: function style(p) {
        return styled.css(["position:relative;flex:1 1 auto;padding:", ";"], modalInnerPadding(p));
      },
      propTypes: {
        children: PropTypes__default.node
      }
    };
  });

  var ModalContent =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'modal-content',
      render: function render(_ref) {
        var Component = _ref.Component,
            _ref2 = _ref.ref,
            onClick = _ref.onClick,
            props = objectWithoutPropertiesLoose(_ref, ["Component", "ref", "onClick"]);

        return React__default.createElement(ModalContext.Consumer, null, function (modalContext) {
          return React__default.createElement(Component, _extends_1({
            tabIndex: "-1",
            ref: function ref(node) {
              if (modalContext) modalContext.contentRef.current = node;
              if (_ref2) _ref2(node);
            },
            onClick: wrapEvent(onClick, stopPropagation)
          }, props));
        });
      },
      style: function style(p) {
        return styled.css(["position:relative;display:flex;flex-direction:column;width:100%;pointer-events:auto;background-color:", ";background-clip:padding-box;border-style:solid;border-width:", ";border-color:", ";border-radius:", ";", ";outline:0;", ";"], modalContentBg(p), modalContentBorderWidth(p), modalContentBorderColor(p), modalContentBorderRadius(p), modalContentBoxShadowXs(p), up('sm', modalContentBoxShadowSmUp(p)));
      },
      propTypes: {
        children: PropTypes__default.node
      }
    };
  });

  var ModalCloseButton =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'modal-close-button',
      defaultComponent: 'button',
      render: function render(_ref) {
        var Component = _ref.Component,
            onClick = _ref.onClick,
            ref = _ref.ref,
            props = objectWithoutPropertiesLoose(_ref, ["Component", "onClick", "ref"]);

        return React__default.createElement(ModalContext.Consumer, null, function (modalContext) {
          return React__default.createElement(Component, _extends_1({
            ref: ref,
            onClick: wrapEvent(onClick, function () {
              if (modalContext) {
                modalContext.onClose();
              }
            })
          }, props), React__default.createElement("span", {
            "aria-hidden": "true"
          }, "\xD7"));
        });
      },
      style: function style(p) {
        return styled.css(["position:absolute;cursor:pointer;top:0.2rem;right:0.2rem;padding:0.3rem 0.8rem;margin:0;background-color:transparent;appearance:none;border:0;font-size:1.5rem;font-weight:700;line-height:1;border-radius:", ";opacity:0.5;", ";&:focus{opacity:1;", ";}&:hover{opacity:1;}"], borderRadius(p), transitionBase(p), baseFocus(primary(p))(p));
      },
      propTypes: {
        children: PropTypes__default.node
      }
    };
  });

  var ModalDialog =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'modal-dialog',
      render: function render(_ref) {
        var Component = _ref.Component,
            props = objectWithoutPropertiesLoose(_ref, ["Component"]);

        return React__default.createElement(Component, _extends_1({
          role: "dialog",
          "aria-modal": "true"
        }, props));
      },
      style: function style(p) {
        var margin = modalDialogMargin(p);
        var upSmMargin = modalDialogMarginYSmUp(p);
        return styled.css(["position:relative;width:auto;margin:", ";pointer-events:none;display:flex;align-items:center;min-height:calc(100% - ", ");", ";"], margin, calc(margin, function (x) {
          return x * 2;
        }), up('sm', styled.css(["max-width:500px;margin:", " auto;min-height:calc(100% - ", ");"], upSmMargin, calc(upSmMargin, function (x) {
          return x * 2;
        }))));
      },
      propTypes: {
        children: PropTypes__default.node
      }
    };
  });

  var ModalFooter =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'modal-footer',
      style: function style(p) {
        return styled.css(["display:flex;align-items:center;justify-content:flex-end;padding:", ";border-top-width:", ";border-top-style:solid;border-top-color:", ";>:not(:first-child){margin-left:0.25rem;}>:not(:last-child){margin-right:0.25rem;}"], modalInnerPadding(p), modalFooterBorderWidth(p), modalFooterBorderColor(p));
      },
      propTypes: {
        children: PropTypes__default.node
      }
    };
  });

  var ModalHeader =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'modal-header',
      style: function style(p) {
        return styled.css(["display:flex;align-items:flex-start;justify-content:space-between;padding:", ";border-bottom-style:solid;border-bottom-width:", ";border-bottom-color:", ";border-top-left-radius:", ";border-top-right-radius:", ";"], modalInnerPadding(p), modalHeaderBorderWidth(p), modalHeaderBorderColor(p), modalContentBorderRadius(p), modalContentBorderRadius(p));
      },
      propTypes: {
        children: PropTypes__default.node
      }
    };
  });

  function useRandomId(prefix) {
    return React.useMemo(function () {
      return "" + (prefix ? prefix + "-" : '') + Math.random().toString(36).substring(7);
    }, [prefix]);
  }

  function usePopper(containerRef, elementRef, _temp) {
    var _ref = _temp === void 0 ? {} : _temp,
        _ref$placement = _ref.placement,
        placement = _ref$placement === void 0 ? 'auto' : _ref$placement;

    React.useEffect(function () {
      var container = containerRef.current;
      var element = elementRef.current;
      if (!container || !element) return undefined;
      var popper = new PopperJs(container, element, {
        placement: placement
      });
      return function () {
        return popper.destroy();
      };
    }, [containerRef, elementRef, placement]);
  }
  var placements = ['auto', 'auto-start', 'auto-end', 'top', 'top-start', 'top-end', 'right', 'right-start', 'right-end', 'bottom', 'bottom-start', 'bottom-end', 'left', 'left-start', 'left-end'];

  function useSetAttribute(elementRef, attribute, value) {
    React.useEffect(function () {
      if (!elementRef.current) return;
      elementRef.current.setAttribute(attribute, value);
    }, [elementRef, attribute, value]);
  }

  function useAsyncRef(defaultValue, forwardedRef) {
    var _useState = React.useState({
      current: defaultValue
    }),
        ref = _useState[0],
        setRef = _useState[1];

    var handleRef = React.useCallback(function (current) {
      setRef({
        current: current
      });

      if (typeof forwardedRef === 'function') {
        forwardedRef(current);
      } else if (forwardedRef) {
        forwardedRef.current = current;
      }
    }, [setRef, forwardedRef]);
    return [ref, handleRef];
  }

  function useClickOutside(ref, handler) {
    var isTouchRef = React.useRef();
    var handlerRef = React.useRef();
    handlerRef.current = handler;
    var refs = React.useMemo(function () {
      return Array.isArray(ref) ? ref : ref;
    }, [ref]);
    React.useEffect(function () {
      function handleEvent(e) {
        if (e.type === 'touchend') {
          isTouchRef.current = true;
        }

        if (e.type === 'click' && isTouchRef.current) return;

        if (refs.every(function (_ref) {
          var el = _ref.current;
          return el && el !== e.target && !el.contains(e.target);
        })) {
          handlerRef.current(e);
        }
      }

      document.addEventListener('touchend', handleEvent, true);
      document.addEventListener('click', handleEvent, true);
      return function () {
        document.removeEventListener('touchend', handleEvent, true);
        document.removeEventListener('click', handleEvent, true);
      };
    }, [refs]);
  }

  var ParentElement =
  /*#__PURE__*/
  React.forwardRef(function (props, ref) {
    var _useState = React.useState(true),
        show = _useState[0],
        setShow = _useState[1];

    var divRef = React.useRef();
    React.useEffect(function () {
      ref.current = divRef.current.parentElement;
      setShow(false);
    }, [divRef, ref]);
    return show ? React__default.createElement("div", {
      ref: divRef
    }) : null;
  });

  function PopoverComponent(_ref) {
    var children = _ref.children,
        visible = _ref.visible,
        idProp = _ref.id,
        _ref$placement = _ref.placement,
        placement = _ref$placement === void 0 ? 'auto' : _ref$placement,
        ref = _ref.ref,
        Component = _ref.Component,
        _ref$hideOnClickOutsi = _ref.hideOnClickOutside,
        hideOnClickOutside = _ref$hideOnClickOutsi === void 0 ? true : _ref$hideOnClickOutsi,
        onHide = _ref.onHide,
        props = objectWithoutPropertiesLoose(_ref, ["children", "visible", "id", "placement", "ref", "Component", "hideOnClickOutside", "onHide"]);

    var parentRef = React.useRef();

    var _useAsyncRef = useAsyncRef(undefined, ref),
        popoverRef = _useAsyncRef[0],
        handlePopoverRef = _useAsyncRef[1];

    var randomId = useRandomId('tooltip');
    var id = idProp || randomId;
    usePopper(parentRef, popoverRef, {
      placement: placement
    });
    useSetAttribute(parentRef, 'aria-haspopup', true);
    useSetAttribute(parentRef, 'aria-expanded', visible);
    useSetAttribute(parentRef, 'aria-controls', id);
    useClickOutside([parentRef, popoverRef], function () {
      if (onHide && hideOnClickOutside) onHide();
    });
    return React__default.createElement(React__default.Fragment, null, React__default.createElement(ParentElement, {
      ref: parentRef
    }), React__default.createElement(Portal, null, React__default.createElement(Component, _extends_1({
      ref: handlePopoverRef,
      id: id,
      "aria-hidden": !visible,
      role: "group"
    }, props), children)));
  }

  var Popover =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'popover',
      render: PopoverComponent,
      defaultComponent: 'div',
      style: function style() {
        return styled.css(["&[aria-hidden='true']{pointer-events:none;visibility:hidden !important;}"]);
      },
      propTypes: {
        children: PropTypes__default.node,
        onHide: PropTypes__default.func,
        placement: PropTypes__default.oneOf(placements),
        visible: PropTypes__default.bool
      }
    };
  });

  var sizeStyle$2 = {
    sm:
    /*#__PURE__*/
    styled.css([".sui-radio-content{width:0.875rem;height:0.875rem;}.sui-radio-circle{width:8px;height:8px;}"]),
    md:
    /*#__PURE__*/
    styled.css([".sui-radio-content{width:1rem;height:1rem;}.sui-radio-circle{width:10px;height:10px;}"]),
    lg:
    /*#__PURE__*/
    styled.css([".sui-radio-content{width:1.25rem;height:1.25rem;}.sui-radio-circle{width:14px;height:14px;}"])
  };

  var validStyle$2 = function validStyle(p) {
    var valid = p.valid;
    if (valid !== true && valid !== false) return null;
    var color = valid ? success(p) : danger(p);
    return styled.css(["input + .sui-radio-content,input:checked + .sui-radio-content{border-color:", ";}input:checked + .sui-radio-content .sui-radio-circle{background-color:", ";}input:focus + .sui-radio-content{border-color:", ";", "}"], color, color, color, controlFocus(color)(p));
  };

  var controlStyle$2 = function controlStyle(p) {
    return styled.css(["input:focus + .sui-radio-content{", "}", ";"], controlFocus(primary(p))(p), validStyle$2(p));
  };

  var containerSystem$1 =
  /*#__PURE__*/
  compose(basics, dimensions, space, flexboxes, positions);
  var contentSystem$1 =
  /*#__PURE__*/
  compose(dimensions, backgrounds, borders);
  var system$2 =
  /*#__PURE__*/
  compose(containerSystem$1, contentSystem$1);
  var Radio =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'radio',
      system: system$2,
      applySystem: null,
      render: function render(_ref) {
        var Component = _ref.Component,
            ref = _ref.ref,
            className = _ref.className,
            size = _ref.size,
            control = _ref.control,
            valid = _ref.valid,
            props = objectWithoutPropertiesLoose(_ref, ["Component", "ref", "className", "size", "control", "valid"]);

        return React__default.createElement(SwitchState, props, function (_ref2) {
          var input = _ref2.input;
          return React__default.createElement(Component, {
            className: className
          }, React__default.createElement("input", _extends_1({
            ref: ref,
            type: "radio"
          }, input)), React__default.createElement("div", {
            className: "sui-radio-content"
          }, React__default.createElement("div", {
            className: "sui-radio-circle"
          })));
        });
      },
      style: function style(p) {
        return styled.css(["display:inline-flex;align-items:center;justify-content:center;position:relative;width:1.5rem;height:1.5rem;position:relative;.sui-radio-content{display:flex;flex-shrink:0;align-items:center;justify-content:center;border-radius:50%;background-color:", ";border-width:", ";border-style:solid;border-color:", ";", ";}input:checked + .sui-radio-content{border-color:", ";.sui-radio-circle{transform:scale(1);}}input:focus + .sui-radio-content{", ";}input:disabled + .sui-radio-content{background-color:", ";}.sui-radio-circle{", ";border-radius:50%;background-color:", ";transform:scale(0);}", ";", ";.sui-radio-content{", ";}", ";"], inputBgColor(p), inputBorderWidth(p), inputBorderColor(p), transitionBase(p), primary(p), baseFocus(primary(p))(p), inputDisabledBgColor(p), transitionBase(p), primary(p), p.size && sizeStyle$2[p.size], containerSystem$1.props, contentSystem$1.props, p.control && controlStyle$2(p));
      },
      propTypes: {
        control: PropTypes__default.bool,
        checked: PropTypes__default.bool,
        disabled: PropTypes__default.bool,
        onChange: PropTypes__default.func,
        size: PropTypes__default.oneOf(['sm', 'md', 'lg']),
        valid: PropTypes__default.bool,
        value: PropTypes__default.string
      },
      defaultProps: {
        size: 'md'
      }
    };
  });

  function create$3() {
    var RadioGroup =
    /*#__PURE__*/
    function (_React$Component) {
      inheritsLoose(RadioGroup, _React$Component);

      function RadioGroup() {
        var _this;

        for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }

        _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
        _this.controls = [];
        return _this;
      }

      var _proto = RadioGroup.prototype;

      _proto.getChildContext = function getChildContext() {
        return {
          suiGroup: this
        };
      };

      _proto.register = function register(control) {
        this.controls.push(control);
      };

      _proto.unregister = function unregister(control) {
        var index = this.controls.indexOf(control);
        if (index !== -1) this.controls.splice(index, 1);
      };

      _proto.notify = function notify(event) {
        this.controls.forEach(function (control) {
          if (control.props.checked === undefined && control.props.name === event.target.name && control.props.value !== event.target.value) {
            control.updateState({
              checked: false
            });
          }
        });
      };

      _proto.render = function render() {
        return this.props.children;
      };

      return RadioGroup;
    }(React__default.Component);

    RadioGroup.propTypes = {
      children: PropTypes__default.node
    };
    RadioGroup.childContextTypes = {
      suiGroup: PropTypes__default.object
    };
    return RadioGroup;
  }

  var RadioGroup = /*#__PURE__*/
  create$3();

  var Row =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'row',
      omitProps: ['gutter'],
      style: function style(p) {
        var _p$gutter = p.gutter,
            gutter = _p$gutter === void 0 ? gridGutter(p) : _p$gutter;
        var negativeGutter = negative(gutter);
        return styled.css(["box-sizing:border-box;flex-grow:1;flex-wrap:wrap;display:flex;", " ", ""], ml({
          ml: negativeGutter
        })(p), mr({
          mr: negativeGutter
        })(p));
      },
      propTypes: {
        children: PropTypes__default.node,
        gutter: getSystemPropTypes(ml).ml
      }
    };
  });

  var sizeStyle$3 = {
    sm: function sm(p) {
      return styled.css(["select{padding:", " ", ";font-size:", ";line-height:", ";border-radius:", ";", ";}.sui-select-arrow{right:0.35rem;width:0.525rem;}"], inputPaddingYSm(p), inputPaddingXSm(p), fontSizeSm(p), inputLineHeightSm(p), borderRadiusSm(p), p.arrow && !p.multiple && 'padding-right: 1.225rem;');
    },
    md: function md(p) {
      return styled.css(["select{padding:", " ", ";font-size:", ";line-height:", ";", ";border-radius:", ";}.sui-select-arrow{right:0.5rem;width:0.6rem;}"], inputPaddingY(p), inputPaddingX(p), fontSizeBase(p), inputLineHeight(p), p.arrow && !p.multiple && 'padding-right: 1.6rem;', borderRadius(p));
    },
    lg: function lg(p) {
      return styled.css(["select{padding:", " ", ";font-size:", ";line-height:", ";border-radius:", ";", ";}.sui-select-arrow{right:0.625rem;width:0.75rem;}"], inputPaddingYLg(p), inputPaddingXLg(p), fontSizeLg(p), inputLineHeightLg(p), borderRadiusLg(p), p.arrow && !p.multiple && 'padding-right: 2rem;');
    }
  };

  var validStyle$3 = function validStyle(p) {
    var valid = p.valid;
    if (valid !== true && valid !== false) return null;
    var color = valid ? success(p) : danger(p);
    return styled.css(["select{border-color:", ";&:focus{border-color:", ";", "}}"], color, color, controlFocus(color)(p));
  };

  var controlStyle$3 = function controlStyle(p) {
    return styled.css(["&,select{display:block;width:100%;}select:focus{", ";}", ";"], controlFocus(primary(p))(p), validStyle$3(p));
  };

  var Arrow = function Arrow() {
    return React__default.createElement("svg", {
      className: "sui-select-arrow",
      viewBox: "0 0 10 5"
    }, React__default.createElement("g", {
      fill: "none",
      fillRule: "evenodd"
    }, React__default.createElement("path", {
      d: "M17 14H-7v-24h24"
    }), React__default.createElement("path", {
      fill: "currentColor",
      opacity: 0.5,
      d: "M0 0l5 5 5-5"
    })));
  };

  var Select =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'select',
      render: function render(_ref) {
        var Component = _ref.Component,
            arrow = _ref.arrow,
            control = _ref.control,
            className = _ref.className,
            size = _ref.size,
            valid = _ref.valid,
            children = _ref.children,
            props = objectWithoutPropertiesLoose(_ref, ["Component", "arrow", "control", "className", "size", "valid", "children"]);

        return React__default.createElement(Component, {
          className: className
        }, arrow && !props.multiple ? React__default.createElement(Arrow, null) : null, React__default.createElement("select", props, children));
      },
      style: function style(p) {
        return styled.css(["display:inline-block;position:relative;select{appearance:none;-webkit-appearance:none;-moz-appearance:none;-webkit-border-radius:0;display:inline-block;border-width:", ";border-color:", ";border-style:solid;line-height:", ";", ";color:", ";&:focus{", ";}&:disabled{background-color:", ";color:", ";}}.sui-select-arrow{position:absolute;top:50%;pointer-events:none;}", ";", ";"], inputBorderWidth(p), inputBorderColor(p), inputLineHeight(p), transitionBase(p), inputTextColor(p), baseFocus(primary(p))(p), inputDisabledBgColor(p), inputDisabledText(p), p.size && sizeStyle$3[p.size] && sizeStyle$3[p.size](p), p.control && controlStyle$3(p));
      },
      propTypes: {
        arrow: PropTypes__default.bool,
        control: PropTypes__default.bool,
        options: PropTypes__default.array,
        size: PropTypes__default.oneOf(['sm', 'md', 'lg']),
        valid: PropTypes__default.bool,
        children: PropTypes__default.node
      },
      defaultProps: {
        arrow: true,
        size: 'md'
      }
    };
  });

  var Switch =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'switch',
      system: system,
      applySystem: null,
      render: function render(_ref) {
        var Component = _ref.Component,
            ref = _ref.ref,
            className = _ref.className,
            _ref$onLabel = _ref.onLabel,
            onLabel = _ref$onLabel === void 0 ? 'ON' : _ref$onLabel,
            _ref$offLabel = _ref.offLabel,
            offLabel = _ref$offLabel === void 0 ? 'OFF' : _ref$offLabel,
            labeled = _ref.labeled,
            props = objectWithoutPropertiesLoose(_ref, ["Component", "ref", "className", "onLabel", "offLabel", "labeled"]);

        return React__default.createElement(SwitchState, props, function (_ref2) {
          var input = _ref2.input;
          return React__default.createElement(Component, {
            className: className
          }, React__default.createElement("input", _extends_1({
            ref: ref,
            type: "checkbox"
          }, input)), React__default.createElement("div", {
            className: "sui-switch-wrapper"
          }, React__default.createElement("div", {
            className: "sui-switch-content"
          }, React__default.createElement("span", {
            className: "sui-switch-label sui-switch-label-on"
          }, labeled ? onLabel : ''), React__default.createElement("div", {
            className: "sui-switch-ball"
          }), React__default.createElement("span", {
            className: "sui-switch-label sui-switch-label-off"
          }, labeled ? offLabel : ''))));
        });
      },
      style: function style(p) {
        return styled.css(["display:inline-block;position:relative;width:50px;height:24px;.sui-switch-wrapper{width:50px;height:24px;border-radius:34px;background-color:", ";overflow:hidden;cursor:pointer;border-width:", ";border-color:", ";border-style:solid;font-size:9px;font-weight:800;", ";}input:focused + .sui-switch-wrapper{", ";}input:checked + .sui-switch-wrapper{background-color:", ";border-color:transparent;.sui-switch-content{transform:translateX(0);}.sui-switch-ball{background-color:", ";}}input:disabled + .sui-switch-wrapper{opacity:0.5;}.sui-switch-ball{flex-shrink:0;background-color:", ";border-radius:50%;width:18px;height:18px;", ";}.sui-switch-content{display:flex;align-items:center;height:22px;transform:translateX(-25px);", ";}.sui-switch-label{flex-shrink:0;width:27px;text-align:center;user-select:none;&.sui-switch-label-on{color:", ";}&.sui-switch-label-off{color:", ";}}.sui-switch-wrapper{", ";}"], gray300(p), inputBorderWidth(p), inputBorderColor(p), transitionBase(p), baseFocus(primary(p))(p), primaryLight(p), primary(p), gray500(p), transitionBase(p), transitionBase(p), primary(p), gray900(p), system.props);
      },
      propTypes: {
        checked: PropTypes__default.bool,
        disabled: PropTypes__default.bool,
        labeled: PropTypes__default.bool,
        onLabel: PropTypes__default.string,
        offLabel: PropTypes__default.string,
        onChange: PropTypes__default.func
      }
    };
  });

  function createTextarea() {
    var Textarea = Input.withComponent('textarea');
    Textarea.propTypes = Input.propTypes;
    return Textarea;
  }

  var Textarea = /*#__PURE__*/
  createTextarea();

  function useToggle(defaultToggled, onToggle) {
    if (defaultToggled === void 0) {
      defaultToggled = false;
    }

    var onToggleRef = React.useRef();
    onToggleRef.current = onToggle;

    var _useState = React.useState(defaultToggled),
        toggled = _useState[0],
        setToggled = _useState[1];

    var toggle = React.useCallback(function (toggled) {
      return setToggled(toggled === undefined ? function (previousToggled) {
        return !previousToggled;
      } : toggled);
    }, [setToggled]);
    React.useEffect(function () {
      if (onToggleRef.current) {
        onToggleRef.current(toggled);
      }
    }, [onToggleRef, toggled]);
    return [toggled, toggle];
  }

  function create$4() {
    function Toggler(_ref) {
      var children = _ref.children,
          defaultToggled = _ref.defaultToggled,
          onToggleProp = _ref.onToggle;
      return children(useToggle(defaultToggled, onToggleProp));
    }

    Toggler.propTypes = {
      children: PropTypes__default.func.isRequired,
      defaultToggled: PropTypes__default.bool,
      onToggle: PropTypes__default.func
    };
    return Toggler;
  }

  var Toggler = /*#__PURE__*/
  create$4();

  function useEventListener(targetRef, event, listener) {
    var listenerRef = React.useRef();
    listenerRef.current = listener;
    React.useEffect(function () {
      if (!targetRef.current) return undefined;
      var target = targetRef.current;

      function handleEvent(event) {
        return listenerRef.current(event);
      }

      target.addEventListener(event, handleEvent);
      return function () {
        return target.removeEventListener(event, handleEvent);
      };
    }, [event, targetRef, listenerRef]);
  }

  function TooltipComponent(_ref) {
    var children = _ref.children,
        idProp = _ref.id,
        _ref$placement = _ref.placement,
        placement = _ref$placement === void 0 ? 'top' : _ref$placement,
        ref = _ref.ref,
        Component = _ref.Component,
        props = objectWithoutPropertiesLoose(_ref, ["children", "id", "placement", "ref", "Component"]);

    var parentRef = React.useRef();

    var _useState = React.useState(false),
        visible = _useState[0],
        setVisible = _useState[1];

    var _useAsyncRef = useAsyncRef(undefined, ref),
        tooltipRef = _useAsyncRef[0],
        handleTooltipRef = _useAsyncRef[1];

    var randomId = useRandomId('tooltip');
    var id = idProp || randomId;

    var makeVisible = function makeVisible() {
      return setVisible(true);
    };

    var makeHidden = function makeHidden() {
      return setVisible(false);
    };

    usePopper(parentRef, tooltipRef, {
      placement: placement
    });
    useSetAttribute(parentRef, 'aria-describedby', id);
    useEventListener(parentRef, 'focus', makeVisible);
    useEventListener(parentRef, 'blur', makeHidden);
    useEventListener(parentRef, 'mouseover', makeVisible);
    useEventListener(parentRef, 'mouseout', makeHidden);
    return React__default.createElement(React__default.Fragment, null, React__default.createElement(ParentElement, {
      ref: parentRef
    }), React__default.createElement(Portal, null, React__default.createElement(Component, _extends_1({
      ref: handleTooltipRef,
      id: id,
      "aria-hidden": !visible,
      role: "tooltip"
    }, props), children)));
  }

  var Tooltip =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'tooltip',
      defaultComponent: 'div',
      render: TooltipComponent,
      style: function style(p) {
        return styled.css(["&[aria-hidden='true']{pointer-events:none;visibility:hidden !important;}margin:0.5em;font-size:0.875em;text-align:center;padding:0.5em 0.75em;color:", ";background-color:", ";border-radius:", ";"], white(p), black(p), borderRadius(p));
      },
      propTypes: {
        children: PropTypes__default.node,
        placement: PropTypes__default.oneOf(placements)
      }
    };
  });

  var variantTags = {
    h1: 'h1',
    h2: 'h2',
    h3: 'h3',
    h4: 'h4',
    h5: 'h5',
    h6: 'h6',
    'display-1': 'h1',
    'display-2': 'h2',
    'display-3': 'h3',
    'display-4': 'h4'
  };

  var commonHeadingStyle = function commonHeadingStyle(p) {
    var fontFamily = headingsFontFamily(p);
    return styled.css(["margin-top:0;margin-bottom:", ";font-weight:", ";line-height:", ";color:", ";", ";"], headingsMarginBottom(p), headingsFontWeight(p), headingsLineHeight(p), headingsColor(p), fontFamily && "font-family: " + fontFamily);
  };

  var variantStyle = {
    h1: function h1(p) {
      return styled.css(["", ";font-size:", ";"], commonHeadingStyle(p), h1FontSize(p));
    },
    h2: function h2(p) {
      return styled.css(["", ";font-size:", ";"], commonHeadingStyle(p), h2FontSize(p));
    },
    h3: function h3(p) {
      return styled.css(["", ";font-size:", ";"], commonHeadingStyle(p), h3FontSize(p));
    },
    h4: function h4(p) {
      return styled.css(["", ";font-size:", ";"], commonHeadingStyle(p), h4FontSize(p));
    },
    h5: function h5(p) {
      return styled.css(["", ";font-size:", ";"], commonHeadingStyle(p), h5FontSize(p));
    },
    h6: function h6(p) {
      return styled.css(["", ";font-size:", ";"], commonHeadingStyle(p), h6FontSize(p));
    },
    'display-1': function display1(p) {
      return styled.css(["", ";font-size:", ";font-weight:", ";line-height:", ";"], commonHeadingStyle(p), display1Size(p), display1Weight(p), displayLineHeight(p));
    },
    'display-2': function display2(p) {
      return styled.css(["", ";font-size:", ";font-weight:", ";line-height:", ";"], commonHeadingStyle(p), display2Size(p), display2Weight(p), displayLineHeight(p));
    },
    'display-3': function display3(p) {
      return styled.css(["", ";font-size:", ";font-weight:", ";line-height:", ";"], commonHeadingStyle(p), display3Size(p), display3Weight(p), displayLineHeight(p));
    },
    'display-4': function display4(p) {
      return styled.css(["", ";font-size:", ";font-weight:", ";line-height:", ";"], commonHeadingStyle(p), display4Size(p), display4Weight(p), displayLineHeight(p));
    }
  };
  var Typography =
  /*#__PURE__*/
  createComponent(function () {
    return {
      name: 'typo',
      defaultComponent: null,
      render: function render(_ref) {
        var className = _ref.className,
            BaseComponent = _ref.Component,
            variant = _ref.variant,
            props = objectWithoutPropertiesLoose(_ref, ["className", "Component", "variant"]);

        var Component = BaseComponent || variantTags[variant] || 'span';
        return React__default.createElement(Component, _extends_1({
          className: className
        }, props));
      },
      style: function style(p) {
        return styled.css(["", ";"], p.variant && variantStyle[p.variant] && variantStyle[p.variant](p));
      },
      propTypes: {
        children: PropTypes__default.node,
        variant: PropTypes__default.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'display-1', 'display-2', 'display-3', 'display-4'])
      }
    };
  });

  var theme = /*#__PURE__*/
  definitionsToTheme(theming);

  Object.keys(styled).forEach(function (key) {
    Object.defineProperty(exports, key, {
      enumerable: true,
      get: function () {
        return styled[key];
      }
    });
  });
  exports.styled = styled__default;
  exports.Alert = Alert;
  exports.Box = Box;
  exports.Breakpoint = Breakpoint;
  exports.Button = Button;
  exports.Checkbox = Checkbox;
  exports.Col = Col;
  exports.ControlFeedback = ControlFeedback;
  exports.FormCheck = FormCheck;
  exports.FormCheckLabel = FormCheckLabel;
  exports.FormGroup = FormGroup;
  exports.Grid = Grid;
  exports.Input = Input;
  exports.Label = Label;
  exports.Modal = Modal;
  exports.ModalBody = ModalBody;
  exports.ModalCloseButton = ModalCloseButton;
  exports.ModalContent = ModalContent;
  exports.ModalContext = ModalContext;
  exports.ModalDialog = ModalDialog;
  exports.ModalFooter = ModalFooter;
  exports.ModalHeader = ModalHeader;
  exports.Normalize = Normalize;
  exports.Popover = Popover;
  exports.Portal = Portal;
  exports.Radio = Radio;
  exports.RadioGroup = RadioGroup;
  exports.Row = Row;
  exports.Select = Select;
  exports.Switch = Switch;
  exports.SwitchState = SwitchState;
  exports.Textarea = Textarea;
  exports.Toggler = Toggler;
  exports.Tooltip = Tooltip;
  exports.Transition = Transition;
  exports.Typography = Typography;
  exports.alertBgLevel = alertBgLevel;
  exports.alertBorderLevel = alertBorderLevel;
  exports.alertColorLevel = alertColorLevel;
  exports.alertMarginBottom = alertMarginBottom;
  exports.alertPaddingX = alertPaddingX;
  exports.alertPaddingY = alertPaddingY;
  exports.alertVariant = alertVariant;
  exports.baseFocus = baseFocus;
  exports.between = between;
  exports.black = black;
  exports.blue = blue;
  exports.borderRadius = borderRadius;
  exports.borderRadiusLg = borderRadiusLg;
  exports.borderRadiusSm = borderRadiusSm;
  exports.borderWidth = borderWidth;
  exports.breakpoints = breakpoints;
  exports.brick = brick;
  exports.btnBorderWidth = btnBorderWidth;
  exports.btnDisabledOpacity = btnDisabledOpacity;
  exports.btnLineHeight = btnLineHeight;
  exports.btnLineHeightLg = btnLineHeightLg;
  exports.btnLineHeightSm = btnLineHeightSm;
  exports.btnPaddingX = btnPaddingX;
  exports.btnPaddingXLg = btnPaddingXLg;
  exports.btnPaddingXSm = btnPaddingXSm;
  exports.btnPaddingY = btnPaddingY;
  exports.btnPaddingYLg = btnPaddingYLg;
  exports.btnPaddingYSm = btnPaddingYSm;
  exports.btnVariant = btnVariant;
  exports.calc = calc;
  exports.cascade = cascade;
  exports.colorInterval = colorInterval;
  exports.colorLevel = colorLevel;
  exports.colorVariant = colorVariant;
  exports.colorVariants = colorVariants;
  exports.colorYik = colorYik;
  exports.colors = colors;
  exports.controlFocus = controlFocus;
  exports.cyan = cyan;
  exports.danger = danger;
  exports.dark = dark;
  exports.definitionsToTheme = definitionsToTheme;
  exports.display1Size = display1Size;
  exports.display1Weight = display1Weight;
  exports.display2Size = display2Size;
  exports.display2Weight = display2Weight;
  exports.display3Size = display3Size;
  exports.display3Weight = display3Weight;
  exports.display4Size = display4Size;
  exports.display4Weight = display4Weight;
  exports.displayLineHeight = displayLineHeight;
  exports.down = down;
  exports.fontFamily = fontFamily;
  exports.fontSizeBase = fontSizeBase;
  exports.fontSizeLg = fontSizeLg;
  exports.fontSizeSm = fontSizeSm;
  exports.fontSizes = fontSizes;
  exports.fontWeightBold = fontWeightBold;
  exports.fontWeightLight = fontWeightLight;
  exports.fontWeightNormal = fontWeightNormal;
  exports.func = func;
  exports.get = get;
  exports.getTheme = getTheme;
  exports.gray100 = gray100;
  exports.gray200 = gray200;
  exports.gray300 = gray300;
  exports.gray400 = gray400;
  exports.gray500 = gray500;
  exports.gray600 = gray600;
  exports.gray700 = gray700;
  exports.gray800 = gray800;
  exports.gray900 = gray900;
  exports.green = green;
  exports.gridColumns = gridColumns;
  exports.gridGutter = gridGutter;
  exports.gridMaxWidths = gridMaxWidths;
  exports.h1FontSize = h1FontSize;
  exports.h2FontSize = h2FontSize;
  exports.h3FontSize = h3FontSize;
  exports.h4FontSize = h4FontSize;
  exports.h5FontSize = h5FontSize;
  exports.h6FontSize = h6FontSize;
  exports.headingsColor = headingsColor;
  exports.headingsFontFamily = headingsFontFamily;
  exports.headingsFontWeight = headingsFontWeight;
  exports.headingsLineHeight = headingsLineHeight;
  exports.headingsMarginBottom = headingsMarginBottom;
  exports.indigo = indigo;
  exports.info = info;
  exports.inputBgColor = inputBgColor;
  exports.inputBorderColor = inputBorderColor;
  exports.inputBorderWidth = inputBorderWidth;
  exports.inputBtnLineHeight = inputBtnLineHeight;
  exports.inputBtnLineHeightLg = inputBtnLineHeightLg;
  exports.inputBtnLineHeightSm = inputBtnLineHeightSm;
  exports.inputBtnPaddingX = inputBtnPaddingX;
  exports.inputBtnPaddingXLg = inputBtnPaddingXLg;
  exports.inputBtnPaddingXSm = inputBtnPaddingXSm;
  exports.inputBtnPaddingY = inputBtnPaddingY;
  exports.inputBtnPaddingYLg = inputBtnPaddingYLg;
  exports.inputBtnPaddingYSm = inputBtnPaddingYSm;
  exports.inputDisabledBgColor = inputDisabledBgColor;
  exports.inputDisabledText = inputDisabledText;
  exports.inputLineHeight = inputLineHeight;
  exports.inputLineHeightLg = inputLineHeightLg;
  exports.inputLineHeightSm = inputLineHeightSm;
  exports.inputPaddingX = inputPaddingX;
  exports.inputPaddingXLg = inputPaddingXLg;
  exports.inputPaddingXSm = inputPaddingXSm;
  exports.inputPaddingY = inputPaddingY;
  exports.inputPaddingYLg = inputPaddingYLg;
  exports.inputPaddingYSm = inputPaddingYSm;
  exports.inputPlaceholderText = inputPlaceholderText;
  exports.inputTextColor = inputTextColor;
  exports.is = is;
  exports.light = light;
  exports.lineHeightBase = lineHeightBase;
  exports.lineHeightLg = lineHeightLg;
  exports.lineHeightSm = lineHeightSm;
  exports.mediaBetweenWidth = mediaBetweenWidth;
  exports.mediaMaxWidth = mediaMaxWidth;
  exports.mediaMinWidth = mediaMinWidth;
  exports.mixin = mixin;
  exports.modalBackdropBg = modalBackdropBg;
  exports.modalContentBg = modalContentBg;
  exports.modalContentBorderColor = modalContentBorderColor;
  exports.modalContentBorderRadius = modalContentBorderRadius;
  exports.modalContentBorderWidth = modalContentBorderWidth;
  exports.modalContentBoxShadowSmUp = modalContentBoxShadowSmUp;
  exports.modalContentBoxShadowXs = modalContentBoxShadowXs;
  exports.modalDialogMargin = modalDialogMargin;
  exports.modalDialogMarginYSmUp = modalDialogMarginYSmUp;
  exports.modalFooterBorderColor = modalFooterBorderColor;
  exports.modalFooterBorderWidth = modalFooterBorderWidth;
  exports.modalHeaderBorderColor = modalHeaderBorderColor;
  exports.modalHeaderBorderWidth = modalHeaderBorderWidth;
  exports.modalInnerPadding = modalInnerPadding;
  exports.modalTransitionDuration = modalTransitionDuration;
  exports.negative = negative;
  exports.num = num;
  exports.omit = omit;
  exports.orange = orange;
  exports.pink = pink;
  exports.primary = primary;
  exports.primaryLight = primaryLight;
  exports.purple = purple;
  exports.px = px;
  exports.red = red;
  exports.secondary = secondary;
  exports.secondaryLight = secondaryLight;
  exports.spaces = spaces;
  exports.stopPropagation = stopPropagation;
  exports.string = string;
  exports.success = success;
  exports.system = universalSystem;
  exports.teal = teal;
  exports.th = th;
  exports.thd = thd;
  exports.theme = theme;
  exports.transition = transition;
  exports.transitionBase = transitionBase;
  exports.transitionDuration = transitionDuration;
  exports.transitionEasingFunc = transitionEasingFunc;
  exports.transitionEnabled = transitionEnabled;
  exports.unit = unit;
  exports.up = up;
  exports.useAsyncRef = useAsyncRef;
  exports.useClickOutside = useClickOutside;
  exports.useEventListener = useEventListener;
  exports.useNode = useNode;
  exports.usePopper = usePopper;
  exports.useRandomId = useRandomId;
  exports.useSetAttribute = useSetAttribute;
  exports.useToggle = useToggle;
  exports.warning = warning;
  exports.white = white;
  exports.wrapEvent = wrapEvent;
  exports.yellow = yellow;
  exports.yikTextDark = yikTextDark;
  exports.yikTextLight = yikTextLight;
  exports.yiqContrastedThreshold = yiqContrastedThreshold;
  exports.zIndexControl = zIndexControl;
  exports.zIndexInnerSwitch = zIndexInnerSwitch;
  exports.zIndexModal = zIndexModal;
  exports.zIndexModalBackdrop = zIndexModalBackdrop;

  Object.defineProperty(exports, '__esModule', { value: true });

}));
